{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-24T15:07:44.345071Z",
     "start_time": "2021-02-24T15:07:32.549360Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spark version:  3.0.0\n",
      "+---------------------------------+---------+-------------------+-------------------------------------+----------------------------+--------------+--------------+----------------+------+------------------------------+---------------+----------------+-------------+---------------+---------+----------+---------+-------------------+----------+-------------------------------------+----------------------+--------------+--------------+---------------+----------------------------+-------------+--------------------+--------+------------------+------------+------------------------+----------+---------------------+----------+--------------------------------+--------------------+----------------+-------------------+-----------------------+--------+-------------+-----------------------+-------------+-----------+-----------------+-----------+----------+------------+---------------+-------+---------------+--------------------+--------------------+---------+\n",
      "|biologicalModelAllelicComposition|variantId|diseaseFromSourceId|diseaseModelAssociatedModelPhenotypes|variantAminoacidDescriptions|clinicalStatus|mutatedSamples|cohortPhenotypes|drugId|variantFunctionalConsequenceId|publicationYear|targetModulation|resourceScore|cohortShortName|pathwayId|literature|oddsRatio|log2FoldChangeValue|confidence|diseaseModelAssociatedHumanPhenotypes|publicationFirstAuthor|pValueExponent|pValueMantissa|       targetId|log2FoldChangePercentileRank|clinicalPhase| textMiningSentences|cohortId|targetFromSourceId|datasourceId|significantDriverMethods|reactionId|clinicalSignificances|datatypeId|biologicalModelGeneticBackground|   diseaseFromSource|diseaseCellLines|allelicRequirements|confidenceIntervalLower|contrast|studyOverview|confidenceIntervalUpper|targetInModel|pathwayName|cohortDescription|variantRsId|studyCases|clinicalUrls|      diseaseId|studyId|studySampleSize|                  id|               score| sourceId|\n",
      "+---------------------------------+---------+-------------------+-------------------------------------+----------------------------+--------------+--------------+----------------+------+------------------------------+---------------+----------------+-------------+---------------+---------+----------+---------+-------------------+----------+-------------------------------------+----------------------+--------------+--------------+---------------+----------------------------+-------------+--------------------+--------+------------------+------------+------------------------+----------+---------------------+----------+--------------------------------+--------------------+----------------+-------------------+-----------------------+--------+-------------+-----------------------+-------------+-----------+-----------------+-----------+----------+------------+---------------+-------+---------------+--------------------+--------------------+---------+\n",
      "|                             null|     null|        EFO_0000764|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.2|           null|     null| [9226175]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000055955|                        null|         null|[[115, 88, abstra...|    null|            Q14624|   europepmc|                    null|      null|                 null|literature|                            null|human immunodefic...|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0000764|   null|           null|0000d2895be2e1a7a...|0.052000000000000005|europepmc|\n",
      "|                             null|     null|         MP_0001845|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          7.2|           null|     null|[26001794]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000102265|                        null|         null|[[125, 114, abstr...|    null|            P01033|   europepmc|                    null|      null|                 null|literature|                            null|        inflammation|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|     MP_0001845|   null|           null|0004f7493fab4dacb...| 0.07200000000000001|europepmc|\n",
      "|                             null|     null|    Orphanet_276525|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          3.6|           null|     null|[27057885]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000157005|                        null|         null|[[55, 40, other, ...|    null|            P61278|   europepmc|                    null|      null|                 null|literature|                            null|    nesidioblastosis|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|Orphanet_276525|   null|           null|00058eed3dcff9a1b...|0.036000000000000004|europepmc|\n",
      "|                             null|     null|    Orphanet_101957|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.4|           null|     null|[18291347]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000064835|                        null|         null|[[280, 259, abstr...|    null|            P28069|   europepmc|                    null|      null|                 null|literature|                            null|pituitary deficie...|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|Orphanet_101957|   null|           null|000890ad2665a8870...|0.054000000000000006|europepmc|\n",
      "|                             null|     null|      MONDO_0001721|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.8|           null|     null|[25888258]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000142405|                        null|         null|[[65, 57, abstrac...|    null|            P59046|   europepmc|                    null|      null|                 null|literature|                            null|           intrinsic|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|  MONDO_0001721|   null|           null|00096e9016f739659...|0.057999999999999996|europepmc|\n",
      "|                             null|     null|        EFO_0004264|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          6.6|           null|     null| [9562351]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000007908|                        null|         null|[[375, 360, abstr...|    null|            P16581|   europepmc|                    null|      null|                 null|literature|                            null|    vascular disease|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0004264|   null|           null|000b7d4fb16d405b3...|               0.066|europepmc|\n",
      "|                             null|     null|        EFO_0006788|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.4|           null|     null|[32439851]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000112715|                        null|         null|[[71, 65, abstrac...|    null|            P15692|   europepmc|                    null|      null|                 null|literature|                            null|             anxiety|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0006788|   null|           null|000cd34fe7146ca26...|0.054000000000000006|europepmc|\n",
      "|                             null|     null|        EFO_0000546|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.4|           null|     null|[26107236]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000100985|                        null|         null|[[122, 118, abstr...|    null|            P14780|   europepmc|                    null|      null|                 null|literature|                            null|               wound|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0000546|   null|           null|000e1b4a4c313f828...|0.054000000000000006|europepmc|\n",
      "|                             null|     null|        EFO_0010649|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.2|           null|     null|[11903595]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000110448|                        null|         null|[[18, 4, abstract...|    null|            P06127|   europepmc|                    null|      null|                 null|literature|                            null|     immunophenotype|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0010649|   null|           null|000f28402ad2366df...|0.052000000000000005|europepmc|\n",
      "|                             null|     null|        EFO_0000673|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|         25.2|           null|     null|[21882048]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000172201|                        null|         null|[[79, 65, title, ...|    null|            P47928|   europepmc|                    null|      null|                 null|literature|                            null|     prostate cancer|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0000673|   null|           null|000f96fc748fd3fa3...|               0.252|europepmc|\n",
      "|                             null|     null|      MONDO_0015175|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.2|           null|     null| [7687985]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000070031|                        null|         null|[[120, 118, abstr...|    null|            P09683|   europepmc|                    null|      null|                 null|literature|                            null|                 AIP|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|  MONDO_0015175|   null|           null|0013dff0df642e47f...|0.052000000000000005|europepmc|\n",
      "|                             null|     null|        EFO_0002427|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          6.8|           null|     null|[10627439]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000115884|                        null|         null|[[48, 42, abstrac...|    null|            P18827|   europepmc|                    null|      null|                 null|literature|                            null|             myeloma|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0002427|   null|           null|001563d859891ba0c...|               0.068|europepmc|\n",
      "|                             null|     null|        EFO_0000731|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|         13.4|           null|     null|[11035987]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000152661|                        null|         null|[[75, 59, title, ...|    null|            P17302|   europepmc|                    null|      null|                 null|literature|                            null|   uterine leiomyoma|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0000731|   null|           null|0015ded4e2838c96a...|               0.134|europepmc|\n",
      "|                             null|     null|        EFO_0005741|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          4.2|           null|     null|[25803100]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000133216|                        null|         null|[[358, 340, other...|    null|            P29323|   europepmc|                    null|      null|                 null|literature|                            null| infectious diseases|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0005741|   null|           null|001672873ad031b84...|               0.042|europepmc|\n",
      "|                             null|     null|        EFO_0000612|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.8|           null|     null|[21092651]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000164867|                        null|         null|[[17, 0, abstract...|    null|            P29474|   europepmc|                    null|      null|                 null|literature|                            null|  Myocardial infarct|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0000612|   null|           null|001697fb5adf3b15b...|0.057999999999999996|europepmc|\n",
      "|                             null|     null|        EFO_0005856|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          2.0|           null|     null|[32332732]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000197822|                        null|         null|[[129, 121, other...|    null|            Q16625|   europepmc|                    null|      null|                 null|literature|                            null|           arthritis|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0005856|   null|           null|0018681a26213b917...|                0.02|europepmc|\n",
      "|                             null|     null|        EFO_0004288|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          6.2|           null|     null|[28096468]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000147162|                        null|         null|[[32, 22, abstrac...|    null|            O15294|   europepmc|                    null|      null|                 null|literature|                            null|         colon tumor|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0004288|   null|           null|001a21dbaf324e853...|               0.062|europepmc|\n",
      "|                             null|     null|      MONDO_0007254|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.6|           null|     null|[18973230]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000066468|                        null|         null|[[62, 50, abstrac...|    null|            P21802|   europepmc|                    null|      null|                 null|literature|                            null|       breast cancer|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|  MONDO_0007254|   null|           null|001a2f1d02f87f20f...|0.055999999999999994|europepmc|\n",
      "|                             null|     null|      MONDO_0001721|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          5.4|           null|     null|[30114701]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000198793|                        null|         null|[[83, 75, abstrac...|    null|            P42345|   europepmc|                    null|      null|                 null|literature|                            null|           intrinsic|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|  MONDO_0001721|   null|           null|001b4b2114eeaade2...|0.054000000000000006|europepmc|\n",
      "|                             null|     null|        EFO_0009373|                                 null|                        null|          null|          null|            null|  null|                          null|           null|            null|          6.6|           null|     null|[19812964]|     null|               null|      null|                                 null|                  null|          null|          null|ENSG00000164867|                        null|         null|[[227, 223, abstr...|    null|            P29474|   europepmc|                    null|      null|                 null|literature|                            null|               edema|            null|               null|                   null|    null|         null|                   null|         null|       null|             null|       null|      null|        null|    EFO_0009373|   null|           null|001ddcbbcde71a527...|               0.066|europepmc|\n",
      "+---------------------------------+---------+-------------------+-------------------------------------+----------------------------+--------------+--------------+----------------+------+------------------------------+---------------+----------------+-------------+---------------+---------+----------+---------+-------------------+----------+-------------------------------------+----------------------+--------------+--------------+---------------+----------------------------+-------------+--------------------+--------+------------------+------------+------------------------+----------+---------------------+----------+--------------------------------+--------------------+----------------+-------------------+-----------------------+--------+-------------+-----------------------+-------------+-----------+-----------------+-----------+----------+------------+---------------+-------+---------------+--------------------+--------------------+---------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pyspark.sql\n",
    "from pyspark.sql.types import *\n",
    "from pyspark.sql.functions import *\n",
    "import pandas as pd\n",
    "\n",
    "global spark\n",
    "spark = (\n",
    "    pyspark.sql\n",
    "     .SparkSession\n",
    "     .builder\n",
    "     .config(\"spark.driver.memory\", \"10g\")\n",
    "     .getOrCreate()\n",
    "    )\n",
    "print('Spark version: ', spark.version)\n",
    "\n",
    "# Reading data:\n",
    "evidence_parquet = '/Users/dsuveges/project_data/ot/evidence_update/iter13_b/'\n",
    "df = spark.read.load(evidence_parquet)\n",
    "df.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding out if Uniprot has phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-24T15:24:13.150438Z",
     "start_time": "2021-02-24T15:24:10.821702Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for uniprot evidence where the cohortPhenotypes is not null:\n",
    "uniprot = (\n",
    "    df\n",
    "    .filter((col('dataSourceID') == 'uniprot_literature') & (col('cohortPhenotypes').isNotNull()))\n",
    ")\n",
    "uniprot.count()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-24T17:40:48.307675Z",
     "start_time": "2021-02-24T17:40:48.269544Z"
    }
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "argument of type 'float' is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-3b67a4bce8ff>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mtarget_activity\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mher3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtarget\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'activity'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0mdisease_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mher3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisease\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m         \u001b[0mvariant\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mher3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvariant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'id'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m     )\n\u001b[1;32m     16\u001b[0m     \u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'literature'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'evidence'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'variant'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'unique_association_fields'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'access_level'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'validated_against_schema_version'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter/lib/python3.6/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[1;32m   4198\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4199\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4200\u001b[0;31m                 \u001b[0mmapped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4201\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4202\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m<ipython-input-29-3b67a4bce8ff>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mtarget_activity\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mher3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtarget\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'activity'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0mdisease_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mher3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisease\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m         \u001b[0mvariant\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mher3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvariant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'id'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m     )\n\u001b[1;32m     16\u001b[0m     \u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'literature'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'evidence'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'variant'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'unique_association_fields'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'access_level'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'validated_against_schema_version'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: argument of type 'float' is not iterable"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "her3 = pd.read_json('/Users/dsuveges/project_data/ot/evidence_input/20.11/uniprot/her3_evidence.json', lines=True)\n",
    "\n",
    "pocok = (\n",
    "    her3.assign(\n",
    "        sourceID = her3.sourceID,\n",
    "        target = her3.unique_association_fields.apply(lambda x: x['target'].split('/')[-1]),\n",
    "        disease = her3.unique_association_fields.apply(lambda x: x['disease_uri'].split('/')[-1]),\n",
    "        disease_acronym = her3.unique_association_fields.apply(lambda x: x['disease_acronym']),\n",
    "        target_activity = her3.target.apply(lambda x: x['activity'].split('/')[-1]),\n",
    "        disease_name = her3.disease.apply(lambda x: x['name']),\n",
    "        variant = her3.variant.apply(lambda x: x['id'].split('/')[-1] if 'id' in x else None)\n",
    "    )\n",
    "    .drop(['literature','evidence','variant','unique_association_fields', 'access_level', 'validated_against_schema_version'], axis=1)\n",
    ")\n",
    "\n",
    "pocok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-24T16:21:35.197366Z",
     "start_time": "2021-02-24T16:18:38.901293Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>column</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>biologicalModelAllelicComposition</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>variantId</td>\n",
       "      <td>phewas_catalog,ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>diseaseFromSourceId</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>diseaseModelAssociatedModelPhenotypes</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>variantAminoacidDescriptions</td>\n",
       "      <td>reactome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>clinicalStatus</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mutatedSamples</td>\n",
       "      <td>intogen,cancer_gene_census</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cohortPhenotypes</td>\n",
       "      <td>genomics_england</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>drugId</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>phewas_catalog,eva,ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>publicationYear</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>targetModulation</td>\n",
       "      <td>reactome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>resourceScore</td>\n",
       "      <td>europepmc,crispr,slapenrich,expression_atlas,i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cohortShortName</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pathwayId</td>\n",
       "      <td>slapenrich,reactome,progeny</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>literature</td>\n",
       "      <td>europepmc,crispr,genomics_england,reactome,sys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>oddsRatio</td>\n",
       "      <td>phewas_catalog,ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>log2FoldChangeValue</td>\n",
       "      <td>expression_atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>confidence</td>\n",
       "      <td>genomics_england,clingen,eva,gene2phenotype,ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>diseaseModelAssociatedHumanPhenotypes</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>publicationFirstAuthor</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pValueExponent</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>pValueMantissa</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>targetId</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>log2FoldChangePercentileRank</td>\n",
       "      <td>expression_atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>clinicalPhase</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>textMiningSentences</td>\n",
       "      <td>europepmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>cohortId</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>targetFromSourceId</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>datasourceId</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>significantDriverMethods</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>reactionId</td>\n",
       "      <td>reactome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>clinicalSignificances</td>\n",
       "      <td>eva,eva_somatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>datatypeId</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>biologicalModelGeneticBackground</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>diseaseFromSource</td>\n",
       "      <td>europepmc,crispr,genomics_england,slapenrich,i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>diseaseCellLines</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>allelicRequirements</td>\n",
       "      <td>genomics_england,clingen,eva,gene2phenotype,ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>confidenceIntervalLower</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>contrast</td>\n",
       "      <td>expression_atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>studyOverview</td>\n",
       "      <td>genomics_england,expression_atlas,sysbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>confidenceIntervalUpper</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>targetInModel</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pathwayName</td>\n",
       "      <td>slapenrich,reactome,sysbio,progeny</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>cohortDescription</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>variantRsId</td>\n",
       "      <td>phewas_catalog,eva,eva_somatic,ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>studyCases</td>\n",
       "      <td>phewas_catalog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>clinicalUrls</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>diseaseId</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>studyId</td>\n",
       "      <td>genomics_england,expression_atlas,intogen,canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>studySampleSize</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>id</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>score</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>sourceId</td>\n",
       "      <td>europepmc,crispr,genomics_england,chembl,slape...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   column  \\\n",
       "0       biologicalModelAllelicComposition   \n",
       "1                               variantId   \n",
       "2                     diseaseFromSourceId   \n",
       "3   diseaseModelAssociatedModelPhenotypes   \n",
       "4            variantAminoacidDescriptions   \n",
       "5                          clinicalStatus   \n",
       "6                          mutatedSamples   \n",
       "7                        cohortPhenotypes   \n",
       "8                                  drugId   \n",
       "9          variantFunctionalConsequenceId   \n",
       "10                        publicationYear   \n",
       "11                       targetModulation   \n",
       "12                          resourceScore   \n",
       "13                        cohortShortName   \n",
       "14                              pathwayId   \n",
       "15                             literature   \n",
       "16                              oddsRatio   \n",
       "17                    log2FoldChangeValue   \n",
       "18                             confidence   \n",
       "19  diseaseModelAssociatedHumanPhenotypes   \n",
       "20                 publicationFirstAuthor   \n",
       "21                         pValueExponent   \n",
       "22                         pValueMantissa   \n",
       "23                               targetId   \n",
       "24           log2FoldChangePercentileRank   \n",
       "25                          clinicalPhase   \n",
       "26                    textMiningSentences   \n",
       "27                               cohortId   \n",
       "28                     targetFromSourceId   \n",
       "29                           datasourceId   \n",
       "30               significantDriverMethods   \n",
       "31                             reactionId   \n",
       "32                  clinicalSignificances   \n",
       "33                             datatypeId   \n",
       "34       biologicalModelGeneticBackground   \n",
       "35                      diseaseFromSource   \n",
       "36                       diseaseCellLines   \n",
       "37                    allelicRequirements   \n",
       "38                confidenceIntervalLower   \n",
       "39                               contrast   \n",
       "40                          studyOverview   \n",
       "41                confidenceIntervalUpper   \n",
       "42                          targetInModel   \n",
       "43                            pathwayName   \n",
       "44                      cohortDescription   \n",
       "45                            variantRsId   \n",
       "46                             studyCases   \n",
       "47                           clinicalUrls   \n",
       "48                              diseaseId   \n",
       "49                                studyId   \n",
       "50                        studySampleSize   \n",
       "51                                     id   \n",
       "52                                  score   \n",
       "53                               sourceId   \n",
       "\n",
       "                                              sources  \n",
       "0                                           phenodigm  \n",
       "1                   phewas_catalog,ot_genetics_portal  \n",
       "2   europepmc,crispr,genomics_england,chembl,slape...  \n",
       "3                                           phenodigm  \n",
       "4                                            reactome  \n",
       "5                                              chembl  \n",
       "6                          intogen,cancer_gene_census  \n",
       "7                                    genomics_england  \n",
       "8                                              chembl  \n",
       "9               phewas_catalog,eva,ot_genetics_portal  \n",
       "10                                 ot_genetics_portal  \n",
       "11                                           reactome  \n",
       "12  europepmc,crispr,slapenrich,expression_atlas,i...  \n",
       "13                                            intogen  \n",
       "14                        slapenrich,reactome,progeny  \n",
       "15  europepmc,crispr,genomics_england,reactome,sys...  \n",
       "16                  phewas_catalog,ot_genetics_portal  \n",
       "17                                   expression_atlas  \n",
       "18  genomics_england,clingen,eva,gene2phenotype,ev...  \n",
       "19                                          phenodigm  \n",
       "20                                 ot_genetics_portal  \n",
       "21                                 ot_genetics_portal  \n",
       "22                                 ot_genetics_portal  \n",
       "23  europepmc,crispr,genomics_england,chembl,slape...  \n",
       "24                                   expression_atlas  \n",
       "25                                             chembl  \n",
       "26                                          europepmc  \n",
       "27                                            intogen  \n",
       "28  europepmc,crispr,genomics_england,chembl,slape...  \n",
       "29  europepmc,crispr,genomics_england,chembl,slape...  \n",
       "30                                            intogen  \n",
       "31                                           reactome  \n",
       "32                                    eva,eva_somatic  \n",
       "33  europepmc,crispr,genomics_england,chembl,slape...  \n",
       "34                                          phenodigm  \n",
       "35  europepmc,crispr,genomics_england,slapenrich,i...  \n",
       "36                                             crispr  \n",
       "37  genomics_england,clingen,eva,gene2phenotype,ev...  \n",
       "38                                 ot_genetics_portal  \n",
       "39                                   expression_atlas  \n",
       "40           genomics_england,expression_atlas,sysbio  \n",
       "41                                 ot_genetics_portal  \n",
       "42                                          phenodigm  \n",
       "43                 slapenrich,reactome,sysbio,progeny  \n",
       "44                                            intogen  \n",
       "45  phewas_catalog,eva,eva_somatic,ot_genetics_portal  \n",
       "46                                     phewas_catalog  \n",
       "47                                             chembl  \n",
       "48  europepmc,crispr,genomics_england,chembl,slape...  \n",
       "49  genomics_england,expression_atlas,intogen,canc...  \n",
       "50                                 ot_genetics_portal  \n",
       "51  europepmc,crispr,genomics_england,chembl,slape...  \n",
       "52  europepmc,crispr,genomics_england,chembl,slape...  \n",
       "53  europepmc,crispr,genomics_england,chembl,slape...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "notnull = []\n",
    "\n",
    "for column in df.columns:\n",
    "    sources = (\n",
    "        df\n",
    "        .filter(f'{column} is not null')\n",
    "        .select('sourceId')\n",
    "        .distinct()\n",
    "        .toPandas()\n",
    "    )\n",
    "    \n",
    "    notnull.append({\n",
    "        'column': column,\n",
    "        'sources': ','.join(sources.sourceId.tolist())\n",
    "    })\n",
    "    \n",
    "sources_df = pd.DataFrame(notnull)\n",
    "sources_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-02T16:14:02.377270Z",
     "start_time": "2020-12-02T16:14:02.370580Z"
    }
   },
   "outputs": [],
   "source": [
    "sources_df.sort_values('column', inplace=True)\n",
    "sources_df.to_csv('/Users/dsuveges/project/random_notebooks/issue-1249_prototyping_new_json_schema/fields_sources.tsv', index=False, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-02T15:05:20.033425Z",
     "start_time": "2020-12-02T15:05:16.728099Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>functionalConsequenceId</th>\n",
       "      <th>numberMutatedSamples</th>\n",
       "      <th>numberSamplesTested</th>\n",
       "      <th>numberSamplesWithMutationType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SO_0001605</td>\n",
       "      <td>1240</td>\n",
       "      <td>9180</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SO_0001059</td>\n",
       "      <td>1240</td>\n",
       "      <td>9180</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SO_0001589</td>\n",
       "      <td>1240</td>\n",
       "      <td>9180</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SO_0001583</td>\n",
       "      <td>1240</td>\n",
       "      <td>9180</td>\n",
       "      <td>1194.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SO_0001059</td>\n",
       "      <td>6</td>\n",
       "      <td>626</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  functionalConsequenceId  numberMutatedSamples  numberSamplesTested  \\\n",
       "0              SO_0001605                  1240                 9180   \n",
       "1              SO_0001059                  1240                 9180   \n",
       "2              SO_0001589                  1240                 9180   \n",
       "3              SO_0001583                  1240                 9180   \n",
       "4              SO_0001059                     6                  626   \n",
       "\n",
       "   numberSamplesWithMutationType  \n",
       "0                            1.0  \n",
       "1                           44.0  \n",
       "2                            1.0  \n",
       "3                         1194.0  \n",
       "4                            6.0  "
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_samples = (\n",
    "    df\n",
    "    .filter('mutatedSamples is not null')\n",
    "    .select(explode(\"mutatedSamples\").alias(\"exploded\"))\n",
    "    .select(\"exploded.*\")\n",
    "    .toPandas()\n",
    ")\n",
    "mutated_samples.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-02T12:33:37.546362Z",
     "start_time": "2020-12-02T12:33:37.517179Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutatedSamples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[(SO_0001605, 1240, 9180, 1), (SO_0001059, 124...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[(SO_0001059, 6, 626, 6), (SO_0001583, 6, 626,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[(SO_0001587, 22, 802, 1), (SO_0001059, 22, 80...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[(SO_0001589, 3, 455, 1), (SO_0001583, 3, 455,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[(SO_0001059, 14, 370, 1), (SO_0001583, 14, 37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65003</th>\n",
       "      <td>[(None, 3, 121, None)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65004</th>\n",
       "      <td>[(None, 13, 58, None)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65005</th>\n",
       "      <td>[(None, 2, 148, None)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65006</th>\n",
       "      <td>[(None, 3, 15, None)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65007</th>\n",
       "      <td>[(None, 2, 148, None)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65008 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          mutatedSamples\n",
       "0      [(SO_0001605, 1240, 9180, 1), (SO_0001059, 124...\n",
       "1      [(SO_0001059, 6, 626, 6), (SO_0001583, 6, 626,...\n",
       "2      [(SO_0001587, 22, 802, 1), (SO_0001059, 22, 80...\n",
       "3      [(SO_0001589, 3, 455, 1), (SO_0001583, 3, 455,...\n",
       "4      [(SO_0001059, 14, 370, 1), (SO_0001583, 14, 37...\n",
       "...                                                  ...\n",
       "65003                             [(None, 3, 121, None)]\n",
       "65004                             [(None, 13, 58, None)]\n",
       "65005                             [(None, 2, 148, None)]\n",
       "65006                              [(None, 3, 15, None)]\n",
       "65007                             [(None, 2, 148, None)]\n",
       "\n",
       "[65008 rows x 1 columns]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-02T15:39:50.490862Z",
     "start_time": "2020-12-02T15:39:50.395251Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- allelicRequirements: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- biologicalModelAllelicComposition: string (nullable = true)\n",
      " |-- biologicalModelGeneticBackground: string (nullable = true)\n",
      " |-- clinicalPhase: long (nullable = true)\n",
      " |-- clinicalSignificances: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- clinicalStatus: string (nullable = true)\n",
      " |-- clinicalUrls: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- niceName: string (nullable = true)\n",
      " |    |    |-- url: string (nullable = true)\n",
      " |-- cohortDescription: string (nullable = true)\n",
      " |-- cohortId: string (nullable = true)\n",
      " |-- cohortPhenotypes: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- cohortShortName: string (nullable = true)\n",
      " |-- confidence: string (nullable = true)\n",
      " |-- confidenceIntervalLower: double (nullable = true)\n",
      " |-- confidenceIntervalUpper: double (nullable = true)\n",
      " |-- contrast: string (nullable = true)\n",
      " |-- datasourceId: string (nullable = true)\n",
      " |-- datatypeId: string (nullable = true)\n",
      " |-- diseaseCellLines: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- diseaseFromSource: string (nullable = true)\n",
      " |-- diseaseFromSourceId: string (nullable = true)\n",
      " |-- diseaseId: string (nullable = true)\n",
      " |-- diseaseModelAssociatedHumanPhenotypes: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |-- diseaseModelAssociatedModelPhenotypes: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |-- drugId: string (nullable = true)\n",
      " |-- id: string (nullable = true)\n",
      " |-- literature: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- log2FoldChangePercentileRank: long (nullable = true)\n",
      " |-- log2FoldChangeValue: double (nullable = true)\n",
      " |-- mutatedSamples: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- functionalConsequenceId: string (nullable = true)\n",
      " |    |    |-- numberMutatedSamples: long (nullable = true)\n",
      " |    |    |-- numberSamplesTested: long (nullable = true)\n",
      " |    |    |-- numberSamplesWithMutationType: long (nullable = true)\n",
      " |-- oddsRatio: double (nullable = true)\n",
      " |-- pValueExponent: long (nullable = true)\n",
      " |-- pValueMantissa: long (nullable = true)\n",
      " |-- pathwayId: string (nullable = true)\n",
      " |-- pathwayName: string (nullable = true)\n",
      " |-- publicationFirstAuthor: string (nullable = true)\n",
      " |-- publicationYear: long (nullable = true)\n",
      " |-- reactionId: string (nullable = true)\n",
      " |-- resourceScore: double (nullable = true)\n",
      " |-- score: double (nullable = true)\n",
      " |-- significantDriverMethods: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- sourceId: string (nullable = true)\n",
      " |-- studyCases: long (nullable = true)\n",
      " |-- studyId: string (nullable = true)\n",
      " |-- studyOverview: string (nullable = true)\n",
      " |-- studySampleSize: long (nullable = true)\n",
      " |-- targetFromSourceId: string (nullable = true)\n",
      " |-- targetId: string (nullable = true)\n",
      " |-- targetInModel: string (nullable = true)\n",
      " |-- targetModulation: string (nullable = true)\n",
      " |-- textMiningSentences: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- dEnd: long (nullable = true)\n",
      " |    |    |-- dStart: long (nullable = true)\n",
      " |    |    |-- section: string (nullable = true)\n",
      " |    |    |-- tEnd: long (nullable = true)\n",
      " |    |    |-- tStart: long (nullable = true)\n",
      " |    |    |-- text: string (nullable = true)\n",
      " |-- variantAminoacidDescriptions: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- variantFunctionalConsequenceId: string (nullable = true)\n",
      " |-- variantId: string (nullable = true)\n",
      " |-- variantRsId: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.select(sorted(df.columns)).printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-02T16:26:27.706274Z",
     "start_time": "2020-12-02T16:26:22.201520Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------------------------+--------+\n",
      "|variantAminoacidDescriptions|sourceId|\n",
      "+----------------------------+--------+\n",
      "|        [L-leucine 389 re...|reactome|\n",
      "|        [L-glutamic acid ...|reactome|\n",
      "|                          []|reactome|\n",
      "|        [L-isoleucine 161...|reactome|\n",
      "|                          []|reactome|\n",
      "|        [L-threonine 680 ...|reactome|\n",
      "|        [L-tyrosine 207 r...|reactome|\n",
      "|        [Insertion of res...|reactome|\n",
      "|                          []|reactome|\n",
      "|                          []|reactome|\n",
      "|                          []|reactome|\n",
      "|                          []|reactome|\n",
      "|        [glycine 12 repla...|reactome|\n",
      "|                          []|reactome|\n",
      "|                          []|reactome|\n",
      "|        [L-valine 1576 re...|reactome|\n",
      "|                          []|reactome|\n",
      "|        [L-asparagine 822...|reactome|\n",
      "|        [L-proline 252 re...|reactome|\n",
      "|        [L-glutamine 61 r...|reactome|\n",
      "+----------------------------+--------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    df\n",
    "    .filter('variantAminoacidDescriptions is not null')\n",
    "    .select('variantAminoacidDescriptions','sourceId')\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-10T17:58:42.562828Z",
     "start_time": "2020-12-10T17:58:40.214043Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"'Crohn's disease, undetermined' vs 'non inflammatory bowel disease control'\",\n",
       " \"'Myotonic dystrophy' vs 'normal'\",\n",
       " \"'rheumatoid arthritis' vs 'normal'\",\n",
       " \"'primary pancreatic ductal adenocarcinoma' vs 'normal' in 'pancreas'\",\n",
       " \"'colon cancer' vs 'normal' in 'colon; Fresh-frozen tissue'\",\n",
       " \"'Ataxia-telangiectasia' vs 'normal'\",\n",
       " \"'tumour tissue' vs 'adjacent normal tissue'\",\n",
       " \"'glioblastoma' vs 'normal'\",\n",
       " \"'esophageal adenocarcinoma' vs 'normal'\",\n",
       " \"'meningeal tuberculosis' vs 'normal'\",\n",
       " \"'bacteriemia; Staphylococcus aureus' vs 'normal; none'\",\n",
       " \"'esophageal adenocarcinoma' vs 'normal'\",\n",
       " \"'non-enhancing margin; glioma' vs 'normal'\",\n",
       " \"'monocyte; meningococcal sepsis' at '0 hour' vs 'monocyte; normal' at '0 hour'\",\n",
       " \"'sepsis' vs 'normal' in 'whole blood'\",\n",
       " \"'tumor tissue' vs 'non-malignant tissue'\",\n",
       " \"'Alzheimers disease' vs 'normal' in 'entorhinal cortex'\",\n",
       " \"'Crohn's disease, macroscopic ileal inflammation with deep ulcer' vs 'non inflammatory bowel disease control'\",\n",
       " \"'colon cancer' vs 'normal' in 'colon; Fresh-frozen tissue'\"]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = (df\n",
    "    .filter(df.contrast.isNotNull())\n",
    "    .select(col('contrast'))\n",
    "    .limit(19).toPandas())\n",
    "\n",
    "\n",
    "x.contrast.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-23T21:12:08.005358Z",
     "start_time": "2021-01-23T21:12:01.517508Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biologicalModelAllelicComposition</th>\n",
       "      <th>variantId</th>\n",
       "      <th>diseaseFromSourceId</th>\n",
       "      <th>diseaseModelAssociatedModelPhenotypes</th>\n",
       "      <th>variantAminoacidDescriptions</th>\n",
       "      <th>clinicalStatus</th>\n",
       "      <th>mutatedSamples</th>\n",
       "      <th>cohortPhenotypes</th>\n",
       "      <th>drugId</th>\n",
       "      <th>variantFunctionalConsequenceId</th>\n",
       "      <th>...</th>\n",
       "      <th>cohortDescription</th>\n",
       "      <th>variantRsId</th>\n",
       "      <th>studyCases</th>\n",
       "      <th>clinicalUrls</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>studyId</th>\n",
       "      <th>studySampleSize</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>sourceId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0005922</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0005922</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0c3ae08069a9f02e7c9bdf01b53ff1ef16eb2507</td>\n",
       "      <td>0.426250</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0000305</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0000305</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106f631b10cad601fd69d041b1ea027d0ab0f2fd</td>\n",
       "      <td>0.464688</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0001378</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0001378</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a85e8f3af29b61c323d52d503a25772bca9b375</td>\n",
       "      <td>0.557813</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0000222</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0000222</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>339addf13f461e55651ce4c566ef3b073dfff18a</td>\n",
       "      <td>0.448889</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0001378</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0001378</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38866d61fb0c0877d91b06a8b31e60547818a4cf</td>\n",
       "      <td>0.514063</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  biologicalModelAllelicComposition variantId diseaseFromSourceId  \\\n",
       "0                              None      None         EFO_0005922   \n",
       "1                              None      None         EFO_0000305   \n",
       "2                              None      None         EFO_0001378   \n",
       "3                              None      None         EFO_0000222   \n",
       "4                              None      None         EFO_0001378   \n",
       "\n",
       "  diseaseModelAssociatedModelPhenotypes variantAminoacidDescriptions  \\\n",
       "0                                  None                         None   \n",
       "1                                  None                         None   \n",
       "2                                  None                         None   \n",
       "3                                  None                         None   \n",
       "4                                  None                         None   \n",
       "\n",
       "  clinicalStatus mutatedSamples cohortPhenotypes drugId  \\\n",
       "0           None           None             None   None   \n",
       "1           None           None             None   None   \n",
       "2           None           None             None   None   \n",
       "3           None           None             None   None   \n",
       "4           None           None             None   None   \n",
       "\n",
       "  variantFunctionalConsequenceId  ...  cohortDescription variantRsId  \\\n",
       "0                           None  ...               None        None   \n",
       "1                           None  ...               None        None   \n",
       "2                           None  ...               None        None   \n",
       "3                           None  ...               None        None   \n",
       "4                           None  ...               None        None   \n",
       "\n",
       "   studyCases clinicalUrls    diseaseId studyId  studySampleSize  \\\n",
       "0         NaN         None  EFO_0005922    None              NaN   \n",
       "1         NaN         None  EFO_0000305    None              NaN   \n",
       "2         NaN         None  EFO_0001378    None              NaN   \n",
       "3         NaN         None  EFO_0000222    None              NaN   \n",
       "4         NaN         None  EFO_0001378    None              NaN   \n",
       "\n",
       "                                         id     score sourceId  \n",
       "0  0c3ae08069a9f02e7c9bdf01b53ff1ef16eb2507  0.426250   crispr  \n",
       "1  106f631b10cad601fd69d041b1ea027d0ab0f2fd  0.464688   crispr  \n",
       "2  1a85e8f3af29b61c323d52d503a25772bca9b375  0.557813   crispr  \n",
       "3  339addf13f461e55651ce4c566ef3b073dfff18a  0.448889   crispr  \n",
       "4  38866d61fb0c0877d91b06a8b31e60547818a4cf  0.514063   crispr  \n",
       "\n",
       "[5 rows x 54 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xf = (\n",
    "    df\n",
    "     .filter(df.datasourceId=='crispr')\n",
    "     .toPandas()\n",
    ")\n",
    "\n",
    "xf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-23T21:22:58.502109Z",
     "start_time": "2021-01-23T21:22:58.447392Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'diseaseFromSourceId': 'EFO_0000305',\n",
       " 'resourceScore': 0.464688,\n",
       " 'literature': ['30971826'],\n",
       " 'targetId': 'ENSG00000138107',\n",
       " 'targetFromSourceId': 'ENSG00000138107',\n",
       " 'datasourceId': 'crispr',\n",
       " 'datatypeId': 'affected_pathway',\n",
       " 'diseaseFromSource': 'Breast Carcinoma',\n",
       " 'diseaseCellLines': ['AU565',\n",
       "  'COLO-824',\n",
       "  'EVSA-T',\n",
       "  'HCC1143',\n",
       "  'HCC1187',\n",
       "  'HCC1806',\n",
       "  'HCC1937',\n",
       "  'HCC1954',\n",
       "  'HCC38',\n",
       "  'Hs-578-T',\n",
       "  'JIMT-1',\n",
       "  'MCF7',\n",
       "  'MDA-MB-361',\n",
       "  'MDA-MB-415',\n",
       "  'MDA-MB-436',\n",
       "  'MDA-MB-453',\n",
       "  'MFM-223',\n",
       "  'OCUB-M',\n",
       "  'UACC-893',\n",
       "  'CAL-51',\n",
       "  'HCC1395',\n",
       "  'HCC70',\n",
       "  'MDA-MB-231',\n",
       "  'T47D',\n",
       "  'MDA-MB-468'],\n",
       " 'diseaseId': 'EFO_0000305',\n",
       " 'id': '106f631b10cad601fd69d041b1ea027d0ab0f2fd',\n",
       " 'score': 0.464688,\n",
       " 'sourceId': 'crispr'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "xf.drop([col for col in xf.columns if not xf[col].any()], axis=1, inplace=True)\n",
    "xf.iloc[1].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-23T21:13:41.461393Z",
     "start_time": "2021-01-23T21:13:41.457094Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['biologicalModelAllelicComposition', 'variantId', 'diseaseFromSourceId',\n",
       "       'diseaseModelAssociatedModelPhenotypes', 'variantAminoacidDescriptions',\n",
       "       'clinicalStatus', 'mutatedSamples', 'cohortPhenotypes', 'drugId',\n",
       "       'variantFunctionalConsequenceId', 'publicationYear', 'targetModulation',\n",
       "       'resourceScore', 'cohortShortName', 'pathwayId', 'literature',\n",
       "       'oddsRatio', 'log2FoldChangeValue', 'confidence',\n",
       "       'diseaseModelAssociatedHumanPhenotypes', 'publicationFirstAuthor',\n",
       "       'pValueExponent', 'pValueMantissa', 'targetId',\n",
       "       'log2FoldChangePercentileRank', 'clinicalPhase', 'textMiningSentences',\n",
       "       'cohortId', 'targetFromSourceId', 'datasourceId',\n",
       "       'significantDriverMethods', 'reactionId', 'clinicalSignificances',\n",
       "       'datatypeId', 'biologicalModelGeneticBackground', 'diseaseFromSource',\n",
       "       'diseaseCellLines', 'allelicRequirements', 'confidenceIntervalLower',\n",
       "       'contrast', 'studyOverview', 'confidenceIntervalUpper', 'targetInModel',\n",
       "       'pathwayName', 'cohortDescription', 'variantRsId', 'studyCases',\n",
       "       'clinicalUrls', 'diseaseId', 'studyId', 'studySampleSize', 'id',\n",
       "       'score', 'sourceId'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
