{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prototyping the new json schema\n",
    "\n",
    "\n",
    "As part of the consolidation of the evidence objects in the backend, we are re-modeling the [json schema](https://github.com/opentargets/json_schema) to reflect the new simplified/flattened design.\n",
    "\n",
    "Based on the meeting we had on 2020.11.11 the following conclusions were reached:\n",
    "\n",
    "* We need to maintain a json schema that guides our data providers and can be used as template to generate evidence strings.\n",
    "* The schema will reflect the concepts of the new platform design, so the units of the schema is going to be data source centric instead of data type.\n",
    "* Each of the valuable columns will be defined in a common section.\n",
    "* For each data source there will only be a list of required fields.\n",
    "* We haven't reached a consensus on how the unique association fields are defined, and at which point of the evidence generation. So for the first iteration of the json schema, the unique_association_fields will be omitted.\n",
    "\n",
    "The schema is written based on the most recent iteration of the [evidence schema review](https://docs.google.com/spreadsheets/d/11jdPCo_vxY3jaP54xKTsXBshR5HMrpUf5oXJNgtbKm8/edit#gid=1735847104) document.\n",
    "\n",
    "The technical approach:\n",
    "\n",
    "* To avoid manual work with the json document, I'm collating information in an excel file and will use that as a source for the definitions.\n",
    "* The same excel file will be used to get the source names from where we are expecting the given field.\n",
    "\n",
    "## The first run completed:\n",
    "\n",
    "- [X] processing the review document to get the rough list of fields\n",
    "- [x] get fields2datasource mapping\n",
    "- [x] generate json schema based on the meeting\n",
    "\n",
    "\n",
    "## The first run didn't cover:\n",
    "\n",
    "- [ ] some fields are missing eg. uniprot id\n",
    "- [ ] some fields shoudl not be here: `score` and `id`\n",
    "- [ ] the precise requirements of the fields are still sparse -> add more data to `field_description.xlsx`\n",
    "- [ ] no structure whatsoever."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Getting the list of data source for every field\n",
    "\n",
    "This information is extracted from the evidence schema review file. The end of the process is a comma separated list of data sources for every field. This column is used later to generate the mandatory list of fields for every data source."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-12T01:32:16.268651Z",
     "start_time": "2020-11-12T01:32:16.247276Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_gene_census</th>\n",
       "      <th>chembl</th>\n",
       "      <th>clingen</th>\n",
       "      <th>crispr</th>\n",
       "      <th>europepmc</th>\n",
       "      <th>eva</th>\n",
       "      <th>eva_somatic</th>\n",
       "      <th>expression_atlas</th>\n",
       "      <th>gene2phenotype</th>\n",
       "      <th>genomics_england</th>\n",
       "      <th>intogen</th>\n",
       "      <th>ot_genetics_portal</th>\n",
       "      <th>phenodigm</th>\n",
       "      <th>phewas_catalog</th>\n",
       "      <th>progeny</th>\n",
       "      <th>reactome</th>\n",
       "      <th>slapenrich</th>\n",
       "      <th>sysbio</th>\n",
       "      <th>uniprot_literature</th>\n",
       "      <th>uniprot_somatic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>allelicRequirement</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biologicalModelAllelicComposition</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>564310</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biologicalModelGeneticBackground</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>564310</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clinicalPhase</th>\n",
       "      <td>0</td>\n",
       "      <td>427943</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clinicalSignificance</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>107532</td>\n",
       "      <td>8173</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cancer_gene_census  chembl  clingen  \\\n",
       "allelicRequirement                                  0       0     1075   \n",
       "biologicalModelAllelicComposition                   0       0        0   \n",
       "biologicalModelGeneticBackground                    0       0        0   \n",
       "clinicalPhase                                       0  427943        0   \n",
       "clinicalSignificance                                0       0        0   \n",
       "\n",
       "                                   crispr  europepmc     eva  eva_somatic  \\\n",
       "allelicRequirement                      0          0       0            0   \n",
       "biologicalModelAllelicComposition       0          0       0            0   \n",
       "biologicalModelGeneticBackground        0          0       0            0   \n",
       "clinicalPhase                           0          0       0            0   \n",
       "clinicalSignificance                    0          0  107532         8173   \n",
       "\n",
       "                                   expression_atlas  gene2phenotype  \\\n",
       "allelicRequirement                                0            2451   \n",
       "biologicalModelAllelicComposition                 0               0   \n",
       "biologicalModelGeneticBackground                  0               0   \n",
       "clinicalPhase                                     0               0   \n",
       "clinicalSignificance                              0               0   \n",
       "\n",
       "                                   genomics_england  intogen  \\\n",
       "allelicRequirement                                0        0   \n",
       "biologicalModelAllelicComposition                 0        0   \n",
       "biologicalModelGeneticBackground                  0        0   \n",
       "clinicalPhase                                     0        0   \n",
       "clinicalSignificance                              0        0   \n",
       "\n",
       "                                   ot_genetics_portal  phenodigm  \\\n",
       "allelicRequirement                                  0          0   \n",
       "biologicalModelAllelicComposition                   0     564310   \n",
       "biologicalModelGeneticBackground                    0     564310   \n",
       "clinicalPhase                                       0          0   \n",
       "clinicalSignificance                                0          0   \n",
       "\n",
       "                                   phewas_catalog  progeny  reactome  \\\n",
       "allelicRequirement                              0        0         0   \n",
       "biologicalModelAllelicComposition               0        0         0   \n",
       "biologicalModelGeneticBackground                0        0         0   \n",
       "clinicalPhase                                   0        0         0   \n",
       "clinicalSignificance                            0        0         0   \n",
       "\n",
       "                                   slapenrich  sysbio  uniprot_literature  \\\n",
       "allelicRequirement                          0       0                   0   \n",
       "biologicalModelAllelicComposition           0       0                   0   \n",
       "biologicalModelGeneticBackground            0       0                   0   \n",
       "clinicalPhase                               0       0                   0   \n",
       "clinicalSignificance                        0       0                   0   \n",
       "\n",
       "                                   uniprot_somatic  \n",
       "allelicRequirement                               0  \n",
       "biologicalModelAllelicComposition                0  \n",
       "biologicalModelGeneticBackground                 0  \n",
       "clinicalPhase                                    0  \n",
       "clinicalSignificance                             0  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Get the source names for every field:\n",
    "import pandas as pd\n",
    "import json\n",
    "from collections import OrderedDict, defaultdict\n",
    "import numpy as np\n",
    "\n",
    "notnull_table = pd.read_csv('iter9_notnull_table.tsv', sep='\\t')\n",
    "notnull_table.index = notnull_table.key.tolist()\n",
    "notnull_table.drop('key', axis=1, inplace=True)\n",
    "notnull_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-11T21:11:03.048756Z",
     "start_time": "2020-11-11T21:11:02.990035Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         clingen,gene2phenotype\n",
       "1                      phenodigm\n",
       "2                      phenodigm\n",
       "3                         chembl\n",
       "4                eva,eva_somatic\n",
       "                 ...            \n",
       "57    cancer_gene_census,intogen\n",
       "58                      reactome\n",
       "59    cancer_gene_census,intogen\n",
       "60    cancer_gene_census,intogen\n",
       "61            cancer_gene_census\n",
       "Length: 62, dtype: object"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looping through all fields and extracting all sources where the data is present:\n",
    "lookup = lambda x: ','.join(notnull_table.loc[x].where(lambda x: x != 0).dropna().index.to_list())\n",
    "pd.Series(notnull_table.index).apply(lookup)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading field description\n",
    "\n",
    "1. Read excel file with the field descriptions\n",
    "2. Parse values. \n",
    "3. Start building json object.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-12T01:08:49.967723Z",
     "start_time": "2020-11-12T01:08:49.943993Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>backend_name</th>\n",
       "      <th>type</th>\n",
       "      <th>description</th>\n",
       "      <th>minimum</th>\n",
       "      <th>exclusiveMinimum</th>\n",
       "      <th>maximum</th>\n",
       "      <th>accepted_values</th>\n",
       "      <th>nullable</th>\n",
       "      <th>pattern</th>\n",
       "      <th>data_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>allelicRequirement</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clingen,gene2phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>biologicalModelAllelicComposition</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>biologicalModelGeneticBackground</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>clinicalPhase</td>\n",
       "      <td>integer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>clinicalSignificance</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eva,eva_somatic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        backend_name     type description  minimum  \\\n",
       "0                 allelicRequirement   string         NaN      NaN   \n",
       "1  biologicalModelAllelicComposition   string         NaN      NaN   \n",
       "2   biologicalModelGeneticBackground   string         NaN      NaN   \n",
       "3                      clinicalPhase  integer         NaN      NaN   \n",
       "4               clinicalSignificance   string         NaN      NaN   \n",
       "\n",
       "   exclusiveMinimum  maximum  accepted_values nullable pattern  \\\n",
       "0               NaN      NaN              NaN      NaN     NaN   \n",
       "1               NaN      NaN              NaN      NaN     NaN   \n",
       "2               NaN      NaN              NaN      NaN     NaN   \n",
       "3               NaN      NaN              NaN      NaN     NaN   \n",
       "4               NaN      NaN              NaN      NaN     NaN   \n",
       "\n",
       "              data_source  \n",
       "0  clingen,gene2phenotype  \n",
       "1               phenodigm  \n",
       "2               phenodigm  \n",
       "3                  chembl  \n",
       "4         eva,eva_somatic  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_description = 'field_descriptions.xlsx'\n",
    "fields_df = pd.read_excel(field_description)\n",
    "fields_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-11T23:21:38.774276Z",
     "start_time": "2020-11-11T23:21:38.759184Z"
    }
   },
   "outputs": [],
   "source": [
    "def parse_data_sources(df):\n",
    "    # Parsing dataframe to get list of fields for each data source:\n",
    "    parsed_sources = defaultdict(list)\n",
    "    for i, row in df.iterrows():\n",
    "        for source in row['data_source'].split(','):\n",
    "            \n",
    "            # Only using ot_genetics for now:\n",
    "            if source != 'ot_genetics_portal':\n",
    "                continue\n",
    "                \n",
    "            parsed_sources[source].append(row['backend_name'])\n",
    "\n",
    "    # Each data source then exploded into schemas:\n",
    "    source_schemas = []\n",
    "    for source, fields in parsed_sources.items():\n",
    "        source_schema = OrderedDict()\n",
    "\n",
    "        # Adding property definitions:\n",
    "        source_schema['properties'] = OrderedDict({'datasourceId': {\"const\": source}})\n",
    "        \n",
    "        for field in fields:\n",
    "            if field == 'datasourceId':\n",
    "                continue\n",
    "                \n",
    "            source_schema['properties'][field] = {\"$ref\": f\"#/definitions/{field}\"}\n",
    "        \n",
    "        source_schema['required'] = fields\n",
    "\n",
    "\n",
    "        # Adding source schema:\n",
    "        source_schemas.append(source_schema)\n",
    "        \n",
    "    return(source_schemas)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-12T02:30:07.542668Z",
     "start_time": "2020-11-12T02:30:07.485868Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
      "  \"title\": \"OpenTargets\",\n",
      "  \"description\": \"OpenTargets evidence objects\",\n",
      "  \"type\": \"object\",\n",
      "  \"oneOf\": [\n",
      "    {\n",
      "      \"properties\": {\n",
      "        \"datasourceId\": {\n",
      "          \"const\": \"ot_genetics_portal\"\n",
      "        },\n",
      "        \"confidenceIntervalLower\": {\n",
      "          \"$ref\": \"#/definitions/confidenceIntervalLower\"\n",
      "        },\n",
      "        \"confidenceIntervalUpper\": {\n",
      "          \"$ref\": \"#/definitions/confidenceIntervalUpper\"\n",
      "        },\n",
      "        \"diseaseFromSource\": {\n",
      "          \"$ref\": \"#/definitions/diseaseFromSource\"\n",
      "        },\n",
      "        \"diseaseId\": {\n",
      "          \"$ref\": \"#/definitions/diseaseId\"\n",
      "        },\n",
      "        \"id\": {\n",
      "          \"$ref\": \"#/definitions/id\"\n",
      "        },\n",
      "        \"literature\": {\n",
      "          \"$ref\": \"#/definitions/literature\"\n",
      "        },\n",
      "        \"locus2GeneScore\": {\n",
      "          \"$ref\": \"#/definitions/locus2GeneScore\"\n",
      "        },\n",
      "        \"oddsRatio\": {\n",
      "          \"$ref\": \"#/definitions/oddsRatio\"\n",
      "        },\n",
      "        \"publicationFirstAuthor\": {\n",
      "          \"$ref\": \"#/definitions/publicationFirstAuthor\"\n",
      "        },\n",
      "        \"publicationYear\": {\n",
      "          \"$ref\": \"#/definitions/publicationYear\"\n",
      "        },\n",
      "        \"resourceScore\": {\n",
      "          \"$ref\": \"#/definitions/resourceScore\"\n",
      "        },\n",
      "        \"resourceScoreExponent\": {\n",
      "          \"$ref\": \"#/definitions/resourceScoreExponent\"\n",
      "        },\n",
      "        \"resourceScoreMantissa\": {\n",
      "          \"$ref\": \"#/definitions/resourceScoreMantissa\"\n",
      "        },\n",
      "        \"score\": {\n",
      "          \"$ref\": \"#/definitions/score\"\n",
      "        },\n",
      "        \"studyId\": {\n",
      "          \"$ref\": \"#/definitions/studyId\"\n",
      "        },\n",
      "        \"studySampleSize\": {\n",
      "          \"$ref\": \"#/definitions/studySampleSize\"\n",
      "        },\n",
      "        \"targetId\": {\n",
      "          \"$ref\": \"#/definitions/targetId\"\n",
      "        },\n",
      "        \"variantFunctionalConsequenceId\": {\n",
      "          \"$ref\": \"#/definitions/variantFunctionalConsequenceId\"\n",
      "        },\n",
      "        \"variantId\": {\n",
      "          \"$ref\": \"#/definitions/variantId\"\n",
      "        },\n",
      "        \"variantRsId\": {\n",
      "          \"$ref\": \"#/definitions/variantRsId\"\n",
      "        }\n",
      "      },\n",
      "      \"required\": [\n",
      "        \"confidenceIntervalLower\",\n",
      "        \"confidenceIntervalUpper\",\n",
      "        \"datasourceId\",\n",
      "        \"diseaseFromSource\",\n",
      "        \"diseaseId\",\n",
      "        \"id\",\n",
      "        \"literature\",\n",
      "        \"locus2GeneScore\",\n",
      "        \"oddsRatio\",\n",
      "        \"publicationFirstAuthor\",\n",
      "        \"publicationYear\",\n",
      "        \"resourceScore\",\n",
      "        \"resourceScoreExponent\",\n",
      "        \"resourceScoreMantissa\",\n",
      "        \"score\",\n",
      "        \"studyId\",\n",
      "        \"studySampleSize\",\n",
      "        \"targetId\",\n",
      "        \"variantFunctionalConsequenceId\",\n",
      "        \"variantId\",\n",
      "        \"variantRsId\"\n",
      "      ]\n",
      "    }\n",
      "  ],\n",
      "  \"definitions\": {\n",
      "    \"confidenceIntervalLower\": {\n",
      "      \"type\": [\n",
      "        \"number\",\n",
      "        \"null\"\n",
      "      ],\n",
      "      \"description\": \"Lower value of the confidence interval\"\n",
      "    },\n",
      "    \"confidenceIntervalUpper\": {\n",
      "      \"type\": [\n",
      "        \"number\",\n",
      "        \"null\"\n",
      "      ],\n",
      "      \"description\": \"Upper value of the confidence interval\"\n",
      "    },\n",
      "    \"datasourceId\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"diseaseFromSource\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"diseaseId\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"id\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"literature\": {\n",
      "      \"type\": [\n",
      "        \"string\",\n",
      "        \"null\"\n",
      "      ]\n",
      "    },\n",
      "    \"locus2GeneScore\": {\n",
      "      \"type\": \"number\",\n",
      "      \"description\": \"Locus to gene score\",\n",
      "      \"maximum\": 1.0,\n",
      "      \"exclusiveMinimum\": 0.0\n",
      "    },\n",
      "    \"oddsRatio\": {\n",
      "      \"type\": [\n",
      "        \"number\",\n",
      "        \"null\"\n",
      "      ],\n",
      "      \"description\": \"Odds ratio of the association\"\n",
      "    },\n",
      "    \"publicationFirstAuthor\": {\n",
      "      \"type\": [\n",
      "        \"string\",\n",
      "        \"null\"\n",
      "      ],\n",
      "      \"description\": \"First author of the publication\"\n",
      "    },\n",
      "    \"publicationYear\": {\n",
      "      \"type\": [\n",
      "        \"integer\",\n",
      "        \"null\"\n",
      "      ],\n",
      "      \"description\": \"Year of publication\"\n",
      "    },\n",
      "    \"resourceScore\": {\n",
      "      \"type\": [\n",
      "        \"number\",\n",
      "        \"null\"\n",
      "      ]\n",
      "    },\n",
      "    \"resourceScoreExponent\": {\n",
      "      \"type\": \"integer\",\n",
      "      \"description\": \"Exponent of the association p-value\",\n",
      "      \"maximum\": 0.0\n",
      "    },\n",
      "    \"resourceScoreMantissa\": {\n",
      "      \"type\": \"number\",\n",
      "      \"description\": \"Mantissa of the association p-value\",\n",
      "      \"exclusiveMinimum\": 0.0\n",
      "    },\n",
      "    \"score\": {\n",
      "      \"type\": \"number\"\n",
      "    },\n",
      "    \"studyId\": {\n",
      "      \"type\": \"string\",\n",
      "      \"description\": \"GWAS study accession\"\n",
      "    },\n",
      "    \"studySampleSize\": {\n",
      "      \"type\": \"integer\",\n",
      "      \"description\": \"Sample size of the GWAS study\",\n",
      "      \"exclusiveMinimum\": 0.0\n",
      "    },\n",
      "    \"targetId\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"variantFunctionalConsequenceId\": {\n",
      "      \"type\": \"string\",\n",
      "      \"description\": \"SO code of the functional consequence of the variant\"\n",
      "    },\n",
      "    \"variantId\": {\n",
      "      \"type\": \"string\",\n",
      "      \"description\": \"Identifier of the variant in the OpenTargets Genetics portal\"\n",
      "    },\n",
      "    \"variantRsId\": {\n",
      "      \"type\": [\n",
      "        \"string\",\n",
      "        \"null\"\n",
      "      ],\n",
      "      \"description\": \"rs identifier of the variant\",\n",
      "      \"pattern\": \"^rs[0-9]{1,}$\"\n",
      "    }\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Reloading possible modifications from the xlsx file:\n",
    "fields_df = pd.read_excel(field_description)\n",
    "\n",
    "# constants for the json schema:\n",
    "schema_obj = OrderedDict({\n",
    "    \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
    "    \"title\": \"OpenTargets\",\n",
    "    \"description\": \"OpenTargets evidence objects\",\n",
    "    \"type\": \"object\",\n",
    "    \"oneOf\": {},\n",
    "    \"definitions\": OrderedDict()\n",
    "})\n",
    "\n",
    "# Adding all the required fields for all data sources:\n",
    "schema_obj[\"oneOf\"] = parse_data_sources(fields_df)\n",
    "\n",
    "# Parsing dataframe to get list of fields for each data source:\n",
    "parsed_sources = defaultdict(list)\n",
    "\n",
    "for i, row in fields_df.iterrows():\n",
    "    field = row['backend_name']\n",
    "    \n",
    "    # Only using ot_genetics for now:\n",
    "    if 'ot_genetics_portal' not in row['data_source']:\n",
    "        continue\n",
    "\n",
    "    field_annotation = OrderedDict()\n",
    "    \n",
    "    # Setting type - maybe nullable:\n",
    "    if row['nullable']:\n",
    "        field_annotation['type'] = [row['type'], \"null\"]\n",
    "    else:\n",
    "        field_annotation['type'] = row['type']\n",
    "    \n",
    "    # Adding description:\n",
    "    if isinstance(row['description'], str):\n",
    "        field_annotation['description'] = row['description']\n",
    "\n",
    "    # Adding minimum:\n",
    "    if not np.isnan(row['minimum']):\n",
    "        field_annotation['minimum'] = row['minimum']\n",
    "        \n",
    "    # Adding description:\n",
    "    if not np.isnan(row['maximum']):\n",
    "        field_annotation['maximum'] = row['maximum']\n",
    "\n",
    "    # Adding minimum:\n",
    "    if not np.isnan(row['exclusiveMinimum']):\n",
    "        field_annotation['exclusiveMinimum'] = row['exclusiveMinimum']\n",
    "        \n",
    "    # Adding pattern:\n",
    "    if isinstance(row['pattern'], str):\n",
    "        field_annotation['pattern'] = row['pattern']\n",
    "        \n",
    "    schema_obj['definitions'][field] = field_annotation\n",
    "    \n",
    "    \n",
    "    \n",
    "print(json.dumps(schema_obj, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
