{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prototyping the new json schema\n",
    "\n",
    "\n",
    "As part of the consolidation of the evidence objects in the backend, we are re-modeling the [json schema](https://github.com/opentargets/json_schema) to reflect the new simplified/flattened design.\n",
    "\n",
    "**Link to ticket: [#1249](https://github.com/opentargets/platform/issues/1249)**\n",
    "\n",
    "Based on the meeting we had on 2020.11.11 the following conclusions were reached:\n",
    "\n",
    "* We need to maintain a json schema that guides our data providers and can be used as template to generate evidence strings.\n",
    "* The schema will reflect the concepts of the new platform design, so the units of the schema is going to be data source centric instead of data type.\n",
    "* Each of the valuable columns will be defined in a common section.\n",
    "* For each data source there will only be a list of required fields.\n",
    "* We haven't reached a consensus on how the unique association fields are defined, and at which point of the evidence generation. So for the first iteration of the json schema, the unique_association_fields will be omitted.\n",
    "\n",
    "The schema is written based on the most recent iteration of the [evidence schema review](https://docs.google.com/spreadsheets/d/11jdPCo_vxY3jaP54xKTsXBshR5HMrpUf5oXJNgtbKm8/edit#gid=1735847104) document.\n",
    "\n",
    "The technical approach:\n",
    "\n",
    "* To avoid manual work with the json document, I'm collating information in an excel file and will use that as a source for the definitions.\n",
    "* The same excel file will be used to get the source names from where we are expecting the given field.\n",
    "\n",
    "## The first run completed:\n",
    "\n",
    "- [X] processing the review document to get the rough list of fields\n",
    "- [x] get fields2datasource mapping\n",
    "- [x] generate json schema based on the meeting\n",
    "\n",
    "\n",
    "## The first run didn't cover:\n",
    "\n",
    "- [ ] some fields are missing eg. uniprot id\n",
    "- [ ] some fields shoudl not be here: `score` and `id`\n",
    "- [ ] the precise requirements of the fields are still sparse -> add more data to `field_description.xlsx`\n",
    "- [ ] no structure whatsoever."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Getting the list of data source for every field\n",
    "\n",
    "This information is extracted from the evidence schema review file. The end of the process is a comma separated list of data sources for every field. This column is used later to generate the mandatory list of fields for every data source."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-12T01:32:16.268651Z",
     "start_time": "2020-11-12T01:32:16.247276Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_gene_census</th>\n",
       "      <th>chembl</th>\n",
       "      <th>clingen</th>\n",
       "      <th>crispr</th>\n",
       "      <th>europepmc</th>\n",
       "      <th>eva</th>\n",
       "      <th>eva_somatic</th>\n",
       "      <th>expression_atlas</th>\n",
       "      <th>gene2phenotype</th>\n",
       "      <th>genomics_england</th>\n",
       "      <th>intogen</th>\n",
       "      <th>ot_genetics_portal</th>\n",
       "      <th>phenodigm</th>\n",
       "      <th>phewas_catalog</th>\n",
       "      <th>progeny</th>\n",
       "      <th>reactome</th>\n",
       "      <th>slapenrich</th>\n",
       "      <th>sysbio</th>\n",
       "      <th>uniprot_literature</th>\n",
       "      <th>uniprot_somatic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>allelicRequirement</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biologicalModelAllelicComposition</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>564310</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biologicalModelGeneticBackground</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>564310</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clinicalPhase</th>\n",
       "      <td>0</td>\n",
       "      <td>427943</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clinicalSignificance</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>107532</td>\n",
       "      <td>8173</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cancer_gene_census  chembl  clingen  \\\n",
       "allelicRequirement                                  0       0     1075   \n",
       "biologicalModelAllelicComposition                   0       0        0   \n",
       "biologicalModelGeneticBackground                    0       0        0   \n",
       "clinicalPhase                                       0  427943        0   \n",
       "clinicalSignificance                                0       0        0   \n",
       "\n",
       "                                   crispr  europepmc     eva  eva_somatic  \\\n",
       "allelicRequirement                      0          0       0            0   \n",
       "biologicalModelAllelicComposition       0          0       0            0   \n",
       "biologicalModelGeneticBackground        0          0       0            0   \n",
       "clinicalPhase                           0          0       0            0   \n",
       "clinicalSignificance                    0          0  107532         8173   \n",
       "\n",
       "                                   expression_atlas  gene2phenotype  \\\n",
       "allelicRequirement                                0            2451   \n",
       "biologicalModelAllelicComposition                 0               0   \n",
       "biologicalModelGeneticBackground                  0               0   \n",
       "clinicalPhase                                     0               0   \n",
       "clinicalSignificance                              0               0   \n",
       "\n",
       "                                   genomics_england  intogen  \\\n",
       "allelicRequirement                                0        0   \n",
       "biologicalModelAllelicComposition                 0        0   \n",
       "biologicalModelGeneticBackground                  0        0   \n",
       "clinicalPhase                                     0        0   \n",
       "clinicalSignificance                              0        0   \n",
       "\n",
       "                                   ot_genetics_portal  phenodigm  \\\n",
       "allelicRequirement                                  0          0   \n",
       "biologicalModelAllelicComposition                   0     564310   \n",
       "biologicalModelGeneticBackground                    0     564310   \n",
       "clinicalPhase                                       0          0   \n",
       "clinicalSignificance                                0          0   \n",
       "\n",
       "                                   phewas_catalog  progeny  reactome  \\\n",
       "allelicRequirement                              0        0         0   \n",
       "biologicalModelAllelicComposition               0        0         0   \n",
       "biologicalModelGeneticBackground                0        0         0   \n",
       "clinicalPhase                                   0        0         0   \n",
       "clinicalSignificance                            0        0         0   \n",
       "\n",
       "                                   slapenrich  sysbio  uniprot_literature  \\\n",
       "allelicRequirement                          0       0                   0   \n",
       "biologicalModelAllelicComposition           0       0                   0   \n",
       "biologicalModelGeneticBackground            0       0                   0   \n",
       "clinicalPhase                               0       0                   0   \n",
       "clinicalSignificance                        0       0                   0   \n",
       "\n",
       "                                   uniprot_somatic  \n",
       "allelicRequirement                               0  \n",
       "biologicalModelAllelicComposition                0  \n",
       "biologicalModelGeneticBackground                 0  \n",
       "clinicalPhase                                    0  \n",
       "clinicalSignificance                             0  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Get the source names for every field:\n",
    "import pandas as pd\n",
    "import json\n",
    "from collections import OrderedDict, defaultdict\n",
    "import numpy as np\n",
    "import requests\n",
    "\n",
    "notnull_table = pd.read_csv('iter9_notnull_table.tsv', sep='\\t')\n",
    "notnull_table.index = notnull_table.key.tolist()\n",
    "notnull_table.drop('key', axis=1, inplace=True)\n",
    "notnull_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-11T21:11:03.048756Z",
     "start_time": "2020-11-11T21:11:02.990035Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         clingen,gene2phenotype\n",
       "1                      phenodigm\n",
       "2                      phenodigm\n",
       "3                         chembl\n",
       "4                eva,eva_somatic\n",
       "                 ...            \n",
       "57    cancer_gene_census,intogen\n",
       "58                      reactome\n",
       "59    cancer_gene_census,intogen\n",
       "60    cancer_gene_census,intogen\n",
       "61            cancer_gene_census\n",
       "Length: 62, dtype: object"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looping through all fields and extracting all sources where the data is present:\n",
    "lookup = lambda x: ','.join(notnull_table.loc[x].where(lambda x: x != 0).dropna().index.to_list())\n",
    "pd.Series(notnull_table.index).apply(lookup)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading field description\n",
    "\n",
    "1. Read excel file with the field descriptions: from [google sheets](https://docs.google.com/spreadsheets/d/1vHoyIsQDBNmUfq2IUdZDoz3G457V5cqW/edit#gid=613969206)\n",
    "2. Parse values. \n",
    "3. Start building json object.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-17T21:51:59.458582Z",
     "start_time": "2020-11-17T21:51:59.436762Z"
    },
    "code_folding": [
     0,
     53,
     69,
     104
    ]
   },
   "outputs": [],
   "source": [
    "def get_dataframe():\n",
    "    '''\n",
    "    This function fetches the field definitinos from google spreadsheets\n",
    "    '''\n",
    "    field_description = 'field_descriptions.tsv'\n",
    "\n",
    "    # The file from now on is stored on google sheets:\n",
    "    url = 'https://docs.google.com/spreadsheets/d/1vHoyIsQDBNmUfq2IUdZDoz3G457V5cqW/export?format=tsv&id=1vHoyIsQDBNmUfq2IUdZDoz3G457V5cqW&gid=613969206'\n",
    "    r = requests.get(url)\n",
    "\n",
    "    with open(field_description, 'w') as f:\n",
    "        f.write(r.text)\n",
    "\n",
    "    fields_df = pd.read_csv(field_description, sep='\\t')\n",
    "    return fields_df\n",
    "\n",
    "\n",
    "def parse_data_sources(df):\n",
    "    '''\n",
    "    Parsing dataframe to get list of fields for each data source:\n",
    "    \n",
    "    input: features_df\n",
    "    output: oneof ordered dictionary\n",
    "    '''\n",
    "    \n",
    "    parsed_sources = defaultdict(list)\n",
    "    for i, row in df.iterrows():\n",
    "        for source in row['data_source'].split(','):                \n",
    "            parsed_sources[source].append(row['field_name'])\n",
    "\n",
    "    # Each data source then exploded into schemas:\n",
    "    source_schemas = []\n",
    "    sorted_items = sorted(parsed_sources.items())\n",
    "    for source, fields in sorted_items:\n",
    "        source_schema = OrderedDict()\n",
    "\n",
    "        # Adding property definitions:\n",
    "        source_schema['properties'] = OrderedDict({'datasourceId': {\"const\": source}})\n",
    "        \n",
    "        for field in fields:\n",
    "            if field == 'datasourceId':\n",
    "                continue\n",
    "                \n",
    "            source_schema['properties'][field] = {\"$ref\": f\"#/definitions/{field}\"}\n",
    "        \n",
    "        source_schema['required'] = []\n",
    "        source_schema[\"additionalItems\"] = False\n",
    "\n",
    "        # Adding source schema:\n",
    "        source_schemas.append(source_schema)\n",
    "        \n",
    "    return(source_schemas)\n",
    "\n",
    "\n",
    "def add_definition(row, fields_df):\n",
    "    '''\n",
    "    This is the main function to add definitions to the schema\n",
    "    '''\n",
    "    \n",
    "    # If the feature is simple:\n",
    "    if row['type'] in ['string', 'integer', 'number']:\n",
    "        return add_simple_definition(row)\n",
    "    \n",
    "    # If the feature is complex:\n",
    "    elif row['type'] == 'array':\n",
    "        print(row[\"field_name\"])\n",
    "        return add_array(row, fields_df)\n",
    "        print(f'complex objectfound: {row[\"field_name\"]}')\n",
    "    \n",
    "    \n",
    "def add_array(row, df):\n",
    "    '''\n",
    "    If the definition is an array, things have to be treated separately\n",
    "    '''\n",
    "    field_annotation = OrderedDict({'type':'array'})\n",
    "    field_name = row['field_name']\n",
    "    \n",
    "    # If there's a description:\n",
    "    if isinstance(row['description'], str):\n",
    "        field_annotation['description'] = row['description']\n",
    "\n",
    "    # Items are objects: \n",
    "    if len(df.loc[df.location == field_name]) > 0:\n",
    "        field_annotation['items'] = OrderedDict({\n",
    "            'type': \"object\",\n",
    "            'properties': OrderedDict()\n",
    "        })\n",
    "        \n",
    "        sub_df = df.loc[df.location == field_name]\n",
    "        for index, sub_row in sub_df.iterrows():\n",
    "             field_annotation['items']['properties'][sub_row['field_name']] = add_definition(sub_row, sub_df)\n",
    "\n",
    "        \n",
    "    # Items are string:\n",
    "    else:\n",
    "        row['type'] = 'string'\n",
    "        row['nullable'] = False\n",
    "\n",
    "        field_annotation['items'] = add_simple_definition(row)\n",
    "        \n",
    "    field_annotation[\"uniqueItems\"] = True\n",
    "    \n",
    "    return field_annotation\n",
    "    \n",
    "    \n",
    "def add_simple_definition(row):\n",
    "    '''\n",
    "    If the feature is a simple object handled easy.\n",
    "    '''\n",
    "    field = row['field_name']\n",
    "\n",
    "    field_annotation = OrderedDict()\n",
    "    \n",
    "    # Setting type - maybe nullable:\n",
    "    if row['nullable']:\n",
    "        field_annotation['type'] = [row['type'], \"null\"]\n",
    "    else:\n",
    "        field_annotation['type'] = row['type']\n",
    "    \n",
    "    # Adding description:\n",
    "    if isinstance(row['description'], str):\n",
    "        field_annotation['description'] = row['description']\n",
    "\n",
    "    # Adding minimum:\n",
    "    if not np.isnan(row['minimum']):\n",
    "        field_annotation['minimum'] = row['minimum']\n",
    "        \n",
    "    # Adding maximum:\n",
    "    if not np.isnan(row['maximum']):\n",
    "        field_annotation['maximum'] = row['maximum']\n",
    "\n",
    "    # Is it exclusive minimum:\n",
    "    if not np.isnan(row['exclusiveMinimum']):\n",
    "        field_annotation['exclusiveMinimum'] = row['exclusiveMinimum']\n",
    "        \n",
    "    # Adding pattern:\n",
    "    if isinstance(row['pattern'], str):\n",
    "        field_annotation['pattern'] = row['pattern']\n",
    "        \n",
    "    # Is there a list of accepted values (might be a list of floats!!):\n",
    "    if isinstance(row['accepted_values'], str):\n",
    "        enum_values = row['accepted_values'].split(';')\n",
    "        try:\n",
    "            field_annotation['enum'] = [float(x) for x in enum_values]\n",
    "        except:\n",
    "            field_annotation['enum'] = enum_values\n",
    "        \n",
    "    return field_annotation\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-17T21:52:02.810567Z",
     "start_time": "2020-11-17T21:52:00.642764Z"
    },
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biologicalModel\n",
      "clinicalSignificance\n",
      "clinicalUrls\n",
      "diseaseModelAssociatedHumanPhenotypes\n",
      "diseaseModelAssociatedModelPhenotypes\n",
      "literature\n",
      "significantDriverMethods\n",
      "textMiningSentences\n",
      "variations\n"
     ]
    }
   ],
   "source": [
    "# Reloading possible modifications from the xlsx file:\n",
    "fields_df = get_dataframe()\n",
    "\n",
    "##\n",
    "## initialize json schema:\n",
    "##\n",
    "schema_obj = OrderedDict({\n",
    "    \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
    "    \"title\": \"OpenTargets\",\n",
    "    \"description\": \"OpenTargets evidence objects\",\n",
    "    \"type\": \"object\",\n",
    "    \"oneOf\": {},\n",
    "    \"definitions\": OrderedDict()\n",
    "})\n",
    "\n",
    "# Adding descxription for all data sources:\n",
    "schema_obj[\"oneOf\"] = parse_data_sources(fields_df.loc[fields_df.location == 'root'])\n",
    "\n",
    "##\n",
    "## Generating field definitions:\n",
    "##\n",
    "    \n",
    "# This will be updated with every row:\n",
    "root_object = OrderedDict() \n",
    "fields_df = get_dataframe()\n",
    "\n",
    "# Looping through all fields that are in the root of the document:\n",
    "for i, row in fields_df.loc[fields_df.location == 'root'].iterrows():\n",
    "    schema_obj['definitions'][row['field_name']] = add_definition(row, fields_df)\n",
    "\n",
    "# Saving object into json file:\n",
    "with open('json_schema_v0.2.json', 'w') as f:\n",
    "    json.dump(schema_obj, f, indent=2)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-17T21:48:41.174013Z",
     "start_time": "2020-11-17T21:48:41.144459Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>field_name</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>description</th>\n",
       "      <th>minimum</th>\n",
       "      <th>exclusiveMinimum</th>\n",
       "      <th>maximum</th>\n",
       "      <th>accepted_values</th>\n",
       "      <th>nullable</th>\n",
       "      <th>pattern</th>\n",
       "      <th>data_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>allelicRequirement</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clingen,gene2phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>biologicalModel</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>clinicalPhase</td>\n",
       "      <td>integer</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>clinicalSignificance</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eva,eva_somatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>clinicalStatus</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>clinicalUrls</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cohortDescription</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cohortId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cohortShortName</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>confidence</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clingen,gene2phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>confidenceIntervalLower</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>Lower value of the confidence interval</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>confidenceIntervalUpper</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>Upper value of the confidence interval</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>contrast</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>expression_atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>datasourceId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cancer_gene_census,chembl,clingen,crispr,europ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>diseaseFromSource</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clingen,eva,eva_somatic,gene2phenotype,genomic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>diseaseId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cancer_gene_census,chembl,clingen,crispr,europ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>diseaseModelAssociatedHumanPhenotypes</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>diseaseModelAssociatedModelPhenotypes</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenodigm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>drugId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>literature</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cancer_gene_census,crispr,europepmc,eva,eva_so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>locus2GeneScore</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>Locus to gene score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>log2FoldChangePercentileRank</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>expression_atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>log2FoldChangeValue</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>expression_atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>oddsRatio</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>Odds ratio of the association</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal,phewas_catalog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>pathwayId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>progeny,reactome,slapenrich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>pathwayName</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>progeny,reactome,slapenrich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>publicationFirstAuthor</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>First author of the publication</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>publicationYear</td>\n",
       "      <td>integer</td>\n",
       "      <td>root</td>\n",
       "      <td>Year of publication</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>recordId</td>\n",
       "      <td>integer</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clingen,eva,eva_somatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>resourceScore</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cancer_gene_census,clingen,crispr,europepmc,ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>resourceScoreExponent</td>\n",
       "      <td>integer</td>\n",
       "      <td>root</td>\n",
       "      <td>Exponent of the association p-value</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>resourceScoreMantissa</td>\n",
       "      <td>number</td>\n",
       "      <td>root</td>\n",
       "      <td>Mantissa of the association p-value</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>resourceScoreType</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cancer_gene_census,clingen,crispr,europepmc,ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>significantDriverMethods</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>studyCases</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phewas_catalog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>studyId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>Identifier of the study</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>expression_atlas,genomics_england,ot_genetics_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>studyOverview</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>expression_atlas,sysbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>studySampleSize</td>\n",
       "      <td>integer</td>\n",
       "      <td>root</td>\n",
       "      <td>Sample size of the GWAS study</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>targetId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cancer_gene_census,chembl,clingen,crispr,europ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>targetModulation</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reactome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>textMiningSentences</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>europepmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>SO code of the functional consequence of the v...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eva,ot_genetics_portal,phewas_catalog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>variantId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>Identifier of the variant in the OpenTargets G...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ot_genetics_portal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>variantRsId</td>\n",
       "      <td>string</td>\n",
       "      <td>root</td>\n",
       "      <td>rs identifier of the variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>^rs[0-9]{1,}$</td>\n",
       "      <td>eva,ot_genetics_portal,phewas_catalog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>variations</td>\n",
       "      <td>array</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intogen,reactome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               field_name     type location  \\\n",
       "0                      allelicRequirement   string     root   \n",
       "1                         biologicalModel    array     root   \n",
       "4                           clinicalPhase  integer     root   \n",
       "5                    clinicalSignificance    array     root   \n",
       "6                          clinicalStatus   string     root   \n",
       "7                            clinicalUrls    array     root   \n",
       "10                      cohortDescription   string     root   \n",
       "11                               cohortId   string     root   \n",
       "12                        cohortShortName   string     root   \n",
       "13                             confidence   string     root   \n",
       "14                confidenceIntervalLower   number     root   \n",
       "15                confidenceIntervalUpper   number     root   \n",
       "16                               contrast   string     root   \n",
       "17                           datasourceId   string     root   \n",
       "18                      diseaseFromSource   string     root   \n",
       "19                              diseaseId   string     root   \n",
       "20  diseaseModelAssociatedHumanPhenotypes    array     root   \n",
       "23  diseaseModelAssociatedModelPhenotypes    array     root   \n",
       "26                                 drugId   string     root   \n",
       "28                             literature    array     root   \n",
       "29                        locus2GeneScore   number     root   \n",
       "30           log2FoldChangePercentileRank   number     root   \n",
       "31                    log2FoldChangeValue   number     root   \n",
       "32                              oddsRatio   number     root   \n",
       "33                              pathwayId   string     root   \n",
       "34                            pathwayName   string     root   \n",
       "35                 publicationFirstAuthor   string     root   \n",
       "36                        publicationYear  integer     root   \n",
       "37                               recordId  integer     root   \n",
       "38                          resourceScore   number     root   \n",
       "39                  resourceScoreExponent  integer     root   \n",
       "40                  resourceScoreMantissa   number     root   \n",
       "41                      resourceScoreType   string     root   \n",
       "43               significantDriverMethods    array     root   \n",
       "44                             studyCases   string     root   \n",
       "45                                studyId   string     root   \n",
       "46                          studyOverview   string     root   \n",
       "47                        studySampleSize  integer     root   \n",
       "48                               targetId   string     root   \n",
       "49                       targetModulation   string     root   \n",
       "50                    textMiningSentences    array     root   \n",
       "57         variantFunctionalConsequenceId   string     root   \n",
       "59                              variantId   string     root   \n",
       "60                            variantRsId   string     root   \n",
       "61                             variations    array     root   \n",
       "\n",
       "                                          description  minimum  \\\n",
       "0                                                 NaN      NaN   \n",
       "1                                                 NaN      NaN   \n",
       "4                                                 NaN      1.0   \n",
       "5                                                 NaN      NaN   \n",
       "6                                                 NaN      NaN   \n",
       "7                                                 NaN      NaN   \n",
       "10                                                NaN      NaN   \n",
       "11                                                NaN      NaN   \n",
       "12                                                NaN      NaN   \n",
       "13                                                NaN      NaN   \n",
       "14             Lower value of the confidence interval      NaN   \n",
       "15             Upper value of the confidence interval      NaN   \n",
       "16                                                NaN      NaN   \n",
       "17                                                NaN      NaN   \n",
       "18                                                NaN      NaN   \n",
       "19                                                NaN      NaN   \n",
       "20                                                NaN      NaN   \n",
       "23                                                NaN      NaN   \n",
       "26                                                NaN      NaN   \n",
       "28                                                NaN      NaN   \n",
       "29                                Locus to gene score      NaN   \n",
       "30                                                NaN      NaN   \n",
       "31                                                NaN      NaN   \n",
       "32                      Odds ratio of the association      NaN   \n",
       "33                                                NaN      NaN   \n",
       "34                                                NaN      NaN   \n",
       "35                    First author of the publication      NaN   \n",
       "36                                Year of publication      NaN   \n",
       "37                                                NaN      NaN   \n",
       "38                                                NaN      NaN   \n",
       "39                Exponent of the association p-value      NaN   \n",
       "40                Mantissa of the association p-value      NaN   \n",
       "41                                                NaN      NaN   \n",
       "43                                                NaN      NaN   \n",
       "44                                                NaN      NaN   \n",
       "45                            Identifier of the study      NaN   \n",
       "46                                                NaN      NaN   \n",
       "47                      Sample size of the GWAS study      NaN   \n",
       "48                                                NaN      NaN   \n",
       "49                                                NaN      NaN   \n",
       "50                                                NaN      NaN   \n",
       "57  SO code of the functional consequence of the v...      NaN   \n",
       "59  Identifier of the variant in the OpenTargets G...      NaN   \n",
       "60                       rs identifier of the variant      NaN   \n",
       "61                                                NaN      NaN   \n",
       "\n",
       "    exclusiveMinimum  maximum  accepted_values nullable        pattern  \\\n",
       "0                NaN      NaN              NaN      NaN            NaN   \n",
       "1                NaN      NaN              NaN      NaN            NaN   \n",
       "4                NaN      4.0              NaN      NaN            NaN   \n",
       "5                NaN      NaN              NaN      NaN            NaN   \n",
       "6                NaN      NaN              NaN      NaN            NaN   \n",
       "7                NaN      NaN              NaN      NaN            NaN   \n",
       "10               NaN      NaN              NaN      NaN            NaN   \n",
       "11               NaN      NaN              NaN      NaN            NaN   \n",
       "12               NaN      NaN              NaN      NaN            NaN   \n",
       "13               NaN      NaN              NaN      NaN            NaN   \n",
       "14               NaN      NaN              NaN     True            NaN   \n",
       "15               NaN      NaN              NaN     True            NaN   \n",
       "16               NaN      NaN              NaN      NaN            NaN   \n",
       "17               NaN      NaN              NaN    False            NaN   \n",
       "18               NaN      NaN              NaN    False            NaN   \n",
       "19               NaN      NaN              NaN    False            NaN   \n",
       "20               NaN      NaN              NaN      NaN            NaN   \n",
       "23               NaN      NaN              NaN      NaN            NaN   \n",
       "26               NaN      NaN              NaN      NaN            NaN   \n",
       "28               NaN      NaN              NaN     True            NaN   \n",
       "29               0.0      1.0              NaN    False            NaN   \n",
       "30               NaN      NaN              NaN      NaN            NaN   \n",
       "31               NaN      NaN              NaN      NaN            NaN   \n",
       "32               NaN      NaN              NaN     True            NaN   \n",
       "33               NaN      NaN              NaN      NaN            NaN   \n",
       "34               NaN      NaN              NaN      NaN            NaN   \n",
       "35               NaN      NaN              NaN     True            NaN   \n",
       "36               NaN      NaN              NaN     True            NaN   \n",
       "37               NaN      NaN              NaN      NaN            NaN   \n",
       "38               NaN      NaN              NaN      NaN            NaN   \n",
       "39               NaN      0.0              NaN    False            NaN   \n",
       "40               0.0      NaN              NaN    False            NaN   \n",
       "41               NaN      NaN              NaN      NaN            NaN   \n",
       "43               NaN      NaN              NaN      NaN            NaN   \n",
       "44               NaN      NaN              NaN      NaN            NaN   \n",
       "45               NaN      NaN              NaN    False            NaN   \n",
       "46               NaN      NaN              NaN      NaN            NaN   \n",
       "47               0.0      NaN              NaN    False            NaN   \n",
       "48               NaN      NaN              NaN    False            NaN   \n",
       "49               NaN      NaN              NaN      NaN            NaN   \n",
       "50               NaN      NaN              NaN      NaN            NaN   \n",
       "57               NaN      NaN              NaN    False            NaN   \n",
       "59               NaN      NaN              NaN    False            NaN   \n",
       "60               NaN      NaN              NaN     True  ^rs[0-9]{1,}$   \n",
       "61               NaN      NaN              NaN      NaN            NaN   \n",
       "\n",
       "                                          data_source  \n",
       "0                              clingen,gene2phenotype  \n",
       "1                                           phenodigm  \n",
       "4                                              chembl  \n",
       "5                                     eva,eva_somatic  \n",
       "6                                              chembl  \n",
       "7                                              chembl  \n",
       "10                                            intogen  \n",
       "11                                            intogen  \n",
       "12                                            intogen  \n",
       "13                             clingen,gene2phenotype  \n",
       "14                                 ot_genetics_portal  \n",
       "15                                 ot_genetics_portal  \n",
       "16                                   expression_atlas  \n",
       "17  cancer_gene_census,chembl,clingen,crispr,europ...  \n",
       "18  clingen,eva,eva_somatic,gene2phenotype,genomic...  \n",
       "19  cancer_gene_census,chembl,clingen,crispr,europ...  \n",
       "20                                          phenodigm  \n",
       "23                                          phenodigm  \n",
       "26                                             chembl  \n",
       "28  cancer_gene_census,crispr,europepmc,eva,eva_so...  \n",
       "29                                 ot_genetics_portal  \n",
       "30                                   expression_atlas  \n",
       "31                                   expression_atlas  \n",
       "32                  ot_genetics_portal,phewas_catalog  \n",
       "33                        progeny,reactome,slapenrich  \n",
       "34                        progeny,reactome,slapenrich  \n",
       "35                                 ot_genetics_portal  \n",
       "36                                 ot_genetics_portal  \n",
       "37                            clingen,eva,eva_somatic  \n",
       "38  cancer_gene_census,clingen,crispr,europepmc,ev...  \n",
       "39                                 ot_genetics_portal  \n",
       "40                                 ot_genetics_portal  \n",
       "41  cancer_gene_census,clingen,crispr,europepmc,ev...  \n",
       "43                                            intogen  \n",
       "44                                     phewas_catalog  \n",
       "45  expression_atlas,genomics_england,ot_genetics_...  \n",
       "46                            expression_atlas,sysbio  \n",
       "47                                 ot_genetics_portal  \n",
       "48  cancer_gene_census,chembl,clingen,crispr,europ...  \n",
       "49                                           reactome  \n",
       "50                                          europepmc  \n",
       "57              eva,ot_genetics_portal,phewas_catalog  \n",
       "59                                 ot_genetics_portal  \n",
       "60              eva,ot_genetics_portal,phewas_catalog  \n",
       "61                                   intogen,reactome  "
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fields_df.loc[fields_df.location == 'root'].sort_values(['field_name'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
