{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bdddde54",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T21:06:55.707448Z",
     "start_time": "2022-05-23T21:06:50.127121Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r--  1 dsuveges  384566875    10M 23 May 22:06 gwas_ancestries_2022-05-23.tsv\n",
      "-rw-r--r--  1 dsuveges  384566875   227M 20 May 06:31 gwas_associations_2022-05-20.tsv\n",
      "-rw-r--r--  1 dsuveges  384566875   227M 23 May 16:41 gwas_associations_2022-05-23.tsv\n",
      "-rw-r--r--  1 dsuveges  384566875    17M 20 May 06:31 gwas_studies_2022-05-20.tsv\n",
      "-rw-r--r--  1 dsuveges  384566875    17M 23 May 16:41 gwas_studies_2022-05-23.tsv\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "\n",
    "today=$(date '+%Y-%m-%d')\n",
    "\n",
    "# wget -q https://www.ebi.ac.uk/gwas/api/search/downloads/alternative \\\n",
    "#     -O gwas_associations_${today}.tsv\n",
    "    \n",
    "# wget -q https://www.ebi.ac.uk/gwas/api/search/downloads/studies_alternative \\\n",
    "#     -O gwas_studies_${today}.tsv\n",
    "    \n",
    "wget -q  https://www.ebi.ac.uk/gwas/api/search/downloads/ancestry    \\\n",
    "    -O gwas_ancestries_${today}.tsv\n",
    "\n",
    "ls -lah *tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b5698f5a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-18T07:51:28.513004Z",
     "start_time": "2022-08-18T07:51:28.292708Z"
    }
   },
   "outputs": [
    {
     "ename": "Py4JJavaError",
     "evalue": "An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.\n: java.lang.AssertionError: assertion failed: Expected hostname or IPv6 IP enclosed in [] but got fe80:0:0:0:ff98:a2e4:f967:1ff7%14\n\tat scala.Predef$.assert(Predef.scala:223)\n\tat org.apache.spark.util.Utils$.checkHost(Utils.scala:1072)\n\tat org.apache.spark.executor.Executor.<init>(Executor.scala:89)\n\tat org.apache.spark.scheduler.local.LocalEndpoint.<init>(LocalSchedulerBackend.scala:64)\n\tat org.apache.spark.scheduler.local.LocalSchedulerBackend.start(LocalSchedulerBackend.scala:132)\n\tat org.apache.spark.scheduler.TaskSchedulerImpl.start(TaskSchedulerImpl.scala:220)\n\tat org.apache.spark.SparkContext.<init>(SparkContext.scala:579)\n\tat org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:58)\n\tat java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)\n\tat java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:247)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:238)\n\tat py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:80)\n\tat py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:69)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPy4JJavaError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-1cb0a925343b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m spark = (pyspark.sql.SparkSession\n\u001b[0m\u001b[1;32m      7\u001b[0m        \u001b[0;34m.\u001b[0m\u001b[0mbuilder\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m        \u001b[0;34m.\u001b[0m\u001b[0mmaster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"local[*]\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pyspark/sql/session.py\u001b[0m in \u001b[0;36mgetOrCreate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    226\u001b[0m                             \u001b[0msparkConf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m                         \u001b[0;31m# This SparkContext may be an existing one.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 228\u001b[0;31m                         \u001b[0msc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSparkContext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetOrCreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msparkConf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    229\u001b[0m                     \u001b[0;31m# Do not update `SparkConf` for existing `SparkContext`, as it's shared\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    230\u001b[0m                     \u001b[0;31m# by all sessions.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pyspark/context.py\u001b[0m in \u001b[0;36mgetOrCreate\u001b[0;34m(cls, conf)\u001b[0m\n\u001b[1;32m    382\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mSparkContext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_lock\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    383\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mSparkContext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_active_spark_context\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 384\u001b[0;31m                 \u001b[0mSparkContext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconf\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mSparkConf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    385\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mSparkContext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_active_spark_context\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    386\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pyspark/context.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, master, appName, sparkHome, pyFiles, environment, batchSize, serializer, conf, gateway, jsc, profiler_cls)\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0mSparkContext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ensure_initialized\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgateway\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgateway\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    145\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 146\u001b[0;31m             self._do_init(master, appName, sparkHome, pyFiles, environment, batchSize, serializer,\n\u001b[0m\u001b[1;32m    147\u001b[0m                           conf, jsc, profiler_cls)\n\u001b[1;32m    148\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pyspark/context.py\u001b[0m in \u001b[0;36m_do_init\u001b[0;34m(self, master, appName, sparkHome, pyFiles, environment, batchSize, serializer, conf, jsc, profiler_cls)\u001b[0m\n\u001b[1;32m    207\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[0;31m# Create the Java SparkContext through Py4J\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 209\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jsc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjsc\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_initialize_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_conf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jconf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    210\u001b[0m         \u001b[0;31m# Reset the SparkConf to the one actually used by the SparkContext in JVM.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    211\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_conf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSparkConf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_jconf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jsc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pyspark/context.py\u001b[0m in \u001b[0;36m_initialize_context\u001b[0;34m(self, jconf)\u001b[0m\n\u001b[1;32m    319\u001b[0m         \u001b[0mInitialize\u001b[0m \u001b[0mSparkContext\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfunction\u001b[0m \u001b[0mto\u001b[0m \u001b[0mallow\u001b[0m \u001b[0msubclass\u001b[0m \u001b[0mspecific\u001b[0m \u001b[0minitialization\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    320\u001b[0m         \"\"\"\n\u001b[0;32m--> 321\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mJavaSparkContext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjconf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    322\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    323\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1566\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1567\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_gateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1568\u001b[0;31m         return_value = get_return_value(\n\u001b[0m\u001b[1;32m   1569\u001b[0m             answer, self._gateway_client, None, self._fqn)\n\u001b[1;32m   1570\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/py4j/protocol.py\u001b[0m in \u001b[0;36mget_return_value\u001b[0;34m(answer, gateway_client, target_id, name)\u001b[0m\n\u001b[1;32m    324\u001b[0m             \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mOUTPUT_CONVERTER\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manswer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgateway_client\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0manswer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mREFERENCE_TYPE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m                 raise Py4JJavaError(\n\u001b[0m\u001b[1;32m    327\u001b[0m                     \u001b[0;34m\"An error occurred while calling {0}{1}{2}.\\n\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m                     format(target_id, \".\", name), value)\n",
      "\u001b[0;31mPy4JJavaError\u001b[0m: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.\n: java.lang.AssertionError: assertion failed: Expected hostname or IPv6 IP enclosed in [] but got fe80:0:0:0:ff98:a2e4:f967:1ff7%14\n\tat scala.Predef$.assert(Predef.scala:223)\n\tat org.apache.spark.util.Utils$.checkHost(Utils.scala:1072)\n\tat org.apache.spark.executor.Executor.<init>(Executor.scala:89)\n\tat org.apache.spark.scheduler.local.LocalEndpoint.<init>(LocalSchedulerBackend.scala:64)\n\tat org.apache.spark.scheduler.local.LocalSchedulerBackend.start(LocalSchedulerBackend.scala:132)\n\tat org.apache.spark.scheduler.TaskSchedulerImpl.start(TaskSchedulerImpl.scala:220)\n\tat org.apache.spark.SparkContext.<init>(SparkContext.scala:579)\n\tat org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:58)\n\tat java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)\n\tat java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:247)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:238)\n\tat py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:80)\n\tat py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:69)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\n"
     ]
    }
   ],
   "source": [
    "import pyspark.sql\n",
    "import pyspark.sql.types as t\n",
    "import pyspark.sql.functions as f\n",
    "\n",
    "\n",
    "spark = (pyspark.sql.SparkSession\n",
    "       .builder\n",
    "       .master(\"local[*]\")\n",
    "       .getOrCreate()\n",
    ")\n",
    "        \n",
    "# Reading GWAS Catalog data into spark dataframe\n",
    "gwas_df = spark.read.csv('gwas_associations_2022-05-20.tsv', sep='\\t', header=True).persist()\n",
    "\n",
    "print(gwas_df.count())\n",
    "gwas_df.show(1,False,True)\n",
    "gwas_df.printSchema()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "90a3211f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T15:45:19.507469Z",
     "start_time": "2022-05-23T15:45:18.203488Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Study count: 15844\n",
      "Split count by variants:\n",
      "   By CHR_ID: 220\n",
      "   By CHR_POS: 220\n",
      "   By CHR_ID or CHR_POS: 220\n",
      "   By SNPS: 1956\n",
      "   By STRONGEST SNP-RISK ALLELE: 1956\n"
     ]
    }
   ],
   "source": [
    "print('Study count: {}'.format(gwas_df.select('STUDY ACCESSION').distinct().count()))\n",
    "print('Split count by variants:')\n",
    "print('   By CHR_ID: {}'.format(gwas_df.filter(f.col('CHR_ID').contains(';')).count()))\n",
    "print('   By CHR_POS: {}'.format(gwas_df.filter(f.col('CHR_POS').contains(';')).count()))\n",
    "print('   By CHR_ID or CHR_POS: {}'.format(gwas_df.filter(f.col('CHR_ID').contains(';') | f.col('CHR_POS').contains(';')).count()))\n",
    "print('   By SNPS: {}'.format(gwas_df.filter(f.col('SNPS').contains(';')).count()))\n",
    "print('   By STRONGEST SNP-RISK ALLELE: {}'.format(gwas_df.filter(f.col('STRONGEST SNP-RISK ALLELE').contains(';')).count()))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "dffea7b6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T16:34:47.687371Z",
     "start_time": "2022-05-23T16:34:47.504900Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0-----------------------------------------------------------------------\n",
      " CHR                       | [6, 6]                                             \n",
      " POS                       | [32353777, 32421871]                               \n",
      " SNPS                      | [rs2395148,  rs2856683,  rs3135363,  rs9357152]    \n",
      " SNP_ID_CURRENT            | null                                               \n",
      " STUDY ACCESSION           | GCST000408                                         \n",
      " STRONGEST SNP-RISK ALLELE | rs2395148-A; rs2856683-C; rs3135363-A; rs9357152-A \n",
      "only showing top 1 row\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    gwas_df\n",
    "    .filter(f.col('CHR_POS').contains(';'))\n",
    "    .select(\n",
    "        f.split(f.col('CHR_ID'), ';').alias('CHR'), \n",
    "        f.split(f.col('CHR_POS'), ';').alias('POS'), \n",
    "        f.split(f.col(\"SNPS\"), ';').alias('SNPS'),\n",
    "        f.col('SNP_ID_CURRENT'),\n",
    "        f.col('STUDY ACCESSION'),\n",
    "        f.col('STRONGEST SNP-RISK ALLELE')\n",
    "    )\n",
    "    .filter(f.size('POS') != f.size('SNPS'))\n",
    "    .show(1, vertical=True, truncate=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6c780cbd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T16:35:15.709351Z",
     "start_time": "2022-05-23T16:35:15.545168Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0-------------------------------------------------------------------\n",
      " CHR_ID                    | null                                           \n",
      " CHR_POS                   | null                                           \n",
      " SNPS                      | rs17435444; rs13116936                         \n",
      " STUDY ACCESSION           | GCST001879                                     \n",
      " SNP_ID_CURRENT            | null                                           \n",
      " STRONGEST SNP-RISK ALLELE | rs17435444-A; rs13116936-G                     \n",
      "-RECORD 1-------------------------------------------------------------------\n",
      " CHR_ID                    | null                                           \n",
      " CHR_POS                   | null                                           \n",
      " SNPS                      | SNP_A-2032347; SNP_A-2196879; rs10871290       \n",
      " STUDY ACCESSION           | GCST000187                                     \n",
      " SNP_ID_CURRENT            | null                                           \n",
      " STRONGEST SNP-RISK ALLELE | SNP_A-2032347-?; SNP_A-2196879-?; rs10871290-? \n",
      "-RECORD 2-------------------------------------------------------------------\n",
      " CHR_ID                    | null                                           \n",
      " CHR_POS                   | null                                           \n",
      " SNPS                      | rs2235276; rs6923965; rs6908214                \n",
      " STUDY ACCESSION           | GCST008413                                     \n",
      " SNP_ID_CURRENT            | null                                           \n",
      " STRONGEST SNP-RISK ALLELE | rs2235276-T; rs6923965-C; rs6908214-C          \n",
      "only showing top 3 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    gwas_df\n",
    "    .filter(\n",
    "        f.col('SNPS').contains(';')\n",
    "        & f.col('CHR_POS').isNull()\n",
    "    )\n",
    "    .select('CHR_ID', 'CHR_POS', 'SNPS', 'STUDY ACCESSION', 'SNP_ID_CURRENT', f.col('STRONGEST SNP-RISK ALLELE'))\n",
    "    .show(3, False, True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "23b5d11d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-20T05:43:25.837697Z",
     "start_time": "2022-05-20T05:43:25.491877Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0-----------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2011-01-14                                                                                                                      \n",
      " PUBMEDID                   | 20801717                                                                                                                        \n",
      " FIRST AUTHOR               | Shatunov A                                                                                                                      \n",
      " DATE                       | 2010-10-01                                                                                                                      \n",
      " JOURNAL                    | Lancet Neurol                                                                                                                   \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/20801717                                                                                            \n",
      " STUDY                      | Chromosome 9p21 in sporadic amyotrophic lateral sclerosis in the UK and seven other countries: a genome-wide association study. \n",
      " DISEASE/TRAIT              | Amyotrophic lateral sclerosis                                                                                                   \n",
      " INITIAL SAMPLE SIZE        | 4,857 European ancestry cases, 8,987 European ancestry controls                                                                 \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                              \n",
      " REGION                     | 9p21.2                                                                                                                          \n",
      " CHR_ID                     | 9;9                                                                                                                             \n",
      " CHR_POS                    | 27556782;27543283                                                                                                               \n",
      " REPORTED GENE(S)           | intergenic                                                                                                                      \n",
      " MAPPED_GENE                | C9orf72; C9orf72                                                                                                                \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                            \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                            \n",
      " SNP_GENE_IDS               | null                                                                                                                            \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                            \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                            \n",
      " STRONGEST SNP-RISK ALLELE  | rs10122902-G; rs3849942-G                                                                                                       \n",
      " SNPS                       | rs10122902; rs3849942                                                                                                           \n",
      " MERGED                     | 0                                                                                                                               \n",
      " SNP_ID_CURRENT             | null                                                                                                                            \n",
      " CONTEXT                    | synonymous_variant; non_coding_transcript_exon_variant                                                                          \n",
      " INTERGENIC                 | null                                                                                                                            \n",
      " RISK ALLELE FREQUENCY      | 0.57                                                                                                                            \n",
      " P-VALUE                    | 5E-11                                                                                                                           \n",
      " PVALUE_MLOG                | 10.301029995663981                                                                                                              \n",
      " P-VALUE (TEXT)             | null                                                                                                                            \n",
      " OR or BETA                 | null                                                                                                                            \n",
      " 95% CI (TEXT)              | null                                                                                                                            \n",
      " PLATFORM [SNPS PASSING QC] | Illumina [227475]                                                                                                               \n",
      " CNV                        | N                                                                                                                               \n",
      " MAPPED_TRAIT               | amyotrophic lateral sclerosis                                                                                                   \n",
      " MAPPED_TRAIT_URI           | http://purl.obolibrary.org/obo/MONDO_0004976                                                                                    \n",
      " STUDY ACCESSION            | GCST000820                                                                                                                      \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                    \n",
      "-RECORD 1-----------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2011-01-14                                                                                                                      \n",
      " PUBMEDID                   | 20801717                                                                                                                        \n",
      " FIRST AUTHOR               | Shatunov A                                                                                                                      \n",
      " DATE                       | 2010-10-01                                                                                                                      \n",
      " JOURNAL                    | Lancet Neurol                                                                                                                   \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/20801717                                                                                            \n",
      " STUDY                      | Chromosome 9p21 in sporadic amyotrophic lateral sclerosis in the UK and seven other countries: a genome-wide association study. \n",
      " DISEASE/TRAIT              | Amyotrophic lateral sclerosis                                                                                                   \n",
      " INITIAL SAMPLE SIZE        | 4,857 European ancestry cases, 8,987 European ancestry controls                                                                 \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                              \n",
      " REGION                     | 9p21.2                                                                                                                          \n",
      " CHR_ID                     | 9;9                                                                                                                             \n",
      " CHR_POS                    | 27556782;27543283                                                                                                               \n",
      " REPORTED GENE(S)           | intergenic                                                                                                                      \n",
      " MAPPED_GENE                | C9orf72; C9orf72                                                                                                                \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                            \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                            \n",
      " SNP_GENE_IDS               | null                                                                                                                            \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                            \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                            \n",
      " STRONGEST SNP-RISK ALLELE  | rs10122902-G; rs3849942-A                                                                                                       \n",
      " SNPS                       | rs10122902; rs3849942                                                                                                           \n",
      " MERGED                     | 0                                                                                                                               \n",
      " SNP_ID_CURRENT             | null                                                                                                                            \n",
      " CONTEXT                    | synonymous_variant; non_coding_transcript_exon_variant                                                                          \n",
      " INTERGENIC                 | null                                                                                                                            \n",
      " RISK ALLELE FREQUENCY      | 0.23                                                                                                                            \n",
      " P-VALUE                    | 8E-10                                                                                                                           \n",
      " PVALUE_MLOG                | 9.096910013008056                                                                                                               \n",
      " P-VALUE (TEXT)             | null                                                                                                                            \n",
      " OR or BETA                 | null                                                                                                                            \n",
      " 95% CI (TEXT)              | null                                                                                                                            \n",
      " PLATFORM [SNPS PASSING QC] | Illumina [227475]                                                                                                               \n",
      " CNV                        | N                                                                                                                               \n",
      " MAPPED_TRAIT               | amyotrophic lateral sclerosis                                                                                                   \n",
      " MAPPED_TRAIT_URI           | http://purl.obolibrary.org/obo/MONDO_0004976                                                                                    \n",
      " STUDY ACCESSION            | GCST000820                                                                                                                      \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                    \n",
      "-RECORD 2-----------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2008-06-27                                                                                                                      \n",
      " PUBMEDID                   | 18488028                                                                                                                        \n",
      " FIRST AUTHOR               | Sulem P                                                                                                                         \n",
      " DATE                       | 2008-05-18                                                                                                                      \n",
      " JOURNAL                    | Nat Genet                                                                                                                       \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/18488028                                                                                            \n",
      " STUDY                      | Two newly identified genetic determinants of pigmentation in Europeans.                                                         \n",
      " DISEASE/TRAIT              | Burning and freckling                                                                                                           \n",
      " INITIAL SAMPLE SIZE        | 5,130 Icelandic individuals                                                                                                     \n",
      " REPLICATION SAMPLE SIZE    | 3,330 European ancestry individuals                                                                                             \n",
      " REGION                     | 20q11.22                                                                                                                        \n",
      " CHR_ID                     | 20;20                                                                                                                           \n",
      " CHR_POS                    | 34150806;34141638                                                                                                               \n",
      " REPORTED GENE(S)           | ASIP                                                                                                                            \n",
      " MAPPED_GENE                | RPS2P1 - ASIP; RPS2P1 - ASIP                                                                                                    \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                            \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                            \n",
      " SNP_GENE_IDS               | null                                                                                                                            \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                            \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                            \n",
      " STRONGEST SNP-RISK ALLELE  | rs1015362-G; rs4911414-T                                                                                                        \n",
      " SNPS                       | rs1015362; rs4911414                                                                                                            \n",
      " MERGED                     | 0                                                                                                                               \n",
      " SNP_ID_CURRENT             | null                                                                                                                            \n",
      " CONTEXT                    | intergenic_variant; regulatory_region_variant                                                                                   \n",
      " INTERGENIC                 | null                                                                                                                            \n",
      " RISK ALLELE FREQUENCY      | 0.08                                                                                                                            \n",
      " P-VALUE                    | 6E-37                                                                                                                           \n",
      " PVALUE_MLOG                | 36.221848749616356                                                                                                              \n",
      " P-VALUE (TEXT)             | null                                                                                                                            \n",
      " OR or BETA                 | 2.56                                                                                                                            \n",
      " 95% CI (TEXT)              | [2.06-3.18]                                                                                                                     \n",
      " PLATFORM [SNPS PASSING QC] | Illumina [316515]                                                                                                               \n",
      " CNV                        | N                                                                                                                               \n",
      " MAPPED_TRAIT               | freckles, sunburn                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0003963, http://www.ebi.ac.uk/efo/EFO_0003958                                                      \n",
      " STUDY ACCESSION            | GCST000196                                                                                                                      \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                    \n",
      "only showing top 3 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    gwas_df\n",
    "    .filter(\n",
    "        f.col('CHR_POS').contains(';')\n",
    "    )\n",
    "    .show(3, False, True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b8374676",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-20T06:10:34.883360Z",
     "start_time": "2022-05-20T06:10:34.658508Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-05-19                                                                                                                        \n",
      " PUBMEDID                   | 31926482                                                                                                                          \n",
      " FIRST AUTHOR               | Shen H                                                                                                                            \n",
      " DATE                       | 2020-01-11                                                                                                                        \n",
      " JOURNAL                    | Neuropsychopharmacology                                                                                                           \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/31926482                                                                                              \n",
      " STUDY                      | Polygenic prediction and GWAS of depression, PTSD, and suicidal ideation/self-harm in a Peruvian cohort.                          \n",
      " DISEASE/TRAIT              | Depression                                                                                                                        \n",
      " INITIAL SAMPLE SIZE        | 1,076 Peruvian cases, 2,328 Peruvian controls                                                                                     \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | NR                                                                                                                                \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [rs2162231]                                                                                                                       \n",
      " SNPS                       | [rs2162231-C]                                                                                                                     \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | 2162231-C                                                                                                                         \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | null                                                                                                                              \n",
      " P-VALUE                    | 3E-6                                                                                                                              \n",
      " PVALUE_MLOG                | 5.522878745280337                                                                                                                 \n",
      " P-VALUE (TEXT)             | null                                                                                                                              \n",
      " OR or BETA                 | 1.306                                                                                                                             \n",
      " 95% CI (TEXT)              | [NR]                                                                                                                              \n",
      " PLATFORM [SNPS PASSING QC] | NR [NR]                                                                                                                           \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | unipolar depression                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0003761                                                                                              \n",
      " STUDY ACCESSION            | GCST009978                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 1-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-07-08                                                                                                                        \n",
      " PUBMEDID                   | 32066663                                                                                                                          \n",
      " FIRST AUTHOR               | Niarchou M                                                                                                                        \n",
      " DATE                       | 2020-02-03                                                                                                                        \n",
      " JOURNAL                    | Transl Psychiatry                                                                                                                 \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32066663                                                                                              \n",
      " STUDY                      | Genome-wide association study of dietary intake in the UK biobank study and its associations with schizophrenia and other traits. \n",
      " DISEASE/TRAIT              | Oily fish consumption                                                                                                             \n",
      " INITIAL SAMPLE SIZE        | 335,576 European ancestry individuals                                                                                             \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | MAPT, KANSL1, ARL17B, LRRC37A, LRRC37A2, ARL17A                                                                                   \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [rs5820605-?]                                                                                                                     \n",
      " SNPS                       | [rs5820605]                                                                                                                       \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | 5820605                                                                                                                           \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 5E-9                                                                                                                              \n",
      " PVALUE_MLOG                | 8.301029995663981                                                                                                                 \n",
      " P-VALUE (TEXT)             | null                                                                                                                              \n",
      " OR or BETA                 | 0.0144678                                                                                                                         \n",
      " 95% CI (TEXT)              | [0.0096-0.0193] unit increase                                                                                                     \n",
      " PLATFORM [SNPS PASSING QC] | Affymetrix [25921788] (imputed)                                                                                                   \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | diet measurement                                                                                                                  \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0008111                                                                                              \n",
      " STUDY ACCESSION            | GCST010135                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 2-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-03-10                                                                                                                        \n",
      " PUBMEDID                   | 32060260                                                                                                                          \n",
      " FIRST AUTHOR               | Campos AI                                                                                                                         \n",
      " DATE                       | 2020-02-14                                                                                                                        \n",
      " JOURNAL                    | Nat Commun                                                                                                                        \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32060260                                                                                              \n",
      " STUDY                      | Insights into the aetiology of snoring from observational and genetic investigations in the UK Biobank.                           \n",
      " DISEASE/TRAIT              | Snoring                                                                                                                           \n",
      " INITIAL SAMPLE SIZE        | 151,836 British ancestry cases, 255,230 British ancestry controls                                                                 \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | NR                                                                                                                                \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr1:87773720-CT]                                                                                                                \n",
      " SNPS                       | [chr1:87773720]                                                                                                                   \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 2E-8                                                                                                                              \n",
      " PVALUE_MLOG                | 7.698970004336019                                                                                                                 \n",
      " P-VALUE (TEXT)             | (adjusted for BMI)                                                                                                                \n",
      " OR or BETA                 | 0.0061                                                                                                                            \n",
      " 95% CI (TEXT)              | [0.004-0.0082] unit increase                                                                                                      \n",
      " PLATFORM [SNPS PASSING QC] | NR [11010159] (imputed)                                                                                                           \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | snoring measurement                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0008341                                                                                              \n",
      " STUDY ACCESSION            | GCST009761                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 3-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-03-10                                                                                                                        \n",
      " PUBMEDID                   | 32060260                                                                                                                          \n",
      " FIRST AUTHOR               | Campos AI                                                                                                                         \n",
      " DATE                       | 2020-02-14                                                                                                                        \n",
      " JOURNAL                    | Nat Commun                                                                                                                        \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32060260                                                                                              \n",
      " STUDY                      | Insights into the aetiology of snoring from observational and genetic investigations in the UK Biobank.                           \n",
      " DISEASE/TRAIT              | Snoring                                                                                                                           \n",
      " INITIAL SAMPLE SIZE        | 151,836 British ancestry cases, 255,230 British ancestry controls                                                                 \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | NR                                                                                                                                \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr1:50819256-CA]                                                                                                                \n",
      " SNPS                       | [chr1:50819256]                                                                                                                   \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 6E-10                                                                                                                             \n",
      " PVALUE_MLOG                | 9.221848749616356                                                                                                                 \n",
      " P-VALUE (TEXT)             | (adjusted for BMI)                                                                                                                \n",
      " OR or BETA                 | 0.006465                                                                                                                          \n",
      " 95% CI (TEXT)              | [0.0044-0.0085] unit increase                                                                                                     \n",
      " PLATFORM [SNPS PASSING QC] | NR [11010159] (imputed)                                                                                                           \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | snoring measurement                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0008341                                                                                              \n",
      " STUDY ACCESSION            | GCST009761                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 4-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-03-10                                                                                                                        \n",
      " PUBMEDID                   | 32060260                                                                                                                          \n",
      " FIRST AUTHOR               | Campos AI                                                                                                                         \n",
      " DATE                       | 2020-02-14                                                                                                                        \n",
      " JOURNAL                    | Nat Commun                                                                                                                        \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32060260                                                                                              \n",
      " STUDY                      | Insights into the aetiology of snoring from observational and genetic investigations in the UK Biobank.                           \n",
      " DISEASE/TRAIT              | Snoring                                                                                                                           \n",
      " INITIAL SAMPLE SIZE        | 151,836 British ancestry cases, 255,230 British ancestry controls                                                                 \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | NR                                                                                                                                \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr12:57088077-CA]                                                                                                               \n",
      " SNPS                       | [chr12:57088077]                                                                                                                  \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 3E-10                                                                                                                             \n",
      " PVALUE_MLOG                | 9.522878745280337                                                                                                                 \n",
      " P-VALUE (TEXT)             | (adjusted for BMI)                                                                                                                \n",
      " OR or BETA                 | 0.007143                                                                                                                          \n",
      " 95% CI (TEXT)              | [0.0049-0.0094] unit increase                                                                                                     \n",
      " PLATFORM [SNPS PASSING QC] | NR [11010159] (imputed)                                                                                                           \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | snoring measurement                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0008341                                                                                              \n",
      " STUDY ACCESSION            | GCST009761                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 5-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-03-10                                                                                                                        \n",
      " PUBMEDID                   | 32060260                                                                                                                          \n",
      " FIRST AUTHOR               | Campos AI                                                                                                                         \n",
      " DATE                       | 2020-02-14                                                                                                                        \n",
      " JOURNAL                    | Nat Commun                                                                                                                        \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32060260                                                                                              \n",
      " STUDY                      | Insights into the aetiology of snoring from observational and genetic investigations in the UK Biobank.                           \n",
      " DISEASE/TRAIT              | Snoring                                                                                                                           \n",
      " INITIAL SAMPLE SIZE        | 90,806 British ancestry male cases, 99,165 British ancestry male controls                                                         \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | NR                                                                                                                                \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr1:50819256-CA]                                                                                                                \n",
      " SNPS                       | [chr1:50819256]                                                                                                                   \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 5E-6                                                                                                                              \n",
      " PVALUE_MLOG                | 5.301029995663981                                                                                                                 \n",
      " P-VALUE (TEXT)             | null                                                                                                                              \n",
      " OR or BETA                 | 0.007486                                                                                                                          \n",
      " 95% CI (TEXT)              | [0.0043-0.0107] unit increase                                                                                                     \n",
      " PLATFORM [SNPS PASSING QC] | NR [11010159] (imputed)                                                                                                           \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | snoring measurement                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0008341                                                                                              \n",
      " STUDY ACCESSION            | GCST009762                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 6-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-03-10                                                                                                                        \n",
      " PUBMEDID                   | 32060260                                                                                                                          \n",
      " FIRST AUTHOR               | Campos AI                                                                                                                         \n",
      " DATE                       | 2020-02-14                                                                                                                        \n",
      " JOURNAL                    | Nat Commun                                                                                                                        \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32060260                                                                                              \n",
      " STUDY                      | Insights into the aetiology of snoring from observational and genetic investigations in the UK Biobank.                           \n",
      " DISEASE/TRAIT              | Snoring                                                                                                                           \n",
      " INITIAL SAMPLE SIZE        | 90,806 British ancestry male cases, 99,165 British ancestry male controls                                                         \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | NR                                                                                                                                \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr1:87773720-CT]                                                                                                                \n",
      " SNPS                       | [chr1:87773720]                                                                                                                   \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 8E-6                                                                                                                              \n",
      " PVALUE_MLOG                | 5.096910013008056                                                                                                                 \n",
      " P-VALUE (TEXT)             | null                                                                                                                              \n",
      " OR or BETA                 | 0.007608                                                                                                                          \n",
      " 95% CI (TEXT)              | [0.0043-0.0109] unit increase                                                                                                     \n",
      " PLATFORM [SNPS PASSING QC] | NR [11010159] (imputed)                                                                                                           \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | snoring measurement                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0008341                                                                                              \n",
      " STUDY ACCESSION            | GCST009762                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 7-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-03-10                                                                                                                        \n",
      " PUBMEDID                   | 32060260                                                                                                                          \n",
      " FIRST AUTHOR               | Campos AI                                                                                                                         \n",
      " DATE                       | 2020-02-14                                                                                                                        \n",
      " JOURNAL                    | Nat Commun                                                                                                                        \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32060260                                                                                              \n",
      " STUDY                      | Insights into the aetiology of snoring from observational and genetic investigations in the UK Biobank.                           \n",
      " DISEASE/TRAIT              | Snoring                                                                                                                           \n",
      " INITIAL SAMPLE SIZE        | 61,792 British ancestry female cases, 156,554 British ancestry female controls                                                    \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | NR                                                                                                                                \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr12:57088077-CA]                                                                                                               \n",
      " SNPS                       | [chr12:57088077]                                                                                                                  \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 1E-8                                                                                                                              \n",
      " PVALUE_MLOG                | 8.0                                                                                                                               \n",
      " P-VALUE (TEXT)             | null                                                                                                                              \n",
      " OR or BETA                 | 0.008514                                                                                                                          \n",
      " 95% CI (TEXT)              | [0.0056-0.0114] unit increase                                                                                                     \n",
      " PLATFORM [SNPS PASSING QC] | NR [11010159] (imputed)                                                                                                           \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | snoring measurement                                                                                                               \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0008341                                                                                              \n",
      " STUDY ACCESSION            | GCST009763                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 8-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-07-14                                                                                                                        \n",
      " PUBMEDID                   | 32154731                                                                                                                          \n",
      " FIRST AUTHOR               | Ripatti P                                                                                                                         \n",
      " DATE                       | 2020-03-10                                                                                                                        \n",
      " JOURNAL                    | Circ Genom Precis Med                                                                                                             \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32154731                                                                                              \n",
      " STUDY                      | Polygenic Hyperlipidemias and Coronary Artery Disease Risk.                                                                       \n",
      " DISEASE/TRAIT              | Low density lipoprotein cholesterol levels                                                                                        \n",
      " INITIAL SAMPLE SIZE        | 324,708 British ancestry individuals                                                                                              \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | ABO                                                                                                                               \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr9:136138765-?]                                                                                                                \n",
      " SNPS                       | [chr9:136138765]                                                                                                                  \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 7E-136                                                                                                                            \n",
      " PVALUE_MLOG                | 135.15490195998575                                                                                                                \n",
      " P-VALUE (TEXT)             | null                                                                                                                              \n",
      " OR or BETA                 | 0.0821                                                                                                                            \n",
      " 95% CI (TEXT)              | [0.076-0.088] unit increase                                                                                                       \n",
      " PLATFORM [SNPS PASSING QC] | Affymetrix [5701386] (imputed)                                                                                                    \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | low density lipoprotein cholesterol measurement                                                                                   \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0004611                                                                                              \n",
      " STUDY ACCESSION            | GCST010204                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "-RECORD 9-------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      " DATE ADDED TO CATALOG      | 2020-07-14                                                                                                                        \n",
      " PUBMEDID                   | 32154731                                                                                                                          \n",
      " FIRST AUTHOR               | Ripatti P                                                                                                                         \n",
      " DATE                       | 2020-03-10                                                                                                                        \n",
      " JOURNAL                    | Circ Genom Precis Med                                                                                                             \n",
      " LINK                       | www.ncbi.nlm.nih.gov/pubmed/32154731                                                                                              \n",
      " STUDY                      | Polygenic Hyperlipidemias and Coronary Artery Disease Risk.                                                                       \n",
      " DISEASE/TRAIT              | Low density lipoprotein cholesterol levels                                                                                        \n",
      " INITIAL SAMPLE SIZE        | 324,708 British ancestry individuals                                                                                              \n",
      " REPLICATION SAMPLE SIZE    | NA                                                                                                                                \n",
      " REGION                     | null                                                                                                                              \n",
      " CHR_ID                     | null                                                                                                                              \n",
      " CHR_POS                    | null                                                                                                                              \n",
      " REPORTED GENE(S)           | DOCK7                                                                                                                             \n",
      " MAPPED_GENE                | null                                                                                                                              \n",
      " UPSTREAM_GENE_ID           | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_ID         | null                                                                                                                              \n",
      " SNP_GENE_IDS               | null                                                                                                                              \n",
      " UPSTREAM_GENE_DISTANCE     | null                                                                                                                              \n",
      " DOWNSTREAM_GENE_DISTANCE   | null                                                                                                                              \n",
      " STRONGEST SNP-RISK ALLELE  | [chr1:62944947-?]                                                                                                                 \n",
      " SNPS                       | [chr1:62944947]                                                                                                                   \n",
      " MERGED                     | 0                                                                                                                                 \n",
      " SNP_ID_CURRENT             | null                                                                                                                              \n",
      " CONTEXT                    | null                                                                                                                              \n",
      " INTERGENIC                 | 1                                                                                                                                 \n",
      " RISK ALLELE FREQUENCY      | NR                                                                                                                                \n",
      " P-VALUE                    | 5E-101                                                                                                                            \n",
      " PVALUE_MLOG                | 100.30102999566398                                                                                                                \n",
      " P-VALUE (TEXT)             | null                                                                                                                              \n",
      " OR or BETA                 | 0.0565                                                                                                                            \n",
      " 95% CI (TEXT)              | [0.052-0.061] unit decrease                                                                                                       \n",
      " PLATFORM [SNPS PASSING QC] | Affymetrix [5701386] (imputed)                                                                                                    \n",
      " CNV                        | N                                                                                                                                 \n",
      " MAPPED_TRAIT               | low density lipoprotein cholesterol measurement                                                                                   \n",
      " MAPPED_TRAIT_URI           | http://www.ebi.ac.uk/efo/EFO_0004611                                                                                              \n",
      " STUDY ACCESSION            | GCST010204                                                                                                                        \n",
      " GENOTYPING TECHNOLOGY      | Genome-wide genotyping array                                                                                                      \n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#     # Split rows by multiple variants\n",
    "#     gwas['CHR_ID'] = gwas['CHR_ID'].astype(str).str.split(';')\n",
    "#     gwas['CHR_POS'] = gwas['CHR_POS'].astype(str).str.split(';')\n",
    "#     gwas['SNPS'] = gwas['SNPS'].astype(str).str.split('; ')\n",
    "#     gwas['STRONGEST SNP-RISK ALLELE'] = gwas['STRONGEST SNP-RISK ALLELE'].astype(str).str.split('; ')\n",
    "gwas_split = (\n",
    "    gwas_df\n",
    "    .withColumn('CHR_ID', f.split('CHR_ID', ';'))\n",
    "    .withColumn('CHR_POS', f.split('CHR_POS', ';'))\n",
    "    .withColumn('SNPS', f.split('SNPS', '; '))\n",
    "    .withColumn('STRONGEST SNP-RISK ALLELE', f.split('STRONGEST SNP-RISK ALLELE', '; '))\n",
    ")\n",
    "\n",
    "(\n",
    "    gwas_split\n",
    "    .filter(f.size('CHR_POS') != f.size('STRONGEST SNP-RISK ALLELE'))\n",
    "    .show(10, False, True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "57d476dd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-20T06:13:31.223639Z",
     "start_time": "2022-05-20T06:13:23.912206Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "372752\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load gwas catalog data\n",
    "gwas_raw = pd.read_csv(\n",
    "    'gwas_associations_2022-05-20.tsv',\n",
    "    sep='\\t',\n",
    "    header=0,\n",
    "    dtype={'P-VALUE': str},\n",
    "    low_memory=False,\n",
    "    # na_values=['']\n",
    ")\n",
    "\n",
    "# Set a row (association) id -> Seriously... the table is alrady indexed.\n",
    "gwas_raw['assoc_id'] = list(range(1, gwas_raw.shape[0] + 1))\n",
    "gwas = gwas_raw.copy()\n",
    "\n",
    "# Split rows by multiple variants\n",
    "gwas['CHR_ID'] = gwas['CHR_ID'].astype(str).str.split(';')\n",
    "gwas['CHR_POS'] = gwas['CHR_POS'].astype(str).str.split(';')\n",
    "gwas['SNPS'] = gwas['SNPS'].astype(str).str.split('; ')\n",
    "gwas['STRONGEST SNP-RISK ALLELE'] = gwas['STRONGEST SNP-RISK ALLELE'].astype(str).str.split('; ')\n",
    "\n",
    "print(len(gwas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "97094f6b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T17:37:37.249644Z",
     "start_time": "2022-05-23T17:37:36.902541Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10387"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    gwas_df\n",
    "    .filter(f.lower(f.col('SNPS')).contains('chr'))\n",
    "    .count()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f72b2c69",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T17:40:23.856141Z",
     "start_time": "2022-05-23T17:40:23.761676Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.7865712323475127"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "10387 / gwas_df.count() * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7619cc56",
   "metadata": {},
   "source": [
    "### Ancestries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb2f6539",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T21:09:45.934152Z",
     "start_time": "2022-05-23T21:09:44.366551Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0---------------------------------------------------------------------\n",
      " STUDY ACCESSION                | GCST009498                                  \n",
      " PUBMEDID                       | 31681408                                    \n",
      " FIRST AUTHOR                   | Yang XL                                     \n",
      " DATE                           | 2019-10-11                                  \n",
      " INITIAL SAMPLE DESCRIPTION     | up to 182,902 European ancestry individuals \n",
      " REPLICATION SAMPLE DESCRIPTION | NA                                          \n",
      " STAGE                          | initial                                     \n",
      " NUMBER OF INDIVDUALS           | 182902                                      \n",
      " BROAD ANCESTRAL CATEGORY       | European                                    \n",
      " COUNTRY OF ORIGIN              | NR                                          \n",
      " COUNTRY OF RECRUITMENT         | NR                                          \n",
      " ADDITONAL ANCESTRY DESCRIPTION | null                                        \n",
      "only showing top 1 row\n",
      "\n",
      "None\n",
      "57229\n",
      "42965\n"
     ]
    }
   ],
   "source": [
    "ancestries = spark.read.csv('gwas_ancestries_2022-05-23.tsv', sep='\\t', header=True).persist()\n",
    "print(ancestries.show(1, False, True))\n",
    "print(ancestries.count())\n",
    "print(ancestries.select('STUDY ACCESSION').distinct().count())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0c78c58c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-09T09:18:02.069334Z",
     "start_time": "2022-06-09T09:18:01.741703Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assoc_files</th>\n",
       "      <th>category</th>\n",
       "      <th>category_index</th>\n",
       "      <th>gc_lambda</th>\n",
       "      <th>num_cases</th>\n",
       "      <th>num_cases_prev</th>\n",
       "      <th>num_controls</th>\n",
       "      <th>num_gw_significant</th>\n",
       "      <th>num_gw_significant_prev</th>\n",
       "      <th>phenocode</th>\n",
       "      <th>phenostring</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...</td>\n",
       "      <td>I Certain infectious and parasitic diseases (A...</td>\n",
       "      <td>1</td>\n",
       "      <td>{'0.001': 0.98843, '0.01': 0.98437, '0.1': 0.9...</td>\n",
       "      <td>102</td>\n",
       "      <td>87</td>\n",
       "      <td>234437</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AB1_AMOEBIASIS</td>\n",
       "      <td>Amoebiasis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...</td>\n",
       "      <td>I Certain infectious and parasitic diseases (A...</td>\n",
       "      <td>1</td>\n",
       "      <td>{'0.001': 0.99329, '0.01': 0.99733, '0.1': 0.9...</td>\n",
       "      <td>1101</td>\n",
       "      <td>891</td>\n",
       "      <td>253346</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AB1_ANOGENITAL_HERPES_SIMPLEX</td>\n",
       "      <td>Anogenital herpesviral [herpes simplex] infection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...</td>\n",
       "      <td>I Certain infectious and parasitic diseases (A...</td>\n",
       "      <td>1</td>\n",
       "      <td>{'0.001': 1.1145, '0.01': 1.0274, '0.1': 0.982...</td>\n",
       "      <td>1763</td>\n",
       "      <td>1450</td>\n",
       "      <td>258642</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>AB1_ARTHROPOD</td>\n",
       "      <td>Arthropod-borne viral fevers and viral haemorr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...</td>\n",
       "      <td>I Certain infectious and parasitic diseases (A...</td>\n",
       "      <td>1</td>\n",
       "      <td>{'0.001': 0.99661, '0.01': 1.0038, '0.1': 0.92...</td>\n",
       "      <td>106</td>\n",
       "      <td>87</td>\n",
       "      <td>255615</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AB1_ASPERGILLOSIS</td>\n",
       "      <td>Aspergillosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...</td>\n",
       "      <td>I Certain infectious and parasitic diseases (A...</td>\n",
       "      <td>1</td>\n",
       "      <td>{'0.001': 0.99495, '0.01': 0.99733, '0.1': 0.9...</td>\n",
       "      <td>176</td>\n",
       "      <td>146</td>\n",
       "      <td>259010</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AB1_ATYPICAL_CNS_VIRUS</td>\n",
       "      <td>Atypical virus infections of central nervous s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         assoc_files  \\\n",
       "0  [/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...   \n",
       "1  [/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...   \n",
       "2  [/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...   \n",
       "3  [/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...   \n",
       "4  [/mnt/nfs/pheweb/r6/generated-by-pheweb/pheno_...   \n",
       "\n",
       "                                            category  category_index  \\\n",
       "0  I Certain infectious and parasitic diseases (A...               1   \n",
       "1  I Certain infectious and parasitic diseases (A...               1   \n",
       "2  I Certain infectious and parasitic diseases (A...               1   \n",
       "3  I Certain infectious and parasitic diseases (A...               1   \n",
       "4  I Certain infectious and parasitic diseases (A...               1   \n",
       "\n",
       "                                           gc_lambda  num_cases  \\\n",
       "0  {'0.001': 0.98843, '0.01': 0.98437, '0.1': 0.9...        102   \n",
       "1  {'0.001': 0.99329, '0.01': 0.99733, '0.1': 0.9...       1101   \n",
       "2  {'0.001': 1.1145, '0.01': 1.0274, '0.1': 0.982...       1763   \n",
       "3  {'0.001': 0.99661, '0.01': 1.0038, '0.1': 0.92...        106   \n",
       "4  {'0.001': 0.99495, '0.01': 0.99733, '0.1': 0.9...        176   \n",
       "\n",
       "  num_cases_prev  num_controls  num_gw_significant num_gw_significant_prev  \\\n",
       "0             87        234437                   0                       0   \n",
       "1            891        253346                   0                       0   \n",
       "2           1450        258642                   4                       4   \n",
       "3             87        255615                   0                       0   \n",
       "4            146        259010                   0                       0   \n",
       "\n",
       "                       phenocode  \\\n",
       "0                 AB1_AMOEBIASIS   \n",
       "1  AB1_ANOGENITAL_HERPES_SIMPLEX   \n",
       "2                  AB1_ARTHROPOD   \n",
       "3              AB1_ASPERGILLOSIS   \n",
       "4         AB1_ATYPICAL_CNS_VIRUS   \n",
       "\n",
       "                                         phenostring  \n",
       "0                                         Amoebiasis  \n",
       "1  Anogenital herpesviral [herpes simplex] infection  \n",
       "2  Arthropod-borne viral fevers and viral haemorr...  \n",
       "3                                      Aspergillosis  \n",
       "4  Atypical virus infections of central nervous s...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests \n",
    "\n",
    "\n",
    "df = \n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b3281559",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-09T09:21:36.595992Z",
     "start_time": "2022-06-09T09:21:36.405501Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assoc_files</th>\n",
       "      <th>category</th>\n",
       "      <th>category_index</th>\n",
       "      <th>gc_lambda</th>\n",
       "      <th>num_cases</th>\n",
       "      <th>num_cases_prev</th>\n",
       "      <th>num_controls</th>\n",
       "      <th>num_gw_significant</th>\n",
       "      <th>num_gw_significant_prev</th>\n",
       "      <th>phenocode</th>\n",
       "      <th>phenostring</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [assoc_files, category, category_index, gc_lambda, num_cases, num_cases_prev, num_controls, num_gw_significant, num_gw_significant_prev, phenocode, phenostring]\n",
       "Index: []"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "(\n",
    "    pd.DataFrame(requests.get('https://r7.finngen.fi/api/phenos').json())\n",
    "    .loc[lambda df: df.phenostring.isnull()]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b3633610",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-18T07:50:52.513668Z",
     "start_time": "2022-08-18T07:50:52.175561Z"
    }
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute '_jvm'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-4ce4ac6d3e23>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# from pyspark.sql.functions import row_number\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mwindowSpec\u001b[0m  \u001b[0;34m=\u001b[0m \u001b[0mWindow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartitionBy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"study\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0morderBy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"pval_text\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pyspark/sql/window.py\u001b[0m in \u001b[0;36mpartitionBy\u001b[0;34m(*cols)\u001b[0m\n\u001b[1;32m     70\u001b[0m         \"\"\"\n\u001b[1;32m     71\u001b[0m         \u001b[0msc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSparkContext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_active_spark_context\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 72\u001b[0;31m         \u001b[0mjspec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0morg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexpressions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mWindow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartitionBy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_to_java_cols\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     73\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mWindowSpec\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjspec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute '_jvm'"
     ]
    }
   ],
   "source": [
    "from pyspark.sql.window import Window\n",
    "\n",
    "\n",
    "# from pyspark.sql.functions import row_number\n",
    "windowSpec  = Window.partitionBy(\"study\").orderBy(\"pval_text\")\n",
    "\n",
    "\n",
    "(\n",
    "    spark.createDataFrame([\n",
    "        {'study':'a', 'pval_text': None},\n",
    "        {'study':'b', 'pval_text': 'pval 1'},\n",
    "        {'study':'b', 'pval_text': 'pval 2'},\n",
    "        {'study':'b', 'pval_text': None},\n",
    "    ])\n",
    "    .withColumn(\"row_number\", f.row_number().over(windowSpec) - 1)\n",
    "    .withColumn(\n",
    "        \"new_study\",\n",
    "        f.when(f.col('row_number') == 0, f.col('study')).otherwise(f.concat_ws('_', 'study', 'row_number'))\n",
    "    )\n",
    "    .drop('row_number')\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4a71f0ee",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-09T21:18:04.570737Z",
     "start_time": "2022-06-09T21:18:04.505146Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_new_study_id(df):\n",
    "    ''' Creates a unique study ID from the origin STUDY ACCESSION for each\n",
    "        new trait\n",
    "    Params:\n",
    "        df (pd.df): dataframe with 2 columns: Study accession, trait name\n",
    "    Returns:\n",
    "        list of strs\n",
    "    '''\n",
    "    cache = {}\n",
    "    new_ids = []\n",
    "    for accession, trait in df.values.tolist():\n",
    "\n",
    "        # Add missing accession to cache\n",
    "        if not accession in cache:\n",
    "            cache[accession] = {'counter': 0}\n",
    "\n",
    "        # Add missing trait to cache\n",
    "        if not trait in cache[accession]:\n",
    "            cache[accession]['counter'] += 1\n",
    "            cache[accession][trait] = cache[accession]['counter']\n",
    "\n",
    "        # Create new id. only add suffix if counter > 1\n",
    "        if cache[accession][trait] == 1:\n",
    "            new_id = accession\n",
    "        elif cache[accession][trait] > 1:\n",
    "            new_id = '{0}_{1}'.format(accession, cache[accession][trait])\n",
    "        else:\n",
    "            sys.exit('Error: counter < 1')\n",
    "        new_ids.append(new_id)\n",
    "\n",
    "    return new_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "16f5658e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-09T21:22:27.201938Z",
     "start_time": "2022-06-09T21:22:27.176271Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| study   | pval_text   | new_study   |\n",
      "|:--------|:------------|:------------|\n",
      "| a       |             | a           |\n",
      "| b       | pval 1      | b           |\n",
      "| b       | pval 2      | b_2         |\n",
      "| b       |             | b_3         |\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    pd.DataFrame([\n",
    "        {'study':'a', 'pval_text': None},\n",
    "        {'study':'b', 'pval_text': 'pval 1'},\n",
    "        {'study':'b', 'pval_text': 'pval 2'},\n",
    "        {'study':'b', 'pval_text': None},\n",
    "    ])\n",
    "    .assign(new_study=lambda df: make_new_study_id(df))\n",
    "    .to_markdown(index=False)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d0767794",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-10T09:54:13.326522Z",
     "start_time": "2022-06-10T09:54:13.262779Z"
    }
   },
   "outputs": [],
   "source": [
    "from random import randint, choice\n",
    "from functools import reduce\n",
    "import string\n",
    "\n",
    "def random_char(y):\n",
    "    \"\"\"Generate random column names with a given lenght:\"\"\"\n",
    "    return ''.join(choice(string.ascii_letters) for _ in range(y))\n",
    "\n",
    "def get_dataframe():\n",
    "    \"\"\"Generate spark dataframe with some fixed and a random columns and random values\"\"\"\n",
    "    randomcol = random_char(5)\n",
    "    return spark.createDataFrame([\n",
    "        {\n",
    "            'a': randint(-1000, 1000), \n",
    "            'b':randint(-1000, 1000),\n",
    "            randomcol : randint(-1000, 1000)\n",
    "        } for _ in range(2)\n",
    "    ])\n",
    "\n",
    "# Generate a set of random dataframes:\n",
    "dataframes = [get_dataframe() for _ in range(3)]\n",
    "\n",
    "# Merging dataframes using reduce:\n",
    "merged = reduce(lambda df1, df2: df1.unionByName(df2, allowMissingColumns=True), dataframes)\n",
    "merged.show()\n",
    "\n",
    "# +-----+----+----+-----+-----+\n",
    "# |Nzbke|   a|   b|iFRbE|UtPKj|\n",
    "# +-----+----+----+-----+-----+\n",
    "# | -702| 994| 544| null| null|\n",
    "# |    0| 700| 922| null| null|\n",
    "# | null|-697| 434|  314| null|\n",
    "# | null|-752| 165| -986| null|\n",
    "# | null|-747|-227| null|  133|\n",
    "# | null|-230| 517| null| -768|\n",
    "# +-----+----+----+-----+-----+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4239dc2a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T10:26:23.304749Z",
     "start_time": "2022-06-17T10:26:21.540666Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----+----+----+-----+-----+\n",
      "|Nzbke|   a|   b|iFRbE|UtPKj|\n",
      "+-----+----+----+-----+-----+\n",
      "| -702| 994| 544| null| null|\n",
      "|    0| 700| 922| null| null|\n",
      "| null|-697| 434|  314| null|\n",
      "| null|-752| 165| -986| null|\n",
      "| null|-747|-227| null|  133|\n",
      "| null|-230| 517| null| -768|\n",
      "+-----+----+----+-----+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "merged.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d61a76b0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T10:41:00.772181Z",
     "start_time": "2022-06-17T10:41:00.077720Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------+--------+------------------+\n",
      "|validationLab|otCrispr|targetFromSourceId|\n",
      "+-------------+--------+------------------+\n",
      "|         true|    true|              PPIA|\n",
      "|         true|    true|              HIRA|\n",
      "|         true|    true|             KMT2D|\n",
      "|         true|    true|             KDM2A|\n",
      "|         true|    true|             CCNB1|\n",
      "|         true|    true|            TADA2B|\n",
      "|         true|    true|              RHOA|\n",
      "|         true|    true|             CCDC6|\n",
      "|         true|    true|             MYBL2|\n",
      "|         true|    true|              WDR7|\n",
      "|         true|    true|              USP7|\n",
      "|         true|    true|              FADD|\n",
      "|         true|    true|              PAK2|\n",
      "|         true|    true|             TRAF2|\n",
      "|         true|    true|            PRPF4B|\n",
      "|         true|    true|             PDCD2|\n",
      "|         true|    true|             USP10|\n",
      "|         true|    true|             CCND1|\n",
      "|         true|    true|           HSP90B1|\n",
      "|         true|    true|             CFLAR|\n",
      "|         true|    true|             CASC3|\n",
      "|         true|    true|             UBE2H|\n",
      "|         true|    true|             UBE2C|\n",
      "|         true|    true|              UBR4|\n",
      "|         true|    true|             VPS4B|\n",
      "|         true|    true|              STX4|\n",
      "|         true|    true|              DNM1|\n",
      "|         true|    true|            STAMBP|\n",
      "|         true|    true|              CHD4|\n",
      "|         true|    true|              SHC1|\n",
      "|         true|    true|            PPP2CA|\n",
      "|         true|    true|              YAP1|\n",
      "|         true|    true|             DDX52|\n",
      "|         true|    true|              MTA2|\n",
      "+-------------+--------+------------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vl = (\n",
    "    spark.read.json(\"/Users/dsuveges/repositories/evidence_datasource_parsers/validationLab_2022-05-20.json.gz\")\n",
    "    .select('diseaseFromSourceMappedId', 'targetFromSourceId')\n",
    "    .withColumn('validationLab', f.lit(True))\n",
    "    .distinct()\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "# vl.show()\n",
    "\n",
    "\n",
    "ot = (\n",
    "    spark.read.json('/Users/dsuveges/repositories/evidence_datasource_parsers/ot_crispr_2022-06-13.json.gz')\n",
    "    .select('diseaseFromSourceMappedId', 'targetFromSourceId')\n",
    "    .withColumn('otCrispr', f.lit(True))\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "gene_disease_pairs = (\n",
    "    vl.select('diseaseFromSourceMappedId', 'targetFromSourceId')\n",
    "    .distinct()\n",
    "    .union(\n",
    "        ot.select('diseaseFromSourceMappedId', 'targetFromSourceId').distinct()\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "(\n",
    "    vl.join(ot, on=['targetFromSourceId'], how='inner')\n",
    "    .select('validationLab','otCrispr','targetFromSourceId')\n",
    "    .distinct()\n",
    "    .show(100)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "54cfef4c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T10:46:40.512254Z",
     "start_time": "2022-06-17T10:46:39.430455Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0 rows)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    spark.read.parquet('/Users/dsuveges/project_data/molecule/')\n",
    "    .select('id', 'name', 'tradeNames', \n",
    "        f.col('linkedTargets.rows').alias('linkedTargets'),\n",
    "        f.explode(f.col('linkedDiseases.rows')).alias('diseaseFromSourceMappedId')\n",
    "    )\n",
    "    .select('*', f.explode(f.col('linkedTargets')).alias('targetFromSourceId'))\n",
    "    .join(gene_disease_pairs, on=['diseaseFromSourceMappedId', 'targetFromSourceId'])\n",
    "    .show(1, False, True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "3ba9bad6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T10:36:26.669746Z",
     "start_time": "2022-06-17T10:36:26.534689Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- id: string (nullable = true)\n",
      " |-- canonicalSmiles: string (nullable = true)\n",
      " |-- inchiKey: string (nullable = true)\n",
      " |-- drugType: string (nullable = true)\n",
      " |-- blackBoxWarning: boolean (nullable = true)\n",
      " |-- name: string (nullable = true)\n",
      " |-- yearOfFirstApproval: long (nullable = true)\n",
      " |-- maximumClinicalTrialPhase: long (nullable = true)\n",
      " |-- parentId: string (nullable = true)\n",
      " |-- hasBeenWithdrawn: boolean (nullable = true)\n",
      " |-- isApproved: boolean (nullable = true)\n",
      " |-- withdrawnNotice: struct (nullable = true)\n",
      " |    |-- countries: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- classes: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- year: long (nullable = true)\n",
      " |-- tradeNames: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- synonyms: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- crossReferences: map (nullable = true)\n",
      " |    |-- key: string\n",
      " |    |-- value: array (valueContainsNull = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |-- childChemblIds: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- linkedTargets: struct (nullable = true)\n",
      " |    |-- rows: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- count: integer (nullable = true)\n",
      " |-- linkedDiseases: struct (nullable = true)\n",
      " |    |-- rows: array (nullable = true)\n",
      " |    |    |-- element: string (containsNull = true)\n",
      " |    |-- count: integer (nullable = true)\n",
      " |-- description: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "spark.read.parquet('/Users/dsuveges/project_data/molecule/').printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "c27d10d2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T10:40:32.264398Z",
     "start_time": "2022-06-17T10:40:31.263324Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------------------+------------------+\n",
      "|diseaseFromSourceMappedId|targetFromSourceId|\n",
      "+-------------------------+------------------+\n",
      "|              EFO_1001951|             GATA3|\n",
      "|              EFO_1001951|            STAMBP|\n",
      "|              EFO_1001951|              RHOA|\n",
      "|              EFO_1001951|             FOXA1|\n",
      "|              EFO_1001951|             SPDEF|\n",
      "|              EFO_1001951|            PRPF4B|\n",
      "|              EFO_1001951|             FDFT1|\n",
      "|              EFO_1001951|             TRAF7|\n",
      "|              EFO_1001951|             BRCA2|\n",
      "|              EFO_1001951|           ARHGEF7|\n",
      "|              EFO_1001951|              E2F3|\n",
      "|              EFO_1001951|             KMT2D|\n",
      "|              EFO_1001951|             FOSL1|\n",
      "|              EFO_1001951|              USP7|\n",
      "|              EFO_1001951|              FADD|\n",
      "|              EFO_1001951|            TADA2B|\n",
      "|              EFO_1001951|             CCDC6|\n",
      "|              EFO_1001951|             UBE2C|\n",
      "|              EFO_1001951|             SPAG5|\n",
      "|              EFO_1001951|              SOX2|\n",
      "+-------------------------+------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    vl.select('diseaseFromSourceMappedId', 'targetFromSourceId')\n",
    "    .distinct()\n",
    "    .union(\n",
    "        ot.select('diseaseFromSourceMappedId', 'targetFromSourceId').distinct()\n",
    "    )\n",
    ").show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "bbd75474",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T22:58:22.049282Z",
     "start_time": "2022-06-17T22:58:21.428522Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th> in number of evidence strings</th>\n",
       "      <th> in number of invalid evidence strings</th>\n",
       "      <th> in number of evidence strings dropped due to duplication</th>\n",
       "      <th> in number of evidence strings dropped due to unresolved target</th>\n",
       "      <th> in number of evidence strings dropped due to unresolved disease</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Nr of evidence strings in 22.04.3</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>europepmc</th>\n",
       "      <th>9746463</th>\n",
       "      <td>-170392</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>crispr</th>\n",
       "      <th>1846</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uniprot_variants</th>\n",
       "      <th>39419</th>\n",
       "      <td>-11035</td>\n",
       "      <td>-1414</td>\n",
       "      <td>-327</td>\n",
       "      <td>0</td>\n",
       "      <td>-1108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genomics_england</th>\n",
       "      <th>35622</th>\n",
       "      <td>-4824</td>\n",
       "      <td>-1449</td>\n",
       "      <td>10</td>\n",
       "      <td>-331</td>\n",
       "      <td>-1165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_burden</th>\n",
       "      <th>18715</th>\n",
       "      <td>-166</td>\n",
       "      <td>29</td>\n",
       "      <td>-18</td>\n",
       "      <td>-62</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     in number of evidence strings  \\\n",
       "                 Nr of evidence strings in 22.04.3                                    \n",
       "europepmc        9746463                                                    -170392   \n",
       "crispr           1846                                                             0   \n",
       "uniprot_variants 39419                                                       -11035   \n",
       "genomics_england 35622                                                        -4824   \n",
       "gene_burden      18715                                                         -166   \n",
       "\n",
       "                                                     in number of invalid evidence strings  \\\n",
       "                 Nr of evidence strings in 22.04.3                                            \n",
       "europepmc        9746463                                                                  0   \n",
       "crispr           1846                                                                     0   \n",
       "uniprot_variants 39419                                                                -1414   \n",
       "genomics_england 35622                                                                -1449   \n",
       "gene_burden      18715                                                                   29   \n",
       "\n",
       "                                                     in number of evidence strings dropped due to duplication  \\\n",
       "                 Nr of evidence strings in 22.04.3                                                               \n",
       "europepmc        9746463                                                                            0            \n",
       "crispr           1846                                                                               0            \n",
       "uniprot_variants 39419                                                                           -327            \n",
       "genomics_england 35622                                                                             10            \n",
       "gene_burden      18715                                                                            -18            \n",
       "\n",
       "                                                     in number of evidence strings dropped due to unresolved target  \\\n",
       "                 Nr of evidence strings in 22.04.3                                                                     \n",
       "europepmc        9746463                                                                            0                  \n",
       "crispr           1846                                                                               0                  \n",
       "uniprot_variants 39419                                                                              0                  \n",
       "genomics_england 35622                                                                           -331                  \n",
       "gene_burden      18715                                                                            -62                  \n",
       "\n",
       "                                                     in number of evidence strings dropped due to unresolved disease  \n",
       "                 Nr of evidence strings in 22.04.3                                                                     \n",
       "europepmc        9746463                                                                            0                  \n",
       "crispr           1846                                                                               0                  \n",
       "uniprot_variants 39419                                                                          -1108                  \n",
       "genomics_england 35622                                                                          -1165                  \n",
       "gene_burden      18715                                                                            101                  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_clipboard()\n",
    "df.reshead()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "6e85d532",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T23:00:32.396241Z",
     "start_time": "2022-06-17T23:00:32.367688Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Nr of evidence strings in 22.04.3</th>\n",
       "      <th> in number of evidence strings</th>\n",
       "      <th> in number of invalid evidence strings</th>\n",
       "      <th> in number of evidence strings dropped due to duplication</th>\n",
       "      <th> in number of evidence strings dropped due to unresolved target</th>\n",
       "      <th> in number of evidence strings dropped due to unresolved disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>europepmc</td>\n",
       "      <td>9746463</td>\n",
       "      <td>-170392</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>crispr</td>\n",
       "      <td>1846</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>uniprot_variants</td>\n",
       "      <td>39419</td>\n",
       "      <td>-11035</td>\n",
       "      <td>-1414</td>\n",
       "      <td>-327</td>\n",
       "      <td>0</td>\n",
       "      <td>-1108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>genomics_england</td>\n",
       "      <td>35622</td>\n",
       "      <td>-4824</td>\n",
       "      <td>-1449</td>\n",
       "      <td>10</td>\n",
       "      <td>-331</td>\n",
       "      <td>-1165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_burden</td>\n",
       "      <td>18715</td>\n",
       "      <td>-166</td>\n",
       "      <td>29</td>\n",
       "      <td>-18</td>\n",
       "      <td>-62</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chembl</td>\n",
       "      <td>594375</td>\n",
       "      <td>-2218</td>\n",
       "      <td>-578</td>\n",
       "      <td>-425</td>\n",
       "      <td>0</td>\n",
       "      <td>-173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>slapenrich</td>\n",
       "      <td>72369</td>\n",
       "      <td>53</td>\n",
       "      <td>-53</td>\n",
       "      <td>0</td>\n",
       "      <td>-53</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>expression_atlas</td>\n",
       "      <td>231165</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>intogen</td>\n",
       "      <td>3299</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>reactome</td>\n",
       "      <td>10345</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>orphanet</td>\n",
       "      <td>6001</td>\n",
       "      <td>-2505</td>\n",
       "      <td>-21</td>\n",
       "      <td>-6</td>\n",
       "      <td>0</td>\n",
       "      <td>-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sysbio</td>\n",
       "      <td>389</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cancer_biomarkers</td>\n",
       "      <td>1299</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cancer_gene_census</td>\n",
       "      <td>74159</td>\n",
       "      <td>-186</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>progeny</td>\n",
       "      <td>378</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>clingen</td>\n",
       "      <td>1636</td>\n",
       "      <td>-330</td>\n",
       "      <td>-59</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>eva</td>\n",
       "      <td>1145558</td>\n",
       "      <td>-396345</td>\n",
       "      <td>-101254</td>\n",
       "      <td>-30989</td>\n",
       "      <td>-1087</td>\n",
       "      <td>-69620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>gene2phenotype</td>\n",
       "      <td>2211</td>\n",
       "      <td>-2243</td>\n",
       "      <td>2293</td>\n",
       "      <td>-4</td>\n",
       "      <td>-5</td>\n",
       "      <td>2304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>uniprot_literature</td>\n",
       "      <td>6272</td>\n",
       "      <td>-1455</td>\n",
       "      <td>-251</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>eva_somatic</td>\n",
       "      <td>11653</td>\n",
       "      <td>-254</td>\n",
       "      <td>-159</td>\n",
       "      <td>-12</td>\n",
       "      <td>0</td>\n",
       "      <td>-147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ot_genetics_portal</td>\n",
       "      <td>693567</td>\n",
       "      <td>-647</td>\n",
       "      <td>647</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>impc</td>\n",
       "      <td>1166212</td>\n",
       "      <td>8676</td>\n",
       "      <td>-8676</td>\n",
       "      <td>-6480</td>\n",
       "      <td>0</td>\n",
       "      <td>-2201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Total</td>\n",
       "      <td>13862953</td>\n",
       "      <td>-583870</td>\n",
       "      <td>-110760</td>\n",
       "      <td>-38251</td>\n",
       "      <td>-1541</td>\n",
       "      <td>-71499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Nr of evidence strings in 22.04.3  \\\n",
       "0            europepmc                            9746463   \n",
       "1               crispr                               1846   \n",
       "2     uniprot_variants                              39419   \n",
       "3     genomics_england                              35622   \n",
       "4          gene_burden                              18715   \n",
       "5               chembl                             594375   \n",
       "6           slapenrich                              72369   \n",
       "7     expression_atlas                             231165   \n",
       "8              intogen                               3299   \n",
       "9             reactome                              10345   \n",
       "10            orphanet                               6001   \n",
       "11              sysbio                                389   \n",
       "12   cancer_biomarkers                               1299   \n",
       "13  cancer_gene_census                              74159   \n",
       "14             progeny                                378   \n",
       "15             clingen                               1636   \n",
       "16                 eva                            1145558   \n",
       "17      gene2phenotype                               2211   \n",
       "18  uniprot_literature                               6272   \n",
       "19         eva_somatic                              11653   \n",
       "20  ot_genetics_portal                             693567   \n",
       "21                impc                            1166212   \n",
       "22               Total                           13862953   \n",
       "\n",
       "     in number of evidence strings   in number of invalid evidence strings  \\\n",
       "0                           -170392                                        0   \n",
       "1                                 0                                        0   \n",
       "2                            -11035                                    -1414   \n",
       "3                             -4824                                    -1449   \n",
       "4                              -166                                       29   \n",
       "5                             -2218                                     -578   \n",
       "6                                53                                      -53   \n",
       "7                                 0                                        0   \n",
       "8                                 0                                        0   \n",
       "9                                 0                                        0   \n",
       "10                            -2505                                      -21   \n",
       "11                                0                                        0   \n",
       "12                                1                                       -1   \n",
       "13                             -186                                      186   \n",
       "14                                0                                        0   \n",
       "15                             -330                                      -59   \n",
       "16                          -396345                                  -101254   \n",
       "17                            -2243                                     2293   \n",
       "18                            -1455                                     -251   \n",
       "19                             -254                                     -159   \n",
       "20                             -647                                      647   \n",
       "21                             8676                                    -8676   \n",
       "22                          -583870                                  -110760   \n",
       "\n",
       "     in number of evidence strings dropped due to duplication  \\\n",
       "0                                                   0            \n",
       "1                                                   0            \n",
       "2                                                -327            \n",
       "3                                                  10            \n",
       "4                                                 -18            \n",
       "5                                                -425            \n",
       "6                                                   0            \n",
       "7                                                   0            \n",
       "8                                                   0            \n",
       "9                                                   0            \n",
       "10                                                 -6            \n",
       "11                                                  0            \n",
       "12                                                  0            \n",
       "13                                                  0            \n",
       "14                                                  0            \n",
       "15                                                  0            \n",
       "16                                             -30989            \n",
       "17                                                 -4            \n",
       "18                                                  0            \n",
       "19                                                -12            \n",
       "20                                                  0            \n",
       "21                                              -6480            \n",
       "22                                             -38251            \n",
       "\n",
       "     in number of evidence strings dropped due to unresolved target  \\\n",
       "0                                                   0                  \n",
       "1                                                   0                  \n",
       "2                                                   0                  \n",
       "3                                                -331                  \n",
       "4                                                 -62                  \n",
       "5                                                   0                  \n",
       "6                                                 -53                  \n",
       "7                                                   0                  \n",
       "8                                                   0                  \n",
       "9                                                   0                  \n",
       "10                                                  0                  \n",
       "11                                                  0                  \n",
       "12                                                  0                  \n",
       "13                                                  0                  \n",
       "14                                                  0                  \n",
       "15                                                 -2                  \n",
       "16                                              -1087                  \n",
       "17                                                 -5                  \n",
       "18                                                 -1                  \n",
       "19                                                  0                  \n",
       "20                                                  0                  \n",
       "21                                                  0                  \n",
       "22                                              -1541                  \n",
       "\n",
       "     in number of evidence strings dropped due to unresolved disease  \n",
       "0                                                   0                  \n",
       "1                                                   0                  \n",
       "2                                               -1108                  \n",
       "3                                               -1165                  \n",
       "4                                                 101                  \n",
       "5                                                -173                  \n",
       "6                                                   0                  \n",
       "7                                                   0                  \n",
       "8                                                   0                  \n",
       "9                                                   0                  \n",
       "10                                                -15                  \n",
       "11                                                  0                  \n",
       "12                                                 -1                  \n",
       "13                                                186                  \n",
       "14                                                  0                  \n",
       "15                                                -57                  \n",
       "16                                             -69620                  \n",
       "17                                               2304                  \n",
       "18                                               -250                  \n",
       "19                                               -147                  \n",
       "20                                                647                  \n",
       "21                                              -2201                  \n",
       "22                                             -71499                  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df.reset_index()\n",
    "\n",
    "(\n",
    "    df2\n",
    "    .assign(\n",
    "        \n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "123e40bc",
   "metadata": {},
   "source": [
    "- 22.04: 18.5, 22.06: 23.\n",
    "- Combined set: 23.3k\n",
    "- Not in 22.06: 320\n",
    "- Not in 22.04: 4874\n",
    "- New MONDO terms: 4720"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "d88c2bd0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-20T16:35:07.105519Z",
     "start_time": "2022-06-20T16:35:04.858385Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18520\n",
      "23074\n"
     ]
    }
   ],
   "source": [
    "d4 = spark.read.parquet('/Users/dsuveges/project_data/diseases_22.04/').persist()\n",
    "d6 = spark.read.parquet('/Users/dsuveges/project_data/diseases_22.06/').persist()\n",
    "\n",
    "print(d4.count())\n",
    "print(d6.count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "7b329c4a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-20T16:38:04.650606Z",
     "start_time": "2022-06-20T16:38:04.597525Z"
    }
   },
   "outputs": [],
   "source": [
    "joined = (\n",
    "    d4\n",
    "    .withColumnRenamed('id', 'id_04')\n",
    "    .select('id_04', 'name')\n",
    "    .join(d6.select('id', 'name'), on='name', how='outer')\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "e986bf7d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-20T16:40:48.612969Z",
     "start_time": "2022-06-20T16:40:48.213154Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4720"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    joined\n",
    "    .filter(f.col('id_04').isNull())\n",
    "    .filter(f.col('id').contains('MONDO'))\n",
    "    .count()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "2fc2f7fe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-21T20:03:47.790003Z",
     "start_time": "2022-06-21T20:03:47.528385Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---+----+-----------------------+\n",
      "|c1 |cp  |complex_column         |\n",
      "+---+----+-----------------------+\n",
      "|1  |cica|{4, {null, null, null}}|\n",
      "|2  |cica|{4, {null, null, null}}|\n",
      "|3  |cica|{4, {null, null, null}}|\n",
      "|4  |cica|{4, {null, null, null}}|\n",
      "+---+----+-----------------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "@f.udf(\n",
    "    t.StructType([\n",
    "        t.StructField('value', t.IntegerType()),\n",
    "        t.StructField('array', t.StructType([\n",
    "            t.StructField('a', t.IntegerType()),\n",
    "            t.StructField('b', t.IntegerType()),\n",
    "            t.StructField('c', t.IntegerType()),\n",
    "        ]))\n",
    "    ])\n",
    ")\n",
    "def get_column():\n",
    "    return {'value': 4, 'array': [{'a': 93}, {'b': 58}, {'c': 36}] + {'d': choice([True, False])}}\n",
    "#     return {\n",
    "#         'value': randint(0,199),\n",
    "#         'array': [{x: randint(0,100)} for x in ['a', 'b', 'c']]\n",
    "#     }\n",
    "\n",
    "df = (\n",
    "    spark.createDataFrame([\n",
    "        {'c1': 1}, {'c1': 2}, {'c1': 3}, {'c1': 4}\n",
    "    ])\n",
    "    .withColumn('cp', f.lit('cica'))\n",
    "    .withColumn('complex_column', get_column())\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "df.show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "ba202f2c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-21T20:09:53.311700Z",
     "start_time": "2022-06-21T20:09:53.287541Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- column: map (nullable = true)\n",
      " |    |-- key: string\n",
      " |    |-- value: long (valueContainsNull = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_column():\n",
    "    return {'value': 4, 'array': [{'a': 93}, {'b': 58}, {'c': 36}]}\n",
    "#     return 'pocok'\n",
    "\n",
    "df = pd.DataFrame([{'column': get_column()} for _ in range(4)])\n",
    "spark.createDataFrame(df).printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "481944a6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-21T20:11:39.901867Z",
     "start_time": "2022-06-21T20:11:39.898306Z"
    }
   },
   "outputs": [],
   "source": [
    "from random import choice, randint\n",
    "import pandas as pd\n",
    "import json\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "9234ae12",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-22T06:40:32.174493Z",
     "start_time": "2022-06-22T06:40:31.980127Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- col1: struct (nullable = true)\n",
      " |    |-- array_column: array (nullable = true)\n",
      " |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |-- a: long (nullable = true)\n",
      " |    |    |    |-- b: long (nullable = true)\n",
      " |    |    |    |-- c: long (nullable = true)\n",
      " |    |    |    |-- d: boolean (nullable = true)\n",
      " |    |-- value_column: long (nullable = true)\n",
      " |-- array_lenght: integer (nullable = false)\n",
      " |-- id: long (nullable = false)\n",
      "\n",
      "+--------------------+------------+---+\n",
      "|                col1|array_lenght| id|\n",
      "+--------------------+------------+---+\n",
      "|{[{664, 100, 900,...|          12|  0|\n",
      "|{[{243, 417, 274,...|          12|  1|\n",
      "|         {null, 614}|           0|  2|\n",
      "|{[{631, 478, 423,...|          12|  3|\n",
      "|{[{474, 716, 43, ...|          12|  4|\n",
      "|{[{205, 459, 373,...|          12|  5|\n",
      "|{[{345, 612, 785,...|          12|  6|\n",
      "|{[{244, 792, 969,...|          12|  7|\n",
      "|{[{281, 729, 56, ...|          12|  8|\n",
      "|{[{301, 602, 677,...|          12|  9|\n",
      "|{[{123, 986, 834,...|          12| 10|\n",
      "|{[{763, 821, 428,...|          12| 11|\n",
      "+--------------------+------------+---+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from typing import Dict, List\n",
    "\n",
    "# Preparing mock data:\n",
    "def generate_entry() -> Dict[str, Dict[int, List[int]]]:\n",
    "    data = {'col1': {'value_column': randint(1,1000)}}\n",
    "\n",
    "    # Some times the array is not even there:\n",
    "    if choice([True, True, False]):\n",
    "        data['col1']['array_column'] = [generate_object() for _ in range(12)]\n",
    "\n",
    "    return data\n",
    "\n",
    "def generate_object() -> Dict[str, List[int]]:\n",
    "    d = {x: randint(1,1000) for x in ['a', 'b', 'c']}\n",
    "    d.update({'d': choice([True, False])})\n",
    "    return d\n",
    "\n",
    "# Saving data into a JSON (I couldn't resolve the issue otherwise):\n",
    "with open('test_data2.json', 'wt') as test_file:\n",
    "    for _ in range(12):\n",
    "        json.dump(generate_entry(), test_file)\n",
    "        test_file.write('\\n')\n",
    "\n",
    "# Reading data back as spark dataframe:\n",
    "df = (\n",
    "    spark.read.json('test_data2.json')\n",
    "    # Adding lenght of the array:\n",
    "    .withColumn(\n",
    "        'array_lenght', \n",
    "        f.when(f.col('col1.array_column').isNotNull(), f.size(f.col('col1.array_column')))\n",
    "        .otherwise(f.lit(0))\n",
    "    )\n",
    "    .withColumn('id', f.monotonically_increasing_id())\n",
    ")\n",
    "\n",
    "df.printSchema()\n",
    "# root\n",
    "#  |-- column: struct (nullable = true)\n",
    "#  |    |-- array: array (nullable = true)\n",
    "#  |    |    |-- element: struct (containsNull = true)\n",
    "#  |    |    |    |-- a: long (nullable = true)\n",
    "#  |    |    |    |-- b: long (nullable = true)\n",
    "#  |    |    |    |-- c: long (nullable = true)\n",
    "#  |    |    |    |-- d: boolean (nullable = true)\n",
    "#  |    |-- value: long (nullable = true)\n",
    "\n",
    "\n",
    "# Adding lenght of the array:\n",
    "(\n",
    "    df.show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "5d91c573",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-22T10:23:38.656410Z",
     "start_time": "2022-06-22T10:23:38.250657Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n",
      "+---+------------+--------------------+---------+\n",
      "| id|array_lenght|                col1|new_count|\n",
      "+---+------------+--------------------+---------+\n",
      "|  0|          12|{[{417, 937, 422,...|        6|\n",
      "|  1|          12|{[{700, 722, 785,...|        6|\n",
      "|  2|           0|           {[], 614}|        0|\n",
      "|  3|          12|{[{237, 739, 174,...|        4|\n",
      "|  4|          12|{[{93, 564, 532, ...|        7|\n",
      "|  5|          12|{[{233, 322, 520,...|        9|\n",
      "|  6|          12|{[{481, 647, 434,...|        4|\n",
      "|  7|          12|{[{475, 8, 987, t...|        4|\n",
      "|  8|          12|{[{394, 122, 353,...|        6|\n",
      "|  9|          12|{[{197, 100, 77, ...|        6|\n",
      "| 10|          12|{[{885, 764, 701,...|        3|\n",
      "| 11|          12|{[{855, 297, 11, ...|        4|\n",
      "+---+------------+--------------------+---------+\n",
      "\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(df.count())\n",
    "xf = (\n",
    "    df\n",
    "\n",
    "    # Filtering arrays:\n",
    "    .select(\n",
    "        'id', 'array_lenght', f.col('col1.value_column').alias('value_column'),\n",
    "        f.explode_outer('col1.array_column').alias('exploded')\n",
    "    )\n",
    "    .filter(\n",
    "        (f.col('exploded.d') == True)\n",
    "        | f.col('exploded').isNull()\n",
    "    )\n",
    "    \n",
    "    # Aggregating stuff:\n",
    "    .groupBy('id', 'array_lenght')\n",
    "    .agg(\n",
    "        f.struct(\n",
    "            f.collect_set(f.col('exploded')).alias('array_column'),\n",
    "            f.first(f.col('value_column')).alias('value_column')\n",
    "        ).alias('col1')\n",
    "    )\n",
    "    .withColumn('new_count', f.size(f.col('col1.array_column')))\n",
    "    .orderBy('id')\n",
    ")\n",
    "print(xf.show())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "16e599d2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-22T06:42:21.453846Z",
     "start_time": "2022-06-22T06:42:21.036364Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---+--------------------+------------+----------+\n",
      "| id|                col1|array_lenght|new_length|\n",
      "+---+--------------------+------------+----------+\n",
      "|  0|{372, [{664, 100,...|          12|         6|\n",
      "|  1|{828, [{228, 387,...|          12|         6|\n",
      "|  2|         {614, null}|           0|         0|\n",
      "|  3|{877, [{881, 714,...|          12|         4|\n",
      "|  4|{595, [{474, 716,...|          12|         7|\n",
      "|  5|{554, [{233, 322,...|          12|         9|\n",
      "|  6|{877, [{481, 647,...|          12|         4|\n",
      "|  7|{881, [{475, 8, 9...|          12|         4|\n",
      "|  8|{905, [{281, 729,...|          12|         6|\n",
      "|  9|{142, [{301, 602,...|          12|         6|\n",
      "| 10|{277, [{854, 226,...|          12|         3|\n",
      "| 11|{576, [{350, 635,...|          12|         4|\n",
      "+---+--------------------+------------+----------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Processing array columns:\n",
    "array_cols = (\n",
    "    df.select('id', f.explode(f.col('col1.array_column')).alias('exploded'))\n",
    "    .filter(f.col('exploded.d') == True)\n",
    "    .groupby('id')\n",
    "    .agg(f.collect_list(f.col('exploded')).alias('array_column'))\n",
    ")\n",
    "\n",
    "# Joining with original data:\n",
    "(\n",
    "    df\n",
    "    .join(array_cols, on='id', how='left')\n",
    "    .withColumn('col1', f.struct(f.col('col1.value_column'), f.col('array_column')))\n",
    "    .drop('array_column')\n",
    "    .withColumn(\n",
    "        'new_length', \n",
    "        f.when(f.col('col1.array_column').isNotNull(), f.size(f.col('col1.array_column')))\n",
    "        .otherwise(f.lit(0))\n",
    "    )\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "1c6d3273",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-22T12:52:35.283828Z",
     "start_time": "2022-06-22T12:52:35.177130Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+------------+---+--------------------+-----+\n",
      "|                col1|array_lenght| id|     filtered_arrays|count|\n",
      "+--------------------+------------+---+--------------------+-----+\n",
      "|{[{664, 100, 900,...|          12|  0|[{664, 100, 900, ...|    6|\n",
      "|{[{243, 417, 274,...|          12|  1|[{228, 387, 656, ...|    6|\n",
      "|         {null, 614}|           0|  2|                null|   -1|\n",
      "|{[{631, 478, 423,...|          12|  3|[{881, 714, 234, ...|    4|\n",
      "|{[{474, 716, 43, ...|          12|  4|[{474, 716, 43, t...|    7|\n",
      "|{[{205, 459, 373,...|          12|  5|[{233, 322, 520, ...|    9|\n",
      "|{[{345, 612, 785,...|          12|  6|[{481, 647, 434, ...|    4|\n",
      "|{[{244, 792, 969,...|          12|  7|[{475, 8, 987, tr...|    4|\n",
      "|{[{281, 729, 56, ...|          12|  8|[{281, 729, 56, t...|    6|\n",
      "|{[{301, 602, 677,...|          12|  9|[{301, 602, 677, ...|    6|\n",
      "|{[{123, 986, 834,...|          12| 10|[{854, 226, 406, ...|    3|\n",
      "|{[{763, 821, 428,...|          12| 11|[{350, 635, 274, ...|    4|\n",
      "+--------------------+------------+---+--------------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    df\n",
    "    .select(\"*\", f.expr(\"filter(col1.array_column, array -> array.d == True)\").alias(\"filtered_arrays\"))\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "1d9cc905",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-23T15:17:22.618528Z",
     "start_time": "2022-06-23T15:17:22.301256Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "def explode(df, columns):\n",
    "    ''' Explodes multiple columns\n",
    "    '''\n",
    "    idx = np.repeat(df.index, df[columns[0]].str.len())\n",
    "    a = df.T.reindex(columns).values\n",
    "    concat = np.concatenate([np.concatenate(a[i]) for i in range(a.shape[0])])\n",
    "    p = pd.DataFrame(concat.reshape(a.shape[0], -1).T, idx, columns)\n",
    "    return pd.concat([df.drop(columns, axis=1), p], axis=1).reset_index(drop=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "e56c1338",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-23T15:22:16.523750Z",
     "start_time": "2022-06-23T15:22:16.509613Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>c</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>h</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>d</td>\n",
       "      <td>o</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>i</td>\n",
       "      <td>z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>e</td>\n",
       "      <td>p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>j</td>\n",
       "      <td>q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>f</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>k</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5</td>\n",
       "      <td>g</td>\n",
       "      <td>v</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  b  c\n",
       "0  1  c  l\n",
       "1  1  h  m\n",
       "2  2  d  o\n",
       "3  2  i  z\n",
       "4  3  e  p\n",
       "5  3  j  q\n",
       "6  4  f  s\n",
       "7  4  k  t\n",
       "8  5  g  v"
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = (\n",
    "    pd.DataFrame({\n",
    "        'a': [1,2,3,4,5],\n",
    "        'b': ['c,h','d,i', 'e,j', 'f,k', 'g'],\n",
    "        'c': ['l,m','o,z', 'p,q', 's,t', 'v']\n",
    "    })\n",
    "    .assign(\n",
    "        b = lambda df: df.b.str.split(','),\n",
    "        c = lambda df: df.c.str.split(',')\n",
    "    )\n",
    ")\n",
    "\n",
    "explode(d, ['b', 'c'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "9a3bdf9d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T19:46:23.017693Z",
     "start_time": "2022-06-24T19:46:22.889223Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+-------------+\n",
      "|column_name|column_name_2|\n",
      "+-----------+-------------+\n",
      "|          1|            4|\n",
      "+-----------+-------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    spark.createDataFrame([{\"COLUMN NAME\":1, \"COLUMN NAME 2\": 4}])\n",
    "    .transform(lambda df: df.select(*[f.col(c).alias(c.lower().replace(' ', '_')) for c in df.columns]))\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "d85bc375",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-24T19:51:10.070489Z",
     "start_time": "2022-06-24T19:51:09.900629Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+-------------+\n",
      "|column_name|column_name_2|\n",
      "+-----------+-------------+\n",
      "|          1|            4|\n",
      "+-----------+-------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import quinn\n",
    "(\n",
    "    spark.createDataFrame([{\"COLUMN NAME\":1, \"COLUMN NAME 2\": 4}])\n",
    "    .transform(quinn.snake_case_col_names)\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "44595556",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-26T21:53:54.993159Z",
     "start_time": "2022-06-26T21:53:54.941926Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 1746032\n",
      "drwxr-xr-x  17 dsuveges  384566875   544B 26 Jun 22:52 .\n",
      "drwxrwxr-x  85 dsuveges  384566875   2.7K 15 Jun 14:30 ..\n",
      "drwxr-xr-x   5 dsuveges  384566875   160B 24 May 23:22 .ipynb_checkpoints\n",
      "-rw-r--r--   1 dsuveges  384566875   252K 20 Jun 14:58 New_layout.ipynb\n",
      "-rw-r--r--   1 dsuveges  384566875   190K 26 Jun 22:50 annotate_gwascat_varaintids.ipynb\n",
      "-rw-r--r--   1 dsuveges  384566875   164M 26 Jun 22:43 gwas-catalog-associations.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875    48M 26 Jun 22:51 gwas-catalog-associations.tsv.1\n",
      "-rw-r--r--   1 dsuveges  384566875    11M 27 May 09:49 gwas_ancestries_new_2022-05-27.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875   164M 26 Jun 22:53 gwas_assoc_22_01_26.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875   191M 26 Jun 22:52 gwas_assoc_22_06_16.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875   229M 27 May 09:49 gwas_associations_2022-05-27.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875    19M 27 May 09:49 gwas_studies_new_2022-05-27.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875   741K 27 May 09:49 gwas_unpublished_ancestries_2022-05-27.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875   456K 27 May 09:49 gwas_unpublished_studies_2022-05-27.tsv\n",
      "-rw-r--r--   1 dsuveges  384566875   6.2K 27 May 09:49 make_gwas_cat_study_table.ipynb\n",
      "-rw-r--r--   1 dsuveges  384566875   4.6K 21 Jun 22:05 test_data.json\n",
      "-rw-r--r--   1 dsuveges  384566875   5.1K 22 Jun 07:40 test_data2.json\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "ls -lah"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "d58e0b5f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T05:14:33.590679Z",
     "start_time": "2022-06-27T05:14:32.867585Z"
    }
   },
   "outputs": [],
   "source": [
    "files = {\n",
    "    'gwas_assoc_22_01_26.tsv': 'Jan_2022',\n",
    "    'gwas_assoc_22_06_16.tsv': 'Jun_2022',\n",
    "    'gwas_assoc_21_06_08.tsv': 'Jun_2021'\n",
    "}\n",
    "\n",
    "\n",
    "def read_data(file, date):\n",
    "    return (\n",
    "        spark.read.csv(file, sep='\\t', header=True)\n",
    "        .filter(f.col('SNPS').startswith('rs'))\n",
    "        .withColumn(f'{date}_mapped', f.when(f.col('CHR_ID').isNotNull(), f.lit(True)).otherwise(f.lit(False)))\n",
    "        .select(\n",
    "            f.col('SNPS').alias('rsIds'),\n",
    "            f.col('PUBMEDID').alias('pmid'),\n",
    "            f'{date}_mapped'\n",
    "        )\n",
    "        .distinct()\n",
    "        .persist()\n",
    "    )\n",
    "    \n",
    "dataframes = [read_data(file, date) for file, date in files.items()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "6de99ac4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T05:16:22.875302Z",
     "start_time": "2022-06-27T05:16:21.281234Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+--------------------+---------------+---------------+---------------+\n",
      "|    pmid|               rsIds|Jan_2022_mapped|Jun_2022_mapped|Jun_2021_mapped|\n",
      "+--------+--------------------+---------------+---------------+---------------+\n",
      "|26634245|         rs142412198|           true|           true|           true|\n",
      "|28604732|           rs2713206|           true|           true|           true|\n",
      "|27082954|          rs11012350|           true|           true|           true|\n",
      "|28928442|           rs4360170|           true|           true|           true|\n",
      "|29292387|           rs4955417|           true|           true|           true|\n",
      "|28928442|         rs139260327|           true|           true|           true|\n",
      "|28429243|          rs57712673|           true|           true|           true|\n",
      "|28448500|          rs10101061|           true|           true|           true|\n",
      "|27903959|rs1547399; rs2213...|           true|           true|           true|\n",
      "|29472613|          rs17753687|           true|           true|           true|\n",
      "|29497042|          rs10748799|           true|           true|           true|\n",
      "|29495422|         rs139236944|           true|           true|           true|\n",
      "|27863252|          rs13228847|           true|           true|           true|\n",
      "|29064472|           rs2299682|           true|           true|           true|\n",
      "|26634245|           rs9668369|           true|           true|           true|\n",
      "|28240269|          rs10895987|           true|           true|           true|\n",
      "|28240269|          rs11072996|           true|           true|           true|\n",
      "|29458411|           rs6733839|           true|           true|           true|\n",
      "|29123153|            rs536348|           true|           true|           true|\n",
      "|27863252|           rs9260639|           true|           true|           true|\n",
      "+--------+--------------------+---------------+---------------+---------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from functools import reduce\n",
    "df = reduce(lambda df1, df2: df1.join(df2, on=['pmid', 'rsIds'], how='inner').persist(), dataframes)\n",
    "\n",
    "df.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "76f79357",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T05:30:41.109663Z",
     "start_time": "2022-06-27T05:30:41.043868Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "@f.udf(t.BooleanType())\n",
    "def is_concordant(b1: bool, b2: bool, b3: bool) -> bool:\n",
    "    if b1 == b2 == b3:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "annotated_df = (\n",
    "    df\n",
    "    .withColumn('is_concordant', is_concordant(f.col('Jan_2022_mapped'),\n",
    "                                               f.col('Jun_2022_mapped'),\n",
    "                                                 f.col('Jun_2021_mapped')))\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "b50b7d5f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T06:00:32.649341Z",
     "start_time": "2022-06-27T06:00:32.185054Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+-------------+---------------+---------------+---------------+-------------+\n",
      "|    pmid|        rsIds|Jan_2022_mapped|Jun_2022_mapped|Jun_2021_mapped|is_concordant|\n",
      "+--------+-------------+---------------+---------------+---------------+-------------+\n",
      "|29874175|  rs1042391-T|          false|          false|          false|         true|\n",
      "|29875488|  rs201274146|          false|          false|          false|         true|\n",
      "|25188341|    rs9406351|          false|          false|          false|         true|\n",
      "|29077507|   rs35077018|          false|          false|          false|         true|\n",
      "|25576479| rs10518889-A|          false|          false|          false|         true|\n",
      "|31043758|   rs2428362|          false|          false|          false|         true|\n",
      "|25576479|rs115231908-C|          false|          false|          false|         true|\n",
      "|25576479| rs10445216-C|          false|          false|          false|         true|\n",
      "|31740837|   rs13004237|          false|          false|          false|         true|\n",
      "|28441456|   rs34106305|          false|          false|          false|         true|\n",
      "|25576479|    rs12938-C|          false|          false|          false|         true|\n",
      "|27989323|   rs66707851|          false|          false|          false|         true|\n",
      "|22864933|     rs361433|          false|          false|          false|         true|\n",
      "|22952603|   rs74347767|          false|          false|          false|         true|\n",
      "|29874175|  rs7188697-G|          false|          false|          false|         true|\n",
      "|31361310|  rs152271219|          false|          false|          false|         true|\n",
      "|29874175|  rs4934956-T|          false|          false|          false|         true|\n",
      "|32193507|   rs7708378a|          false|          false|          false|         true|\n",
      "|31251335|  rs11650927|          false|          false|          false|         true|\n",
      "|31201950|    rs9608227|          false|          false|          false|         true|\n",
      "+--------+-------------+---------------+---------------+---------------+-------------+\n",
      "only showing top 20 rows\n",
      "\n",
      "175\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    annotated_df\n",
    "    .filter((f.col('is_concordant') == True) & (f.col('Jan_2022_mapped') == False))\n",
    "    .show()\n",
    ")\n",
    "print(\n",
    "    annotated_df\n",
    "    .filter((f.col('is_concordant') == True) & (f.col('Jan_2022_mapped') == False))\n",
    "    .count()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "39883549",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T05:33:16.543103Z",
     "start_time": "2022-06-27T05:33:16.266085Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+------------+---------------+---------------+---------------+-------------+\n",
      "|    pmid|       rsIds|Jan_2022_mapped|Jun_2022_mapped|Jun_2021_mapped|is_concordant|\n",
      "+--------+------------+---------------+---------------+---------------+-------------+\n",
      "|29874175| rs1042391-T|          false|          false|          false|         true|\n",
      "|29874175| rs7188697-G|          false|          false|          false|         true|\n",
      "|29874175| rs4934956-T|          false|          false|          false|         true|\n",
      "|29874175| rs2074518-A|          false|          false|          false|         true|\n",
      "|29874175| rs1361754-G|          false|          false|          false|         true|\n",
      "|29874175| rs7439869-T|          false|          false|          false|         true|\n",
      "|29874175| rs3189030-A|          false|          false|          false|         true|\n",
      "|29874175|rs55910611-A|          false|          false|          false|         true|\n",
      "|29874175|rs17608766-C|          false|          false|          false|         true|\n",
      "|29874175|rs72648998-T|          false|          false|          false|         true|\n",
      "|29874175| rs1077514-G|          false|          false|          false|         true|\n",
      "|29874175| rs3807989-A|          false|          false|          false|         true|\n",
      "|29874175| rs9912468-G|          false|          false|          false|         true|\n",
      "|29874175|  rs709209-G|          false|          false|          false|         true|\n",
      "|29874175|  rs737154-C|          false|          false|          false|         true|\n",
      "|29874175|rs12053903-C|          false|          false|          false|         true|\n",
      "|29874175| rs2074238-T|          false|          false|          false|         true|\n",
      "|29874175| rs2926743-A|          false|          false|          false|         true|\n",
      "|29874175|rs12143842-T|          false|          false|          false|         true|\n",
      "|29874175| rs1801725-T|          false|          false|          false|         true|\n",
      "+--------+------------+---------------+---------------+---------------+-------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    annotated_df\n",
    "    .filter(f.col('pmid') == '29874175')\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "787fe783",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T05:40:31.979160Z",
     "start_time": "2022-06-27T05:40:31.366275Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+------------+-------------------------+------+-------+\n",
      "|PUBMEDID|        SNPS|STRONGEST SNP-RISK ALLELE|CHR_ID|CHR_POS|\n",
      "+--------+------------+-------------------------+------+-------+\n",
      "|29874175| rs7188697-G|                rs7188697|  null|   null|\n",
      "|29874175| rs2074518-A|                rs2074518|  null|   null|\n",
      "|29874175| rs9912468-G|                rs9912468|  null|   null|\n",
      "|29874175|rs17779747-T|               rs17779747|  null|   null|\n",
      "|29874175| rs6762208-A|                rs6762208|  null|   null|\n",
      "|29874175|  rs737154-C|                 rs737154|  null|   null|\n",
      "|29874175| rs9470361-A|                rs9470361|  null|   null|\n",
      "|29874175| rs2926743-A|                rs2926743|  null|   null|\n",
      "|29874175| rs1361754-G|                rs1361754|  null|   null|\n",
      "|29874175|rs55910611-A|               rs55910611|  null|   null|\n",
      "|29874175| rs1801725-T|                rs1801725|  null|   null|\n",
      "|29874175| rs4934956-T|                rs4934956|  null|   null|\n",
      "|29874175| rs3189030-A|                rs3189030|  null|   null|\n",
      "|29874175|rs17608766-C|               rs17608766|  null|   null|\n",
      "|29874175|  rs709209-G|                 rs709209|  null|   null|\n",
      "|29874175| rs1047640-C|                rs1047640|  null|   null|\n",
      "|29874175|rs72648998-T|               rs72648998|  null|   null|\n",
      "|29874175| rs7439869-T|                rs7439869|  null|   null|\n",
      "|29874175| rs1042391-T|                rs1042391|  null|   null|\n",
      "|29874175| rs1805123-G|                rs1805123|  null|   null|\n",
      "+--------+------------+-------------------------+------+-------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    spark.read.csv(file, sep='\\t', header=True)\n",
    "    .filter(f.col('PUBMEDID') == '29874175')\n",
    "    .select('PUBMEDID', 'SNPS', 'STRONGEST SNP-RISK ALLELE', 'CHR_ID', 'CHR_POS')\n",
    "    .show()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "8e8dabf5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T05:48:24.937009Z",
     "start_time": "2022-06-27T05:48:24.394147Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+-------------+-------------------------+------+-------+\n",
      "|PUBMEDID|         SNPS|STRONGEST SNP-RISK ALLELE|CHR_ID|CHR_POS|\n",
      "+--------+-------------+-------------------------+------+-------+\n",
      "|25576479|rs114539181-G|              rs114539181|  null|   null|\n",
      "|25576479|rs115231908-C|              rs115231908|  null|   null|\n",
      "|25576479| rs10518889-A|               rs10518889|  null|   null|\n",
      "|25576479|rs114747981-C|              rs114747981|  null|   null|\n",
      "|25576479|    rs12938-C|                  rs12938|  null|   null|\n",
      "|25576479|rs116715409-G|              rs116715409|  null|   null|\n",
      "|25576479|  rs3917768-C|                rs3917768|  null|   null|\n",
      "|25576479|rs116121798-A|              rs116121798|  null|   null|\n",
      "|25576479|    rs12938-C|                  rs12938|  null|   null|\n",
      "|25576479| rs11632650-C|               rs11632650|  null|   null|\n",
      "|25576479|   rs228966-A|                 rs228966|  null|   null|\n",
      "|25576479|    rs12938-C|                  rs12938|  null|   null|\n",
      "|25576479|   rs621437-C|                 rs621437|  null|   null|\n",
      "|25576479|rs145194793-A|              rs145194793|  null|   null|\n",
      "|25576479|  rs7195004-T|                rs7195004|  null|   null|\n",
      "|25576479|rs143440180-T|              rs143440180|  null|   null|\n",
      "|25576479|    rs12938-C|                  rs12938|  null|   null|\n",
      "|25576479|   rs621437-C|                 rs621437|  null|   null|\n",
      "|25576479| rs10445216-C|               rs10445216|  null|   null|\n",
      "|25576479| rs11544990-T|               rs11544990|  null|   null|\n",
      "|25576479|  rs1007294-T|                rs1007294|  null|   null|\n",
      "|25576479| rs10048853-T|               rs10048853|  null|   null|\n",
      "|25576479|   rs902790-T|                 rs902790|  null|   null|\n",
      "|29874175|  rs2074518-A|                rs2074518|  null|   null|\n",
      "|29874175|  rs9912468-G|                rs9912468|  null|   null|\n",
      "|29874175| rs17779747-T|               rs17779747|  null|   null|\n",
      "|29874175|  rs6762208-A|                rs6762208|  null|   null|\n",
      "|29874175|  rs7188697-G|                rs7188697|  null|   null|\n",
      "|29874175|  rs9470361-A|                rs9470361|  null|   null|\n",
      "|29874175|  rs2926743-A|                rs2926743|  null|   null|\n",
      "|29874175|  rs1361754-G|                rs1361754|  null|   null|\n",
      "|29874175| rs55910611-A|               rs55910611|  null|   null|\n",
      "|29874175|  rs1801725-T|                rs1801725|  null|   null|\n",
      "|29874175|  rs4934956-T|                rs4934956|  null|   null|\n",
      "|29874175|  rs3189030-A|                rs3189030|  null|   null|\n",
      "|29874175| rs17608766-C|               rs17608766|  null|   null|\n",
      "|29874175|   rs737154-C|                 rs737154|  null|   null|\n",
      "|29874175|  rs1047640-C|                rs1047640|  null|   null|\n",
      "|29874175| rs72648998-T|               rs72648998|  null|   null|\n",
      "|29874175|  rs7439869-T|                rs7439869|  null|   null|\n",
      "|29874175|  rs1042391-T|                rs1042391|  null|   null|\n",
      "|29874175|  rs1805123-G|                rs1805123|  null|   null|\n",
      "|29874175| rs11068997-A|               rs11068997|  null|   null|\n",
      "|29874175|  rs1077514-G|                rs1077514|  null|   null|\n",
      "|29874175| rs12143842-T|               rs12143842|  null|   null|\n",
      "|29874175| rs10919071-G|               rs10919071|  null|   null|\n",
      "|29874175| rs12053903-C|               rs12053903|  null|   null|\n",
      "|29874175|  rs7689609-C|                rs7689609|  null|   null|\n",
      "|29874175| rs11153730-C|               rs11153730|  null|   null|\n",
      "|29874175|  rs3807989-A|                rs3807989|  null|   null|\n",
      "|29874175|  rs2074238-T|                rs2074238|  null|   null|\n",
      "|29874175|     rs1535-G|                   rs1535|  null|   null|\n",
      "|29874175|  rs1886512-A|                rs1886512|  null|   null|\n",
      "|29874175|  rs8049607-T|                rs8049607|  null|   null|\n",
      "|29874175|    rs30208-T|                  rs30208|  null|   null|\n",
      "|29874175|   rs709209-G|                 rs709209|  null|   null|\n",
      "|30175238| rs41273537-G|               rs41273537|  null|   null|\n",
      "|30371334|  rs7769879-?|                rs7769879|  null|   null|\n",
      "|31589552|  rs2307111-C|              rs2307111-C|  null|   null|\n",
      "|31926482|  rs2162231-C|                rs2162231|  null|   null|\n",
      "+--------+-------------+-------------------------+------+-------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    spark.read.csv(file, sep='\\t', header=True)\n",
    "    .filter(f.col('SNPS').contains('-') & f.col('SNPS').startswith('rs'))\n",
    "    .select('PUBMEDID', 'SNPS', 'STRONGEST SNP-RISK ALLELE', 'CHR_ID', 'CHR_POS')\n",
    "    .orderBy('PUBMEDID')\n",
    "    .show(1000)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "id": "d1b6e7e4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T07:12:40.669215Z",
     "start_time": "2022-06-27T07:12:39.903862Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+------------------------+---------------------------+------+-------+\n",
      "|PUBMEDID|SNPS                    |STRONGEST SNP-RISK ALLELE  |CHR_ID|CHR_POS|\n",
      "+--------+------------------------+---------------------------+------+-------+\n",
      "|32193507|\"rs2630445a\"            |rs2630445a-T               |null  |null   |\n",
      "|29874175|\"rs1535-G\"              |rs1535                     |null  |null   |\n",
      "|31251335|\"rs41274896\"           |rs41274896-G             |null  |null   |\n",
      "|29874175|\"rs8049607-T\"           |rs8049607                  |null  |null   |\n",
      "|29874175|\"rs737154-C\"            |rs737154                   |null  |null   |\n",
      "|25576479|\"rs621437-C\"            |rs621437                   |null  |null   |\n",
      "|30573740|\"rs11639303_G_A\"        |rs11639303_G_A-A           |null  |null   |\n",
      "|29874175|\"rs6762208-A\"           |rs6762208                  |null  |null   |\n",
      "|31251335|\"rs10179193\"           |rs10179193-C             |null  |null   |\n",
      "|31251335|\"rs857958\"             |rs857958-T               |null  |null   |\n",
      "|31043758|\"rs2428362\"            |rs2428362-T               |null  |null   |\n",
      "|25576479|\"rs115231908-C\"         |rs115231908                |null  |null   |\n",
      "|25576479|\"rs11632650-C\"          |rs11632650                 |null  |null   |\n",
      "|25576479|\"rs114539181-G\"         |rs114539181                |null  |null   |\n",
      "|29874175|\"rs30208-T\"             |rs30208                    |null  |null   |\n",
      "|25576479|\"rs10048853-T\"          |rs10048853                 |null  |null   |\n",
      "|29874175|\"rs72648998-T\"          |rs72648998                 |null  |null   |\n",
      "|31251335|\"rs11650927\"           |rs11650927-A             |null  |null   |\n",
      "|31251335|\"rs1907608\"            |rs1907608-A              |null  |null   |\n",
      "|29874175|\"rs1801725-T\"           |rs1801725                  |null  |null   |\n",
      "|31251335|\"rs10119861\"           |rs10119861-G             |null  |null   |\n",
      "|31251335|\"rs1993919\"            |rs1993919-G              |null  |null   |\n",
      "|29874175|\"rs11153730-C\"          |rs11153730                 |null  |null   |\n",
      "|23830517|\"rs12524487\"           |rs12524487-T               |null  |null   |\n",
      "|25576479|\"rs3917768-C\"           |rs3917768                  |null  |null   |\n",
      "|31589552|\"rs2307111-C\"           |rs2307111-C                |null  |null   |\n",
      "|25576479|\"rs7195004-T\"           |rs7195004                  |null  |null   |\n",
      "|25576479|\"rs228966-A\"            |rs228966                   |null  |null   |\n",
      "|31251335|\"rs1784128\"            |rs1784128-A              |null  |null   |\n",
      "|24165912|\"rs7658266b\"            |rs7658266-T                |null  |null   |\n",
      "|30371334|\"rs7769879-?\"           |rs7769879                  |null  |null   |\n",
      "|30175238|\"rs41273537-G\"          |rs41273537                 |null  |null   |\n",
      "|35512355|\"rs77992257\"           |rs77992257-?              |null  |null   |\n",
      "|29874175|\"rs4934956-T\"           |rs4934956                  |null  |null   |\n",
      "|34503513|\"rs66801830:IND\"        |rs66801830:IND-A           |null  |null   |\n",
      "|29874175|\"rs3189030-A\"           |rs3189030                  |null  |null   |\n",
      "|34503513|\"rs71661463:IND\"        |rs71661463:IND-G           |null  |null   |\n",
      "|25576479|\"rs902790-T\"            |rs902790                   |null  |null   |\n",
      "|32193507|\"rs7708378a\"            |rs7708378a-T               |null  |null   |\n",
      "|31251335|\"rs9967959\"            |rs9967959-T              |null  |null   |\n",
      "|25576479|\"rs143440180-T\"         |rs143440180                |null  |null   |\n",
      "|29874175|\"rs2074518-A\"           |rs2074518                  |null  |null   |\n",
      "|29874175|\"rs9912468-G\"           |rs9912468                  |null  |null   |\n",
      "|31251335|\"rs2346689\"            |rs2346689-A              |null  |null   |\n",
      "|29874175|\"rs1047640-C\"           |rs1047640                  |null  |null   |\n",
      "|29874175|\"rs1805123-G\"           |rs1805123                  |null  |null   |\n",
      "|29874175|\"rs7188697-G\"           |rs7188697                  |null  |null   |\n",
      "|29874175|\"rs12053903-C\"          |rs12053903                 |null  |null   |\n",
      "|25576479|\"rs10445216-C\"          |rs10445216                 |null  |null   |\n",
      "|31926482|\"rs2162231-C\"           |rs2162231                  |null  |null   |\n",
      "|31251335|\"rs73596019\"           |rs73596019-T             |null  |null   |\n",
      "|31251335|\"rs4834946\"            |rs4834946-C              |null  |null   |\n",
      "|29874175|\"rs11068997-A\"          |rs11068997                 |null  |null   |\n",
      "|29874175|\"rs10919071-G\"          |rs10919071                 |null  |null   |\n",
      "|25576479|\"rs11544990-T\"          |rs11544990                 |null  |null   |\n",
      "|29874175|\"rs3807989-A\"           |rs3807989                  |null  |null   |\n",
      "|29874175|\"rs709209-G\"            |rs709209                   |null  |null   |\n",
      "|29874175|\"rs2926743-A\"           |rs2926743                  |null  |null   |\n",
      "|29874175|\"rs7689609-C\"           |rs7689609                  |null  |null   |\n",
      "|31251335|\"rs75673924\"           |rs75673924-C             |null  |null   |\n",
      "|29874175|\"rs17779747-T\"          |rs17779747                 |null  |null   |\n",
      "|31251335|\"rs9616637\"            |rs9616637-A              |null  |null   |\n",
      "|29874175|\"rs1077514-G\"           |rs1077514                  |null  |null   |\n",
      "|25576479|\"rs116715409-G\"         |rs116715409                |null  |null   |\n",
      "|29892016|\"rs61890184d\"           |rs61890184-A               |null  |null   |\n",
      "|31251335|\"rs1883048\"            |rs1883048-A              |null  |null   |\n",
      "|29874175|\"rs7439869-T\"           |rs7439869                  |null  |null   |\n",
      "|29874175|\"rs2074238-T\"           |rs2074238                  |null  |null   |\n",
      "|32193507|\"rs9506725a\"            |rs9506725a-T               |null  |null   |\n",
      "|33723403|\"rs80103986c\"           |rs80103986c-T              |null  |null   |\n",
      "|25576479|\"rs145194793-A\"         |rs145194793                |null  |null   |\n",
      "|29874175|\"rs17608766-C\"          |rs17608766                 |null  |null   |\n",
      "|29874175|\"rs1886512-A\"           |rs1886512                  |null  |null   |\n",
      "|32193507|\"rs11204213a\"           |rs11204213a-T              |null  |null   |\n",
      "|25576479|\"rs10518889-A\"          |rs10518889                 |null  |null   |\n",
      "|25576479|\"rs1007294-T\"           |rs1007294                  |null  |null   |\n",
      "|32193507|\"rs60078183a\"           |rs60078183a-A              |null  |null   |\n",
      "|34075027|\"rs7948848:\"            |rs7948848:-A               |null  |null   |\n",
      "|34503513|\"rs66801830:IND\"        |rs66801830:IND-G           |null  |null   |\n",
      "|34503513|\"rs71218095:IND\"        |rs71218095:IND-T           |null  |null   |\n",
      "|31251335|\"rs7337023\"            |rs7337023-T              |null  |null   |\n",
      "|29874175|\"rs1361754-G\"           |rs1361754                  |null  |null   |\n",
      "|29874175|\"rs55910611-A\"          |rs55910611                 |null  |null   |\n",
      "|25576479|\"rs12938-C\"             |rs12938                    |null  |null   |\n",
      "|25576479|\"rs116121798-A\"         |rs116121798                |null  |null   |\n",
      "|29874175|\"rs12143842-T\"          |rs12143842                 |null  |null   |\n",
      "|32581359|\"rs71198528,rs747888955\"|rs71198528,rs747888955-CCAG|null  |null   |\n",
      "|29874175|\"rs9470361-A\"           |rs9470361                  |null  |null   |\n",
      "|29874175|\"rs1042391-T\"           |rs1042391                  |null  |null   |\n",
      "|25576479|\"rs114747981-C\"         |rs114747981                |null  |null   |\n",
      "+--------+------------------------+---------------------------+------+-------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "(\n",
    "    spark.read.csv(file, sep='\\t', header=True)\n",
    "    .filter(f.col('SNPS').rlike(r'rs\\d+\\D+'))\n",
    "    .filter(~ f.col('SNPS').rlike(r'[x;]'))\n",
    "    .filter(f.col('CHR_ID').isNull())\n",
    "    .withColumn('SNPS', f.udf(lambda x: f'\"{x}\"', t.StringType())(f.col('SNPS')))\n",
    "    .select('PUBMEDID', 'SNPS', 'STRONGEST SNP-RISK ALLELE', 'CHR_ID', 'CHR_POS')\n",
    "    .orderBy('PUBMEDID')\n",
    "    .distinct()\n",
    "    .show(500, truncate=False)\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "id": "c6e4befc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-27T09:18:01.106611Z",
     "start_time": "2022-06-27T09:18:00.452240Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "99"
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    spark.read.csv(file, sep='\\t', header=True)\n",
    "#     .filter(f.col('MERGED') == 1)\n",
    "    .filter(~f.col('SNP_ID_CURRENT').rlike(r'^\\d+$'))\n",
    "#     .withColumn('cicaful', f.format_string('rs%s', f.col('SNP_ID_CURRENT')))\n",
    "#     .filter(f.col('SNP_ID_CURRENT').isNotNull())\n",
    "#     .filter(f.concat(f.lit('rs') + f.col('SNP_ID_CURRENT')) != f.col('SNPS'))\n",
    "    .select(f.format_string('\"%s\"', f.col('SNP_ID_CURRENT')), 'SNPS')\n",
    "#     .show(100)\n",
    "    .count()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "id": "34ce5c52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-28T09:53:09.064045Z",
     "start_time": "2022-06-28T09:53:09.053925Z"
    }
   },
   "outputs": [],
   "source": [
    "bases = ('ACTG', 'TGAC')\n",
    "tr = {ord(a):ord(b) for a, b in zip(*bases)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "id": "cd04ed5e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-28T09:56:25.211716Z",
     "start_time": "2022-06-28T09:56:25.098441Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+----+\n",
      "|   a|   r|\n",
      "+----+----+\n",
      "|TTGC|AACG|\n",
      "+----+----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "spark.createDataFrame([('TTGC',)], ['a']).select('a', f.translate('a', \"ACTG\", \"TGAC\").alias('r')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdf6e92d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
