{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Uniprot variant review release\n",
    "\n",
    "**Test protein:** \n",
    "* Uniprot id: [Q9BYX4](https://www.uniprot.org/uniprot/Q9BYX4)\n",
    "* Ensembl: [IFIH1 (ENSG00000115267)](https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000115267;r=2:162267074-162318684)\n",
    "\n",
    "**Conclusions:**\n",
    "\n",
    "1. Consequences for only one transcript are captured. Although I might get it wrong: on Ensembl this gene has 7 transcripts, out of which 4 are protein coding. When looking for this gene on Uniprot, I see 3 entries ([Q9BYX4](https://www.uniprot.org/uniprot/Q9BYX4), [A0A3B3IRK8](https://www.uniprot.org/uniprot/A0A3B3IRK8), [A0A7P0Z4A9](https://www.uniprot.org/uniprot/A0A7P0Z4A9)) for each a different transcript is associated, except Q9BYX4, which has two transcripts associated: ENST00000421365 and ENST00000649979. So is it a reasonable assumption to expect consequences calculated for these two transcripts?\n",
    "3. For certain multiallelic variants (eg. `rs778858888`) the consequences are not fully captured. However others (eg. `rs553228814`) are properly represented. Is it due to the fact synonymous variants are not included? (consequences in the files are: `missense`, `frame shift`, `stop gained`) Is there a consequence based filter applied?\n",
    "4. The key `genomicLocation` is not a genomic location per se, it's HGV variant identifier (which by the way has genomic location), but I feel it would make sense to have both this identifier and the genomic location like Ensembl does: \n",
    "\n",
    "insetead:\n",
    "```\n",
    "    <genomicLocation>NC_000002.12:g.162310899C&gt;G</genomicLocation>\n",
    "```\n",
    "rather:\n",
    "```\n",
    "    <genomicLocation seq_region_name=\"2\" start=\"162310899\">\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-23T15:11:11.469047Z",
     "start_time": "2021-07-23T15:11:11.146475Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import requests\n",
    "from itertools import chain\n",
    "import pandas as pd\n",
    "\n",
    "import xml.etree.ElementTree as ET\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-23T15:11:11.721717Z",
     "start_time": "2021-07-23T15:11:11.672439Z"
    }
   },
   "outputs": [],
   "source": [
    "uniprot_file = 'Q9BYX4_mod.xml'\n",
    "\n",
    "tree = ET.parse(uniprot_file)\n",
    "assert isinstance(tree, ET.ElementTree)\n",
    "\n",
    "root = tree.getroot()\n",
    "assert isinstance(root, ET.Element)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-13T22:48:07.649489Z",
     "start_time": "2021-07-13T22:48:07.566863Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location</th>\n",
       "      <th>uniprotId</th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>db_refs</th>\n",
       "      <th>polyPhen_type</th>\n",
       "      <th>polyPhen_score</th>\n",
       "      <th>sift_type</th>\n",
       "      <th>sift_score</th>\n",
       "      <th>HGVS</th>\n",
       "      <th>cytogenicBand</th>\n",
       "      <th>variantLocation</th>\n",
       "      <th>codon</th>\n",
       "      <th>consequence</th>\n",
       "      <th>somaticStatus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs778858888', 'source': 'ExAC'}]</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.015</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.22</td>\n",
       "      <td>NC_000002.12:g.162318299A&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Asn3Lys'}...</td>\n",
       "      <td>AAT/AAA</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'RCV000729683', 'source': 'ClinVar'}, ...</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.4</td>\n",
       "      <td>NC_000002.12:g.162318297C&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Gly4Glu'}...</td>\n",
       "      <td>GGG/GAG</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs1485039163', 'source': 'TOPMed'}]</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.18</td>\n",
       "      <td>NC_000002.12:g.162318294T&gt;C</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Tyr5Cys'}...</td>\n",
       "      <td>TAT/TGT</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs1576243763', 'source': 'Ensembl'}]</td>\n",
       "      <td>probably damaging</td>\n",
       "      <td>0.936</td>\n",
       "      <td>deleterious - low confidence</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NC_000002.12:g.162318291G&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Ser6Tyr'}...</td>\n",
       "      <td>TCC/TAC</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs375179358', 'source': 'ESP'}, {'id'...</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.56</td>\n",
       "      <td>NC_000002.12:g.162318283C&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Glu9Lys'}...</td>\n",
       "      <td>GAG/AAG</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>None</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs777469992', 'source': 'ExAC'}, {'id...</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NC_000002.12:g.162267225C&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Cys1018Ty...</td>\n",
       "      <td>TGC/TAC</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>None</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs755779570', 'source': 'ExAC'}, {'id...</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.228</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.7</td>\n",
       "      <td>NC_000002.12:g.162267218T&gt;G</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Leu1020Ph...</td>\n",
       "      <td>TTA/TTC</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>None</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs1345435216', 'source': 'TOPMed'}]</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.118</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.41</td>\n",
       "      <td>NC_000002.12:g.162267219A&gt;G</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Leu1020Se...</td>\n",
       "      <td>TTA/TCA</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>947</th>\n",
       "      <td>None</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs1385374454', 'source': 'gnomAD'}]</td>\n",
       "      <td>probably damaging</td>\n",
       "      <td>0.999</td>\n",
       "      <td>deleterious - low confidence</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NC_000002.12:g.162267212A&gt;C</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Ser1022Ar...</td>\n",
       "      <td>AGT/AGG</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>None</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs1559808617', 'source': 'Ensembl'}]</td>\n",
       "      <td>probably damaging</td>\n",
       "      <td>0.997</td>\n",
       "      <td>deleterious - low confidence</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NC_000002.12:g.162267211C&gt;G</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Asp1023Hi...</td>\n",
       "      <td>GAT/CAT</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>942 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    location uniprotId geneSymbol  \\\n",
       "0          3    Q9BYX4      IFIH1   \n",
       "1          4    Q9BYX4      IFIH1   \n",
       "2          5    Q9BYX4      IFIH1   \n",
       "3          6    Q9BYX4      IFIH1   \n",
       "4          9    Q9BYX4      IFIH1   \n",
       "..       ...       ...        ...   \n",
       "944     None    Q9BYX4      IFIH1   \n",
       "945     None    Q9BYX4      IFIH1   \n",
       "946     None    Q9BYX4      IFIH1   \n",
       "947     None    Q9BYX4      IFIH1   \n",
       "948     None    Q9BYX4      IFIH1   \n",
       "\n",
       "                                               db_refs      polyPhen_type  \\\n",
       "0            [{'id': 'rs778858888', 'source': 'ExAC'}]             benign   \n",
       "1    [{'id': 'RCV000729683', 'source': 'ClinVar'}, ...             benign   \n",
       "2         [{'id': 'rs1485039163', 'source': 'TOPMed'}]             benign   \n",
       "3        [{'id': 'rs1576243763', 'source': 'Ensembl'}]  probably damaging   \n",
       "4    [{'id': 'rs375179358', 'source': 'ESP'}, {'id'...             benign   \n",
       "..                                                 ...                ...   \n",
       "944  [{'id': 'rs777469992', 'source': 'ExAC'}, {'id...             benign   \n",
       "945  [{'id': 'rs755779570', 'source': 'ExAC'}, {'id...             benign   \n",
       "946       [{'id': 'rs1345435216', 'source': 'TOPMed'}]             benign   \n",
       "947       [{'id': 'rs1385374454', 'source': 'gnomAD'}]  probably damaging   \n",
       "948      [{'id': 'rs1559808617', 'source': 'Ensembl'}]  probably damaging   \n",
       "\n",
       "    polyPhen_score                     sift_type sift_score  \\\n",
       "0            0.015    tolerated - low confidence       0.22   \n",
       "1            0.001    tolerated - low confidence        0.4   \n",
       "2              0.0    tolerated - low confidence       0.18   \n",
       "3            0.936  deleterious - low confidence        0.0   \n",
       "4              0.0                     tolerated       0.56   \n",
       "..             ...                           ...        ...   \n",
       "944            0.0                     tolerated        1.0   \n",
       "945          0.228    tolerated - low confidence        0.7   \n",
       "946          0.118    tolerated - low confidence       0.41   \n",
       "947          0.999  deleterious - low confidence        0.0   \n",
       "948          0.997  deleterious - low confidence        0.0   \n",
       "\n",
       "                            HGVS cytogenicBand  \\\n",
       "0    NC_000002.12:g.162318299A>T        2q24.2   \n",
       "1    NC_000002.12:g.162318297C>T        2q24.2   \n",
       "2    NC_000002.12:g.162318294T>C        2q24.2   \n",
       "3    NC_000002.12:g.162318291G>T        2q24.2   \n",
       "4    NC_000002.12:g.162318283C>T        2q24.2   \n",
       "..                           ...           ...   \n",
       "944  NC_000002.12:g.162267225C>T        2q24.2   \n",
       "945  NC_000002.12:g.162267218T>G        2q24.2   \n",
       "946  NC_000002.12:g.162267219A>G        2q24.2   \n",
       "947  NC_000002.12:g.162267212A>C        2q24.2   \n",
       "948  NC_000002.12:g.162267211C>G        2q24.2   \n",
       "\n",
       "                                       variantLocation    codon consequence  \\\n",
       "0    [{'id': 'ENST00000649979', 'loc': 'p.Asn3Lys'}...  AAT/AAA    missense   \n",
       "1    [{'id': 'ENST00000649979', 'loc': 'p.Gly4Glu'}...  GGG/GAG    missense   \n",
       "2    [{'id': 'ENST00000649979', 'loc': 'p.Tyr5Cys'}...  TAT/TGT    missense   \n",
       "3    [{'id': 'ENST00000649979', 'loc': 'p.Ser6Tyr'}...  TCC/TAC    missense   \n",
       "4    [{'id': 'ENST00000649979', 'loc': 'p.Glu9Lys'}...  GAG/AAG    missense   \n",
       "..                                                 ...      ...         ...   \n",
       "944  [{'id': 'ENST00000649979', 'loc': 'p.Cys1018Ty...  TGC/TAC    missense   \n",
       "945  [{'id': 'ENST00000649979', 'loc': 'p.Leu1020Ph...  TTA/TTC    missense   \n",
       "946  [{'id': 'ENST00000649979', 'loc': 'p.Leu1020Se...  TTA/TCA    missense   \n",
       "947  [{'id': 'ENST00000649979', 'loc': 'p.Ser1022Ar...  AGT/AGG    missense   \n",
       "948  [{'id': 'ENST00000649979', 'loc': 'p.Asp1023Hi...  GAT/CAT    missense   \n",
       "\n",
       "    somaticStatus  \n",
       "0           false  \n",
       "1           false  \n",
       "2           false  \n",
       "3           false  \n",
       "4           false  \n",
       "..            ...  \n",
       "944         false  \n",
       "945         false  \n",
       "946         false  \n",
       "947         false  \n",
       "948         false  \n",
       "\n",
       "[942 rows x 14 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = []\n",
    "\n",
    "# General annotation:\n",
    "uniprot_id = root.find('accession').text\n",
    "hgnc_symbol = root.find('geneName').text\n",
    "\n",
    "for variant in root.findall('feature'):\n",
    "\n",
    "    db_refs = [{'id': x.get('id'), 'source': x.get('type')} for x in variant.findall('dbReference')]\n",
    "    location = variant.find('location').find('position').get('position') if variant.find('location').find('position') is not None else None\n",
    "\n",
    "    # Loop through all variant:\n",
    "    for v in variant.findall('variant'):\n",
    "        var_data = {\n",
    "            'location': location,\n",
    "            'uniprotId': uniprot_id,\n",
    "            'geneSymbol': hgnc_symbol,\n",
    "            'db_refs': db_refs,\n",
    "\n",
    "            # Polyphen annotation:\n",
    "            'polyPhen_type': v.find(\"variantPrediction[@predAlgorithmNameType='PolyPhen']\").get('predictionValType') if v.find(\"variantPrediction[@predAlgorithmNameType='PolyPhen']\") is not None else None,\n",
    "            'polyPhen_score': v.find(\"variantPrediction[@predAlgorithmNameType='PolyPhen']\").get('score') if v.find(\"variantPrediction[@predAlgorithmNameType='PolyPhen']\") is not None else None,\n",
    "\n",
    "            # SIFT annotation:\n",
    "            'sift_type': v.find(\"variantPrediction[@predAlgorithmNameType='SIFT']\").get('predictionValType') if v.find(\"variantPrediction[@predAlgorithmNameType='PolyPhen']\") is not None else None,\n",
    "            'sift_score': v.find(\"variantPrediction[@predAlgorithmNameType='SIFT']\").get('score') if v.find(\"variantPrediction[@predAlgorithmNameType='PolyPhen']\") is not None else None,\n",
    "\n",
    "            # Genomic location:\n",
    "            'HGVS': v.find('genomicLocation').text,\n",
    "            'cytogenicBand': v.find('cytogeneticBand').text,\n",
    "            'variantLocation': [{'id': x.get('seqId'), 'loc': x.get('loc')} for x in v.findall('variantLocation')],\n",
    "            \n",
    "            # Effects:\n",
    "            'codon': v.find('codon').text if v.find('codon')  is not None else None,\n",
    "            'consequence': v.find('consequenceType').text if v.find('consequenceType')  is not None else None,\n",
    "            \n",
    "            # Phenotype annotation:\n",
    "            'somaticStatus': v.find('somaticStatus').text if v.find('somaticStatus')  is not None else None,\n",
    "        }\n",
    "        \n",
    "        \n",
    "        variants.append(var_data)\n",
    "\n",
    "xf = pd.DataFrame(variants)\n",
    "xf.loc[xf.codon.notna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-13T14:52:48.497154Z",
     "start_time": "2021-07-13T14:52:48.479338Z"
    }
   },
   "source": [
    "## Checking overlapping transcript consequences\n",
    "\n",
    "In this test we look for a given variant, if we have consequence terms calculated for one or more overlapping transcripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-13T15:40:39.613440Z",
     "start_time": "2021-07-13T15:40:38.871996Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variant entries for rs778858888: 1\n",
      "          rsId         transcript consequence    codon\n",
      "0  rs778858888  [ENST00000649979]    missense  AAT/AAA\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>consequence_terms</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>codons</th>\n",
       "      <th>alleles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[synonymous_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000421365</td>\n",
       "      <td>aaT/aaC</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000421365</td>\n",
       "      <td>aaT/aaA</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>[synonymous_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000648433</td>\n",
       "      <td>aaT/aaC</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000648433</td>\n",
       "      <td>aaT/aaA</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>[synonymous_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000649979</td>\n",
       "      <td>aaT/aaC</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000649979</td>\n",
       "      <td>aaT/aaA</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>[upstream_gene_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000679938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>[upstream_gene_variant]</td>\n",
       "      <td>ENSG00000115267</td>\n",
       "      <td>ENST00000679938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A/G/T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          consequence_terms          gene_id    transcript_id   codons alleles\n",
       "0      [synonymous_variant]  ENSG00000115267  ENST00000421365  aaT/aaC   A/G/T\n",
       "1        [missense_variant]  ENSG00000115267  ENST00000421365  aaT/aaA   A/G/T\n",
       "8      [synonymous_variant]  ENSG00000115267  ENST00000648433  aaT/aaC   A/G/T\n",
       "9        [missense_variant]  ENSG00000115267  ENST00000648433  aaT/aaA   A/G/T\n",
       "10     [synonymous_variant]  ENSG00000115267  ENST00000649979  aaT/aaC   A/G/T\n",
       "11       [missense_variant]  ENSG00000115267  ENST00000649979  aaT/aaA   A/G/T\n",
       "12  [upstream_gene_variant]  ENSG00000115267  ENST00000679938      NaN   A/G/T\n",
       "13  [upstream_gene_variant]  ENSG00000115267  ENST00000679938      NaN   A/G/T"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_id = 'rs778858888'\n",
    "\n",
    "# Checking parsed xml data:\n",
    "print(f'Number of variant entries for {variant_id}: {len(xf.loc[xf.rsId == variant_id])}')\n",
    "print(xf.loc[xf.rsId == variant_id,['rsId', 'transcript', 'consequence', 'codon']])\n",
    "\n",
    "# Checking consequences in ENSEMBL REST API:\n",
    "url = f'https://rest.ensembl.org/vep/human/id/{variant_id}?content-type=application/json'\n",
    "consequences = requests.get(url)\n",
    "consequences_data = consequences.json()\n",
    "consequences_df = (\n",
    "    pd.DataFrame(consequences_data[0]['transcript_consequences'])\n",
    "    .loc[lambda df: df.gene_symbol == hgnc_symbol, ['consequence_terms', 'gene_id', 'transcript_id', 'codons']]\n",
    "    .assign(alleles = consequences_data[0]['allele_string'])\n",
    ")\n",
    "\n",
    "consequences_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-13T15:21:05.599202Z",
     "start_time": "2021-07-13T15:21:05.558606Z"
    }
   },
   "source": [
    "So, apparently not all transcripts are covered for `rs778858888`:\n",
    "* In the xml file: `ENST00000649979`\n",
    "* From Ensembl REST API: `ENST00000421365`, `ENST00000648433`, `ENST00000649979`, `ENST00000679938`\n",
    "\n",
    "Moreover, `rs778858888` is a multiallelic variant, and only the missense (`AAT/AAA`) is captured, while the synonymous (`aaT/aaC`) is not. How about the other variant with multiallelic nature eg. `rs553228814`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-13T15:40:47.287315Z",
     "start_time": "2021-07-13T15:40:47.278941Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           rsId         transcript consequence    codon  \\\n",
      "17  rs553228814  [ENST00000649979]    missense  AGG/AGT   \n",
      "18  rs553228814  [ENST00000649979]    missense  AGG/AGC   \n",
      "\n",
      "                           HGVS  \n",
      "17  NC_000002.12:g.162318245C>A  \n",
      "18  NC_000002.12:g.162318245C>G  \n"
     ]
    }
   ],
   "source": [
    "print(xf.loc[xf.rsId == 'rs553228814',['rsId', 'transcript', 'consequence', 'codon', 'HGVS']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-13T15:46:56.507345Z",
     "start_time": "2021-07-13T15:46:56.490750Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsId</th>\n",
       "      <th>transcript</th>\n",
       "      <th>consequence</th>\n",
       "      <th>codon</th>\n",
       "      <th>HGVS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>rs1178412309</td>\n",
       "      <td>[ENST00000649979]</td>\n",
       "      <td>missense</td>\n",
       "      <td>GGT/GCT</td>\n",
       "      <td>NC_000002.12:g.162310899C&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>rs1178412309</td>\n",
       "      <td>[ENST00000649979]</td>\n",
       "      <td>missense</td>\n",
       "      <td>GGT/GAT</td>\n",
       "      <td>NC_000002.12:g.162310899C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>rs1178412309</td>\n",
       "      <td>[ENST00000649979]</td>\n",
       "      <td>missense</td>\n",
       "      <td>GGT/GTT</td>\n",
       "      <td>NC_000002.12:g.162310899C&gt;A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             rsId         transcript consequence    codon  \\\n",
       "153  rs1178412309  [ENST00000649979]    missense  GGT/GCT   \n",
       "154  rs1178412309  [ENST00000649979]    missense  GGT/GAT   \n",
       "156  rs1178412309  [ENST00000649979]    missense  GGT/GTT   \n",
       "\n",
       "                            HGVS  \n",
       "153  NC_000002.12:g.162310899C>G  \n",
       "154  NC_000002.12:g.162310899C>T  \n",
       "156  NC_000002.12:g.162310899C>A  "
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiallelics = xf.rsId.value_counts().loc[lambda x: x > 2].index\n",
    "xf.loc[xf.rsId.isin(multiallelics), ['rsId', 'transcript', 'consequence', 'codon', 'HGVS']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-13T22:48:56.921331Z",
     "start_time": "2021-07-13T22:48:56.887161Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location</th>\n",
       "      <th>uniprotId</th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>db_refs</th>\n",
       "      <th>polyPhen_type</th>\n",
       "      <th>polyPhen_score</th>\n",
       "      <th>sift_type</th>\n",
       "      <th>sift_score</th>\n",
       "      <th>HGVS</th>\n",
       "      <th>cytogenicBand</th>\n",
       "      <th>variantLocation</th>\n",
       "      <th>codon</th>\n",
       "      <th>consequence</th>\n",
       "      <th>somaticStatus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs778858888', 'source': 'ExAC'}]</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.015</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.22</td>\n",
       "      <td>NC_000002.12:g.162318299A&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Asn3Lys'}...</td>\n",
       "      <td>AAT/AAA</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'RCV000729683', 'source': 'ClinVar'}, ...</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.4</td>\n",
       "      <td>NC_000002.12:g.162318297C&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Gly4Glu'}...</td>\n",
       "      <td>GGG/GAG</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs1485039163', 'source': 'TOPMed'}]</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tolerated - low confidence</td>\n",
       "      <td>0.18</td>\n",
       "      <td>NC_000002.12:g.162318294T&gt;C</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Tyr5Cys'}...</td>\n",
       "      <td>TAT/TGT</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs1576243763', 'source': 'Ensembl'}]</td>\n",
       "      <td>probably damaging</td>\n",
       "      <td>0.936</td>\n",
       "      <td>deleterious - low confidence</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NC_000002.12:g.162318291G&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Ser6Tyr'}...</td>\n",
       "      <td>TCC/TAC</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'rs375179358', 'source': 'ESP'}, {'id'...</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.56</td>\n",
       "      <td>NC_000002.12:g.162318283C&gt;T</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Glu9Lys'}...</td>\n",
       "      <td>GAG/AAG</td>\n",
       "      <td>missense</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>589</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'RCV000652114', 'source': 'ClinVar'}, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000002.12:g.162278213dup</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Ala589fs'}]</td>\n",
       "      <td>None</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>596</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'RCV000768232', 'source': 'ClinVar'}, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000002.12:g.162277662_162277672del</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Lys596fs'}]</td>\n",
       "      <td>None</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>673</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'RCV000704039', 'source': 'ClinVar'}, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000002.12:g.162277443del</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Asp673fs'}]</td>\n",
       "      <td>None</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>674</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'RCV000419118', 'source': 'ClinVar'}, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000002.12:g.162277436_162277439ATCT[1]</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Arg674fs'}]</td>\n",
       "      <td>None</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>679</td>\n",
       "      <td>Q9BYX4</td>\n",
       "      <td>IFIH1</td>\n",
       "      <td>[{'id': 'RCV000790933', 'source': 'ClinVar'}, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000002.12:g.162277424_162277425del</td>\n",
       "      <td>2q24.2</td>\n",
       "      <td>[{'id': 'ENST00000649979', 'loc': 'p.Leu679fs'}]</td>\n",
       "      <td>None</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    location uniprotId geneSymbol  \\\n",
       "0          3    Q9BYX4      IFIH1   \n",
       "1          4    Q9BYX4      IFIH1   \n",
       "2          5    Q9BYX4      IFIH1   \n",
       "3          6    Q9BYX4      IFIH1   \n",
       "4          9    Q9BYX4      IFIH1   \n",
       "..       ...       ...        ...   \n",
       "536      589    Q9BYX4      IFIH1   \n",
       "548      596    Q9BYX4      IFIH1   \n",
       "630      673    Q9BYX4      IFIH1   \n",
       "632      674    Q9BYX4      IFIH1   \n",
       "639      679    Q9BYX4      IFIH1   \n",
       "\n",
       "                                               db_refs      polyPhen_type  \\\n",
       "0            [{'id': 'rs778858888', 'source': 'ExAC'}]             benign   \n",
       "1    [{'id': 'RCV000729683', 'source': 'ClinVar'}, ...             benign   \n",
       "2         [{'id': 'rs1485039163', 'source': 'TOPMed'}]             benign   \n",
       "3        [{'id': 'rs1576243763', 'source': 'Ensembl'}]  probably damaging   \n",
       "4    [{'id': 'rs375179358', 'source': 'ESP'}, {'id'...             benign   \n",
       "..                                                 ...                ...   \n",
       "536  [{'id': 'RCV000652114', 'source': 'ClinVar'}, ...               None   \n",
       "548  [{'id': 'RCV000768232', 'source': 'ClinVar'}, ...               None   \n",
       "630  [{'id': 'RCV000704039', 'source': 'ClinVar'}, ...               None   \n",
       "632  [{'id': 'RCV000419118', 'source': 'ClinVar'}, ...               None   \n",
       "639  [{'id': 'RCV000790933', 'source': 'ClinVar'}, ...               None   \n",
       "\n",
       "    polyPhen_score                     sift_type sift_score  \\\n",
       "0            0.015    tolerated - low confidence       0.22   \n",
       "1            0.001    tolerated - low confidence        0.4   \n",
       "2              0.0    tolerated - low confidence       0.18   \n",
       "3            0.936  deleterious - low confidence        0.0   \n",
       "4              0.0                     tolerated       0.56   \n",
       "..             ...                           ...        ...   \n",
       "536           None                          None       None   \n",
       "548           None                          None       None   \n",
       "630           None                          None       None   \n",
       "632           None                          None       None   \n",
       "639           None                          None       None   \n",
       "\n",
       "                                          HGVS cytogenicBand  \\\n",
       "0                  NC_000002.12:g.162318299A>T        2q24.2   \n",
       "1                  NC_000002.12:g.162318297C>T        2q24.2   \n",
       "2                  NC_000002.12:g.162318294T>C        2q24.2   \n",
       "3                  NC_000002.12:g.162318291G>T        2q24.2   \n",
       "4                  NC_000002.12:g.162318283C>T        2q24.2   \n",
       "..                                         ...           ...   \n",
       "536                NC_000002.12:g.162278213dup        2q24.2   \n",
       "548      NC_000002.12:g.162277662_162277672del        2q24.2   \n",
       "630                NC_000002.12:g.162277443del        2q24.2   \n",
       "632  NC_000002.12:g.162277436_162277439ATCT[1]        2q24.2   \n",
       "639      NC_000002.12:g.162277424_162277425del        2q24.2   \n",
       "\n",
       "                                       variantLocation    codon consequence  \\\n",
       "0    [{'id': 'ENST00000649979', 'loc': 'p.Asn3Lys'}...  AAT/AAA    missense   \n",
       "1    [{'id': 'ENST00000649979', 'loc': 'p.Gly4Glu'}...  GGG/GAG    missense   \n",
       "2    [{'id': 'ENST00000649979', 'loc': 'p.Tyr5Cys'}...  TAT/TGT    missense   \n",
       "3    [{'id': 'ENST00000649979', 'loc': 'p.Ser6Tyr'}...  TCC/TAC    missense   \n",
       "4    [{'id': 'ENST00000649979', 'loc': 'p.Glu9Lys'}...  GAG/AAG    missense   \n",
       "..                                                 ...      ...         ...   \n",
       "536   [{'id': 'ENST00000649979', 'loc': 'p.Ala589fs'}]     None  frameshift   \n",
       "548   [{'id': 'ENST00000649979', 'loc': 'p.Lys596fs'}]     None  frameshift   \n",
       "630   [{'id': 'ENST00000649979', 'loc': 'p.Asp673fs'}]     None  frameshift   \n",
       "632   [{'id': 'ENST00000649979', 'loc': 'p.Arg674fs'}]     None  frameshift   \n",
       "639   [{'id': 'ENST00000649979', 'loc': 'p.Leu679fs'}]     None  frameshift   \n",
       "\n",
       "    somaticStatus  \n",
       "0           false  \n",
       "1           false  \n",
       "2           false  \n",
       "3           false  \n",
       "4           false  \n",
       "..            ...  \n",
       "536         false  \n",
       "548         false  \n",
       "630         false  \n",
       "632         false  \n",
       "639         false  \n",
       "\n",
       "[124 rows x 14 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xf.loc[xf.location.notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-24T07:24:04.640023Z",
     "start_time": "2021-07-24T07:24:04.188477Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "454045"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assoc.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```xml\n",
    "<xs:schema attributeFormDefault=\"unqualified\" elementFormDefault=\"qualified\" xmlns:xs=\"http://www.w3.org/2001/XMLSchema\">\n",
    "  <xs:element name=\"entryFeature\">\n",
    "    <xs:complexType>\n",
    "      <xs:sequence>\n",
    "        <xs:element type=\"xs:string\" name=\"name\"/>\n",
    "        <xs:element type=\"xs:string\" name=\"accession\"/>\n",
    "        <xs:element type=\"xs:string\" name=\"proteinName\"/>\n",
    "        <xs:element type=\"xs:string\" name=\"geneName\"/>\n",
    "        <xs:element type=\"xs:byte\" name=\"proteinExistence\"/>\n",
    "        <xs:element name=\"sequence\">\n",
    "          <xs:complexType>\n",
    "            <xs:simpleContent>\n",
    "              <xs:extension base=\"xs:string\">\n",
    "                <xs:attribute type=\"xs:long\" name=\"checksum\"/>\n",
    "                <xs:attribute type=\"xs:byte\" name=\"version\"/>\n",
    "              </xs:extension>\n",
    "            </xs:simpleContent>\n",
    "          </xs:complexType>\n",
    "        </xs:element>\n",
    "        <xs:element type=\"xs:short\" name=\"taxid\"/>\n",
    "        <xs:element type=\"xs:string\" name=\"organismName\"/>\n",
    "        <xs:element name=\"feature\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "          <xs:complexType>\n",
    "            <xs:sequence>\n",
    "              <xs:element name=\"location\">\n",
    "                <xs:complexType>\n",
    "                  <xs:sequence>\n",
    "                    <xs:element name=\"position\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:simpleContent>\n",
    "                          <xs:extension base=\"xs:string\">\n",
    "                            <xs:attribute type=\"xs:short\" name=\"position\" use=\"optional\"/>\n",
    "                          </xs:extension>\n",
    "                        </xs:simpleContent>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                    <xs:element name=\"begin\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:simpleContent>\n",
    "                          <xs:extension base=\"xs:string\">\n",
    "                            <xs:attribute type=\"xs:short\" name=\"position\" use=\"optional\"/>\n",
    "                          </xs:extension>\n",
    "                        </xs:simpleContent>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                    <xs:element name=\"end\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:simpleContent>\n",
    "                          <xs:extension base=\"xs:string\">\n",
    "                            <xs:attribute type=\"xs:short\" name=\"position\" use=\"optional\"/>\n",
    "                          </xs:extension>\n",
    "                        </xs:simpleContent>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                  </xs:sequence>\n",
    "                </xs:complexType>\n",
    "              </xs:element>\n",
    "              <xs:element name=\"evidence\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                <xs:complexType>\n",
    "                  <xs:sequence>\n",
    "                    <xs:element name=\"dbReference\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:simpleContent>\n",
    "                          <xs:extension base=\"xs:string\">\n",
    "                            <xs:attribute type=\"xs:int\" name=\"id\" use=\"optional\"/>\n",
    "                            <xs:attribute type=\"xs:string\" name=\"type\" use=\"optional\"/>\n",
    "                          </xs:extension>\n",
    "                        </xs:simpleContent>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                  </xs:sequence>\n",
    "                  <xs:attribute type=\"xs:string\" name=\"code\" use=\"optional\"/>\n",
    "                </xs:complexType>\n",
    "              </xs:element>\n",
    "              <xs:element name=\"dbReference\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                <xs:complexType>\n",
    "                  <xs:simpleContent>\n",
    "                    <xs:extension base=\"xs:string\">\n",
    "                      <xs:attribute type=\"xs:string\" name=\"id\" use=\"optional\"/>\n",
    "                      <xs:attribute type=\"xs:string\" name=\"type\" use=\"optional\"/>\n",
    "                    </xs:extension>\n",
    "                  </xs:simpleContent>\n",
    "                </xs:complexType>\n",
    "              </xs:element>\n",
    "              <xs:element name=\"variant\">\n",
    "                <xs:complexType>\n",
    "                  <xs:sequence>\n",
    "                    <xs:element type=\"xs:string\" name=\"cytogeneticBand\"/>\n",
    "                    <xs:element type=\"xs:string\" name=\"genomicLocation\"/>\n",
    "                    <xs:element name=\"variantLocation\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:simpleContent>\n",
    "                          <xs:extension base=\"xs:string\">\n",
    "                            <xs:attribute type=\"xs:string\" name=\"loc\" use=\"optional\"/>\n",
    "                            <xs:attribute type=\"xs:string\" name=\"seqId\" use=\"optional\"/>\n",
    "                            <xs:attribute type=\"xs:string\" name=\"source\" use=\"optional\"/>\n",
    "                          </xs:extension>\n",
    "                        </xs:simpleContent>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                    <xs:element type=\"xs:string\" name=\"codon\" minOccurs=\"0\"/>\n",
    "                    <xs:element type=\"xs:string\" name=\"consequenceType\"/>\n",
    "                    <xs:element type=\"xs:string\" name=\"wildType\"/>\n",
    "                    <xs:element type=\"xs:string\" name=\"mutatedType\"/>\n",
    "                    <xs:element name=\"populationFrequencies\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:simpleContent>\n",
    "                          <xs:extension base=\"xs:string\">\n",
    "                            <xs:attribute type=\"xs:string\" name=\"populationName\" use=\"optional\"/>\n",
    "                            <xs:attribute type=\"xs:float\" name=\"frequency\" use=\"optional\"/>\n",
    "                            <xs:attribute type=\"xs:string\" name=\"source\" use=\"optional\"/>\n",
    "                          </xs:extension>\n",
    "                        </xs:simpleContent>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                    <xs:element name=\"variantPrediction\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:sequence>\n",
    "                          <xs:element type=\"xs:string\" name=\"source\"/>\n",
    "                        </xs:sequence>\n",
    "                        <xs:attribute type=\"xs:string\" name=\"predictionValType\" use=\"optional\"/>\n",
    "                        <xs:attribute type=\"xs:string\" name=\"predictorType\" use=\"optional\"/>\n",
    "                        <xs:attribute type=\"xs:float\" name=\"score\" use=\"optional\"/>\n",
    "                        <xs:attribute type=\"xs:string\" name=\"predAlgorithmNameType\" use=\"optional\"/>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                    <xs:element type=\"xs:string\" name=\"somaticStatus\"/>\n",
    "                    <xs:element name=\"variantClinicalSignificance\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:sequence>\n",
    "                          <xs:element type=\"xs:string\" name=\"source\" maxOccurs=\"unbounded\" minOccurs=\"0\"/>\n",
    "                        </xs:sequence>\n",
    "                        <xs:attribute type=\"xs:string\" name=\"type\" use=\"optional\"/>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                    <xs:element name=\"association\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:sequence>\n",
    "                          <xs:element type=\"xs:string\" name=\"name\"/>\n",
    "                          <xs:element type=\"xs:string\" name=\"description\"/>\n",
    "                          <xs:element name=\"evidence\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                            <xs:complexType>\n",
    "                              <xs:sequence>\n",
    "                                <xs:element name=\"dbReference\">\n",
    "                                  <xs:complexType>\n",
    "                                    <xs:simpleContent>\n",
    "                                      <xs:extension base=\"xs:string\">\n",
    "                                        <xs:attribute type=\"xs:string\" name=\"id\" use=\"optional\"/>\n",
    "                                        <xs:attribute type=\"xs:string\" name=\"type\" use=\"optional\"/>\n",
    "                                      </xs:extension>\n",
    "                                    </xs:simpleContent>\n",
    "                                  </xs:complexType>\n",
    "                                </xs:element>\n",
    "                              </xs:sequence>\n",
    "                              <xs:attribute type=\"xs:string\" name=\"code\" use=\"optional\"/>\n",
    "                            </xs:complexType>\n",
    "                          </xs:element>\n",
    "                          <xs:element name=\"dbReference\">\n",
    "                            <xs:complexType>\n",
    "                              <xs:simpleContent>\n",
    "                                <xs:extension base=\"xs:string\">\n",
    "                                  <xs:attribute type=\"xs:int\" name=\"id\" use=\"optional\"/>\n",
    "                                  <xs:attribute type=\"xs:string\" name=\"type\" use=\"optional\"/>\n",
    "                                </xs:extension>\n",
    "                              </xs:simpleContent>\n",
    "                            </xs:complexType>\n",
    "                          </xs:element>\n",
    "                        </xs:sequence>\n",
    "                        <xs:attribute type=\"xs:string\" name=\"isDisease\" use=\"optional\"/>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                    <xs:element name=\"description\" maxOccurs=\"unbounded\" minOccurs=\"0\">\n",
    "                      <xs:complexType>\n",
    "                        <xs:sequence>\n",
    "                          <xs:element type=\"xs:string\" name=\"value\"/>\n",
    "                          <xs:element type=\"xs:string\" name=\"source\"/>\n",
    "                        </xs:sequence>\n",
    "                      </xs:complexType>\n",
    "                    </xs:element>\n",
    "                  </xs:sequence>\n",
    "                  <xs:attribute type=\"xs:string\" name=\"sourceType\" use=\"optional\"/>\n",
    "                </xs:complexType>\n",
    "              </xs:element>\n",
    "            </xs:sequence>\n",
    "            <xs:attribute type=\"xs:string\" name=\"type\" use=\"optional\"/>\n",
    "            <xs:attribute type=\"xs:string\" name=\"id\" use=\"optional\"/>\n",
    "          </xs:complexType>\n",
    "        </xs:element>\n",
    "      </xs:sequence>\n",
    "    </xs:complexType>\n",
    "  </xs:element>\n",
    "</xs:schema>\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
