{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter for canonical genes\n",
    "\n",
    "\n",
    "To get a more reasonable view on the gene synonyms and labels, using the gene index labels filtered for genes that localized on canonical chromosomes.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T11:14:40.090696Z",
     "start_time": "2021-03-02T11:14:40.083963Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spark version: 3.0.0\n"
     ]
    }
   ],
   "source": [
    "import pyspark.sql\n",
    "from pyspark.sql.types import *\n",
    "from pyspark.sql.functions import *\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "\n",
    "global spark\n",
    "spark = (pyspark.sql.SparkSession\n",
    "    .builder\n",
    "    .config(\"spark.driver.bindAddress\", \"localhost\")\n",
    "    .getOrCreate()\n",
    ")\n",
    "print(f'Spark version: {spark.version}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T16:30:35.566194Z",
     "start_time": "2021-03-02T16:30:33.880113Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl</th>\n",
       "      <th>alias</th>\n",
       "      <th>alias_type</th>\n",
       "      <th>bioType</th>\n",
       "      <th>chrom</th>\n",
       "      <th>id</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>OTTHUMG00000183507</td>\n",
       "      <td>vega_id</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>P04217</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>NM_130786</td>\n",
       "      <td>refseq_accession</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>P04217</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>5</td>\n",
       "      <td>hgnc_id</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>P04217</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>69417</td>\n",
       "      <td>rgd_id</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>P04217</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>138670</td>\n",
       "      <td>omim_id</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>P04217</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Ensembl               alias        alias_type         bioType  \\\n",
       "0  ENSG00000121410  OTTHUMG00000183507           vega_id  protein_coding   \n",
       "1  ENSG00000121410           NM_130786  refseq_accession  protein_coding   \n",
       "2  ENSG00000121410                   5           hgnc_id  protein_coding   \n",
       "3  ENSG00000121410               69417            rgd_id  protein_coding   \n",
       "4  ENSG00000121410              138670           omim_id  protein_coding   \n",
       "\n",
       "  chrom      id _merge  \n",
       "0    19  P04217   both  \n",
       "1    19  P04217   both  \n",
       "2    19  P04217   both  \n",
       "3    19  P04217   both  \n",
       "4    19  P04217   both  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_index_file = '/Users/dsuveges/project_data/ot/target_index/targets'\n",
    "\n",
    "targets = (\n",
    "    spark.read.parquet(target_index_file)\n",
    "    .select(\n",
    "        col('id').alias('Ensembl'),\n",
    "        col('bioType'),\n",
    "        col('genomicLocation.chromosome').alias('chrom'),\n",
    "        col('proteinAnnotations.id'))\n",
    "    .toPandas()\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "\n",
    "merged = (\n",
    "    pd.read_csv(labels_file, sep='\\t')\n",
    "    .merge(targets, on='Ensembl', how='outer', indicator=True)\n",
    ")\n",
    "\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T16:06:45.017958Z",
     "start_time": "2021-03-02T16:06:44.853378Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes in the gene index: 60616\n",
      "Number of genes labels file: 41873\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of genes in the gene index: {len(targets.Ensembl.unique())}')\n",
    "print(f'Number of genes labels file: {len(merged.loc[merged.alias.notna()].Ensembl.unique())}')\n",
    "print(f'Number of genes in the gene index: {}')\n",
    "# print(f'Number of genes in the gene index: {}')\n",
    "# print(f'Number of genes in the gene index: {}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T16:34:43.635900Z",
     "start_time": "2021-03-02T16:34:43.587591Z"
    }
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'numpy.ndarray' object has no attribute 'apply'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-78-cd1c6fc32b04>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m (\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mmerged\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0;34m.\u001b[0m\u001b[0mbioType\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'\\n{x}\\n{merged.loc[merged.bioType == x]._merge.value_counts()}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'apply'"
     ]
    }
   ],
   "source": [
    "(\n",
    "    merged\n",
    "    .bioType\n",
    "    .unique()\n",
    "    .apply(lambda x: print(f'\\n{x}\\n{merged.loc[merged.bioType == x]._merge.value_counts()}'))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T16:38:02.520712Z",
     "start_time": "2021-03-02T16:38:00.923735Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "protein_coding\n",
      "both          19439\n",
      "right_only      505\n",
      "left_only         0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "lncRNA\n",
      "right_only    11163\n",
      "both           5726\n",
      "left_only         0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "transcribed_unprocessed_pseudogene\n",
      "both          725\n",
      "right_only    213\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "processed_pseudogene\n",
      "both          7187\n",
      "right_only    2973\n",
      "left_only        0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "transcribed_processed_pseudogene\n",
      "both          341\n",
      "right_only    159\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "transcribed_unitary_pseudogene\n",
      "both          120\n",
      "right_only     18\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "unprocessed_pseudogene\n",
      "both          1686\n",
      "right_only     926\n",
      "left_only        0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "unitary_pseudogene\n",
      "both          79\n",
      "right_only    19\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "nan\n",
      "both          0\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "polymorphic_pseudogene\n",
      "both          48\n",
      "right_only     0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "scRNA\n",
      "both          1\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "TEC\n",
      "right_only    1031\n",
      "both            26\n",
      "left_only        0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_C_gene\n",
      "both          14\n",
      "right_only     0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_D_gene\n",
      "both          37\n",
      "right_only     0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_C_pseudogene\n",
      "both          9\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_J_gene\n",
      "both          18\n",
      "right_only     0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_J_pseudogene\n",
      "both          3\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_V_gene\n",
      "both          143\n",
      "right_only      2\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_V_pseudogene\n",
      "both          184\n",
      "right_only      3\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "miRNA\n",
      "both          1854\n",
      "right_only      25\n",
      "left_only        0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "Mt_rRNA\n",
      "both          2\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "Mt_tRNA\n",
      "both          22\n",
      "right_only     0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "misc_RNA\n",
      "right_only    1175\n",
      "both          1036\n",
      "left_only        0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "translated_unprocessed_pseudogene\n",
      "both          1\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "translated_processed_pseudogene\n",
      "both          2\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "ribozyme\n",
      "both          5\n",
      "right_only    3\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "snRNA\n",
      "both          1830\n",
      "right_only      71\n",
      "left_only        0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "rRNA\n",
      "both          44\n",
      "right_only     3\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "rRNA_pseudogene\n",
      "both          496\n",
      "right_only      0\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "snoRNA\n",
      "right_only    501\n",
      "both          442\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "scaRNA\n",
      "right_only    29\n",
      "both          20\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "TR_C_gene\n",
      "both          6\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "TR_J_gene\n",
      "both          79\n",
      "right_only     0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "TR_J_pseudogene\n",
      "both          4\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "TR_V_gene\n",
      "both          106\n",
      "right_only      0\n",
      "left_only       0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "TR_V_pseudogene\n",
      "both          33\n",
      "right_only     0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "TR_D_gene\n",
      "both          4\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "vault_RNA\n",
      "both          1\n",
      "right_only    0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "sRNA\n",
      "right_only    4\n",
      "both          1\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "pseudogene\n",
      "right_only    18\n",
      "both           0\n",
      "left_only      0\n",
      "Name: _merge, dtype: int64\n",
      "\n",
      "IG_pseudogene\n",
      "right_only    1\n",
      "both          0\n",
      "left_only     0\n",
      "Name: _merge, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0     None\n",
       "1     None\n",
       "2     None\n",
       "3     None\n",
       "4     None\n",
       "5     None\n",
       "6     None\n",
       "7     None\n",
       "8     None\n",
       "9     None\n",
       "10    None\n",
       "11    None\n",
       "12    None\n",
       "13    None\n",
       "14    None\n",
       "15    None\n",
       "16    None\n",
       "17    None\n",
       "18    None\n",
       "19    None\n",
       "20    None\n",
       "21    None\n",
       "22    None\n",
       "23    None\n",
       "24    None\n",
       "25    None\n",
       "26    None\n",
       "27    None\n",
       "28    None\n",
       "29    None\n",
       "30    None\n",
       "31    None\n",
       "32    None\n",
       "33    None\n",
       "34    None\n",
       "35    None\n",
       "36    None\n",
       "37    None\n",
       "38    None\n",
       "39    None\n",
       "40    None\n",
       "dtype: object"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    pd.Series(\n",
    "        merged\n",
    "        .bioType\n",
    "        .unique()\n",
    "    )\n",
    "    .apply(lambda x: print(f'\\n{x}\\n{merged.loc[merged.bioType == x,[\"Ensembl\",\"_merge\"]].drop_duplicates()._merge.value_counts()}'))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T16:05:47.719766Z",
     "start_time": "2021-03-02T16:05:45.347178Z"
    }
   },
   "outputs": [],
   "source": [
    "targets_pooled = (\n",
    "    targets\n",
    "    .groupby(['Ensembl'])\n",
    "    .agg(uniprotIds=('id', lambda x: x.tolist()),\n",
    "        chroms=('chrom', lambda x: x.tolist()))\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T20:41:37.481209Z",
     "start_time": "2021-03-02T20:41:37.461595Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl</th>\n",
       "      <th>alias</th>\n",
       "      <th>alias_type</th>\n",
       "      <th>bioType</th>\n",
       "      <th>chrom</th>\n",
       "      <th>id</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13701</th>\n",
       "      <td>ENSG00000279355</td>\n",
       "      <td>OTTHUMG00000152859</td>\n",
       "      <td>vega_id</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13702</th>\n",
       "      <td>ENSG00000279355</td>\n",
       "      <td>NR_024277</td>\n",
       "      <td>refseq_accession</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13703</th>\n",
       "      <td>ENSG00000279355</td>\n",
       "      <td>20988</td>\n",
       "      <td>hgnc_id</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13704</th>\n",
       "      <td>ENSG00000279355</td>\n",
       "      <td>AGPAT4-IT1</td>\n",
       "      <td>symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13705</th>\n",
       "      <td>ENSG00000279355</td>\n",
       "      <td>AGPAT4 intronic transcript 1</td>\n",
       "      <td>name</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749531</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>LOC107987373</td>\n",
       "      <td>symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749532</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>39S ribosomal protein L23, mitochondrial</td>\n",
       "      <td>description</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749533</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>-</td>\n",
       "      <td>symbol_from_nomenclature_authority</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749534</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>-</td>\n",
       "      <td>full_name_from_nomenclature_authority</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749535</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>39S ribosomal protein L23, mitochondrial</td>\n",
       "      <td>other_designations</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1125 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Ensembl                                     alias  \\\n",
       "13701   ENSG00000279355                        OTTHUMG00000152859   \n",
       "13702   ENSG00000279355                                 NR_024277   \n",
       "13703   ENSG00000279355                                     20988   \n",
       "13704   ENSG00000279355                                AGPAT4-IT1   \n",
       "13705   ENSG00000279355              AGPAT4 intronic transcript 1   \n",
       "...                 ...                                       ...   \n",
       "749531  ENSG00000276345                              LOC107987373   \n",
       "749532  ENSG00000276345  39S ribosomal protein L23, mitochondrial   \n",
       "749533  ENSG00000276345                                         -   \n",
       "749534  ENSG00000276345                                         -   \n",
       "749535  ENSG00000276345  39S ribosomal protein L23, mitochondrial   \n",
       "\n",
       "                                   alias_type bioType chrom   id     _merge  \n",
       "13701                                 vega_id     NaN   NaN  NaN  left_only  \n",
       "13702                        refseq_accession     NaN   NaN  NaN  left_only  \n",
       "13703                                 hgnc_id     NaN   NaN  NaN  left_only  \n",
       "13704                                  symbol     NaN   NaN  NaN  left_only  \n",
       "13705                                    name     NaN   NaN  NaN  left_only  \n",
       "...                                       ...     ...   ...  ...        ...  \n",
       "749531                                 symbol     NaN   NaN  NaN  left_only  \n",
       "749532                            description     NaN   NaN  NaN  left_only  \n",
       "749533     symbol_from_nomenclature_authority     NaN   NaN  NaN  left_only  \n",
       "749534  full_name_from_nomenclature_authority     NaN   NaN  NaN  left_only  \n",
       "749535                     other_designations     NaN   NaN  NaN  left_only  \n",
       "\n",
       "[1125 rows x 7 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# targets_pooled.loc[targets_pooled.chroms.apply(lambda x: len(x) > 1)]\n",
    "merged.loc[merged._merge == 'left_only']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T21:02:05.391470Z",
     "start_time": "2021-03-02T21:02:05.182338Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"3\" halign=\"left\">count</th>\n",
       "      <th>recovery</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_merge</th>\n",
       "      <th>left_only</th>\n",
       "      <th>right_only</th>\n",
       "      <th>both</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bioType</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IG_C_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IG_C_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IG_D_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IG_J_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IG_J_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IG_V_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>143.0</td>\n",
       "      <td>0.986207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IG_V_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>0.983957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IG_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mt_rRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mt_tRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TEC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1031.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.024598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TR_C_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TR_D_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TR_J_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TR_J_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TR_V_gene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TR_V_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lncRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>11163.0</td>\n",
       "      <td>5726.0</td>\n",
       "      <td>0.339037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>miRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1854.0</td>\n",
       "      <td>0.986695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>misc_RNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>1036.0</td>\n",
       "      <td>0.468566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>polymorphic_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>processed_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2973.0</td>\n",
       "      <td>7187.0</td>\n",
       "      <td>0.707382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_coding</th>\n",
       "      <td>NaN</td>\n",
       "      <td>505.0</td>\n",
       "      <td>19439.0</td>\n",
       "      <td>0.974679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.936170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rRNA_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>496.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribozyme</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>29.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.408163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1830.0</td>\n",
       "      <td>0.962651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoRNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>501.0</td>\n",
       "      <td>442.0</td>\n",
       "      <td>0.468717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcribed_processed_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>159.0</td>\n",
       "      <td>341.0</td>\n",
       "      <td>0.682000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcribed_unitary_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.869565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcribed_unprocessed_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>213.0</td>\n",
       "      <td>725.0</td>\n",
       "      <td>0.772921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>translated_processed_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>translated_unprocessed_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unitary_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>0.806122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unprocessed_pseudogene</th>\n",
       "      <td>NaN</td>\n",
       "      <td>926.0</td>\n",
       "      <td>1686.0</td>\n",
       "      <td>0.645482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vault_RNA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       count                      recovery\n",
       "_merge                             left_only right_only     both          \n",
       "bioType                                                                   \n",
       "IG_C_gene                                NaN        NaN     14.0       NaN\n",
       "IG_C_pseudogene                          NaN        NaN      9.0       NaN\n",
       "IG_D_gene                                NaN        NaN     37.0       NaN\n",
       "IG_J_gene                                NaN        NaN     18.0       NaN\n",
       "IG_J_pseudogene                          NaN        NaN      3.0       NaN\n",
       "IG_V_gene                                NaN        2.0    143.0  0.986207\n",
       "IG_V_pseudogene                          NaN        3.0    184.0  0.983957\n",
       "IG_pseudogene                            NaN        1.0      NaN       NaN\n",
       "Mt_rRNA                                  NaN        NaN      2.0       NaN\n",
       "Mt_tRNA                                  NaN        NaN     22.0       NaN\n",
       "TEC                                      NaN     1031.0     26.0  0.024598\n",
       "TR_C_gene                                NaN        NaN      6.0       NaN\n",
       "TR_D_gene                                NaN        NaN      4.0       NaN\n",
       "TR_J_gene                                NaN        NaN     79.0       NaN\n",
       "TR_J_pseudogene                          NaN        NaN      4.0       NaN\n",
       "TR_V_gene                                NaN        NaN    106.0       NaN\n",
       "TR_V_pseudogene                          NaN        NaN     33.0       NaN\n",
       "lncRNA                                   NaN    11163.0   5726.0  0.339037\n",
       "miRNA                                    NaN       25.0   1854.0  0.986695\n",
       "misc_RNA                                 NaN     1175.0   1036.0  0.468566\n",
       "polymorphic_pseudogene                   NaN        NaN     48.0       NaN\n",
       "processed_pseudogene                     NaN     2973.0   7187.0  0.707382\n",
       "protein_coding                           NaN      505.0  19439.0  0.974679\n",
       "pseudogene                               NaN       18.0      NaN       NaN\n",
       "rRNA                                     NaN        3.0     44.0  0.936170\n",
       "rRNA_pseudogene                          NaN        NaN    496.0       NaN\n",
       "ribozyme                                 NaN        3.0      5.0  0.625000\n",
       "sRNA                                     NaN        4.0      1.0  0.200000\n",
       "scRNA                                    NaN        NaN      1.0       NaN\n",
       "scaRNA                                   NaN       29.0     20.0  0.408163\n",
       "snRNA                                    NaN       71.0   1830.0  0.962651\n",
       "snoRNA                                   NaN      501.0    442.0  0.468717\n",
       "transcribed_processed_pseudogene         NaN      159.0    341.0  0.682000\n",
       "transcribed_unitary_pseudogene           NaN       18.0    120.0  0.869565\n",
       "transcribed_unprocessed_pseudogene       NaN      213.0    725.0  0.772921\n",
       "translated_processed_pseudogene          NaN        NaN      2.0       NaN\n",
       "translated_unprocessed_pseudogene        NaN        NaN      1.0       NaN\n",
       "unitary_pseudogene                       NaN       19.0     79.0  0.806122\n",
       "unprocessed_pseudogene                   NaN      926.0   1686.0  0.645482\n",
       "vault_RNA                                NaN        NaN      1.0       NaN"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = (\n",
    "    merged[['Ensembl', '_merge', 'bioType']]\n",
    "    .drop_duplicates()\n",
    "    .groupby(['bioType','_merge'])\n",
    "    .agg(count=(\"Ensembl\",\"count\"))\n",
    "    .unstack()\n",
    "    .rename(columns=str)\n",
    "    .assign(recovery = lambda x: x['count']['both'] / (x['count']['both'] + x['count']['right_only']))\n",
    ")\n",
    "\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T21:00:04.719839Z",
     "start_time": "2021-03-02T21:00:04.698814Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>bioType</th>\n",
       "      <th colspan=\"3\" halign=\"left\">count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_merge</th>\n",
       "      <th></th>\n",
       "      <th>left_only</th>\n",
       "      <th>right_only</th>\n",
       "      <th>both</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IG_C_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IG_D_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IG_J_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IG_J_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IG_V_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>184.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IG_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TEC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1031.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TR_C_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TR_J_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TR_J_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TR_V_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TR_V_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11163.0</td>\n",
       "      <td>5726.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1854.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>1036.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>polymorphic_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2973.0</td>\n",
       "      <td>7187.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>505.0</td>\n",
       "      <td>19439.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>496.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ribozyme</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>sRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>scRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>scaRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>snRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1830.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>snoRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>501.0</td>\n",
       "      <td>442.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>159.0</td>\n",
       "      <td>341.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>transcribed_unitary_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>213.0</td>\n",
       "      <td>725.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>translated_processed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>translated_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>unitary_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>926.0</td>\n",
       "      <td>1686.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>vault_RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   bioType     count                    \n",
       "_merge                                     left_only right_only     both\n",
       "0                                IG_C_gene       NaN        NaN     14.0\n",
       "1                          IG_C_pseudogene       NaN        NaN      9.0\n",
       "2                                IG_D_gene       NaN        NaN     37.0\n",
       "3                                IG_J_gene       NaN        NaN     18.0\n",
       "4                          IG_J_pseudogene       NaN        NaN      3.0\n",
       "5                                IG_V_gene       NaN        2.0    143.0\n",
       "6                          IG_V_pseudogene       NaN        3.0    184.0\n",
       "7                            IG_pseudogene       NaN        1.0      NaN\n",
       "8                                  Mt_rRNA       NaN        NaN      2.0\n",
       "9                                  Mt_tRNA       NaN        NaN     22.0\n",
       "10                                     TEC       NaN     1031.0     26.0\n",
       "11                               TR_C_gene       NaN        NaN      6.0\n",
       "12                               TR_D_gene       NaN        NaN      4.0\n",
       "13                               TR_J_gene       NaN        NaN     79.0\n",
       "14                         TR_J_pseudogene       NaN        NaN      4.0\n",
       "15                               TR_V_gene       NaN        NaN    106.0\n",
       "16                         TR_V_pseudogene       NaN        NaN     33.0\n",
       "17                                  lncRNA       NaN    11163.0   5726.0\n",
       "18                                   miRNA       NaN       25.0   1854.0\n",
       "19                                misc_RNA       NaN     1175.0   1036.0\n",
       "20                  polymorphic_pseudogene       NaN        NaN     48.0\n",
       "21                    processed_pseudogene       NaN     2973.0   7187.0\n",
       "22                          protein_coding       NaN      505.0  19439.0\n",
       "23                              pseudogene       NaN       18.0      NaN\n",
       "24                                    rRNA       NaN        3.0     44.0\n",
       "25                         rRNA_pseudogene       NaN        NaN    496.0\n",
       "26                                ribozyme       NaN        3.0      5.0\n",
       "27                                    sRNA       NaN        4.0      1.0\n",
       "28                                   scRNA       NaN        NaN      1.0\n",
       "29                                  scaRNA       NaN       29.0     20.0\n",
       "30                                   snRNA       NaN       71.0   1830.0\n",
       "31                                  snoRNA       NaN      501.0    442.0\n",
       "32        transcribed_processed_pseudogene       NaN      159.0    341.0\n",
       "33          transcribed_unitary_pseudogene       NaN       18.0    120.0\n",
       "34      transcribed_unprocessed_pseudogene       NaN      213.0    725.0\n",
       "35         translated_processed_pseudogene       NaN        NaN      2.0\n",
       "36       translated_unprocessed_pseudogene       NaN        NaN      1.0\n",
       "37                      unitary_pseudogene       NaN       19.0     79.0\n",
       "38                  unprocessed_pseudogene       NaN      926.0   1686.0\n",
       "39                               vault_RNA       NaN        NaN      1.0"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.rename(columns=str).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T20:51:44.652466Z",
     "start_time": "2021-03-02T20:51:44.644248Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9.0"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.iloc[1]['count']['both']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
